{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "35cac7c0-4912-4bdb-8c0d-2a9dd9768097",
   "metadata": {},
   "source": [
    "# Generative models for protein sequences\n",
    "## Authors: Eren Bardak, Daniel Gonzalez\n",
    "### In this notebook we investigate two different methods for generating novel protein sequences. In the first part, we extended and fine-tuned the [Struct2Seq](https://github.com/jingraham/neurips19-graph-protein-design) model which leverages a Transformer and Message Passing Neural Network (instead of attention) to predict novel [Multiple Sequence Alignments](https://en.wikipedia.org/wiki/Multiple_sequence_alignment) from a given protein sequence. In the second part, we extend and fine-tune [ProtGPT2](https://huggingface.co/nferruz/ProtGPT2/blob/main/README.md) to predict completely novel protein sequences.\n",
    "Install required packages before beginning:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e550d507-e54f-46e3-b9ed-da4ff0006a02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install biopython\n",
    "# !pip install nglview\n",
    "# !conda install -c conda-forge nglview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e6f2f82-892d-4b85-9618-271da9e1024b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "14954185b0504807880bf262e6ec5235",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "import nglview as nv\n",
    "from Bio.PDB import PDBParser\n",
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import nglview as nv\n",
    "from IPython.display import display as ipy_display\n",
    "import ipywidgets as widgets\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55d35c76-33c5-469e-8b0f-8f7b5064e8fc",
   "metadata": {},
   "source": [
    "## Dataset download\n",
    "### The [ProteinNet CASP12 dataset](https://github.com/aqlaboratory/proteinnet/tree/master) is large ASCII encoded file of protein sequences paired with PSSM and other structural information. Documentation is available at the above link. First we will need to parse this data into a useable format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2b52ff76-1806-419d-8ad8-042ae837ed80",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_protein_data(file_path, n):\n",
    "    # Initialize lists to store data\n",
    "    ids = []\n",
    "    sequences = []\n",
    "    evolutionary = []\n",
    "    tertiary = []\n",
    "    masks = []\n",
    "    \n",
    "    # Initialize counters and temporary storage for current protein data\n",
    "    current_protein = {}\n",
    "    count = 0\n",
    "    \n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in tqdm(file):\n",
    "            line = line.strip()\n",
    "            if line.startswith('[ID]'):\n",
    "                # Save previous protein data if exists\n",
    "                if current_protein:\n",
    "                    ids.append(current_protein.get('ID', ''))\n",
    "                    sequences.append(current_protein.get('PRIMARY', ''))\n",
    "                    evolutionary.append(current_protein.get('EVOLUTIONARY', []))\n",
    "                    tertiary.append(current_protein.get('TERTIARY', []))\n",
    "                    masks.append(current_protein.get('MASK', ''))\n",
    "                    count += 1\n",
    "                    # Break the loop if the required number of proteins have been parsed\n",
    "                    if count >= n:\n",
    "                        break\n",
    "                    current_protein = {}\n",
    "            \n",
    "            # Identify the section and append the data to the current protein\n",
    "            if line.startswith('['):\n",
    "                key = line[1:line.find(']')]\n",
    "                current_protein[key] = []\n",
    "            elif current_protein:\n",
    "                # Split TSV data into lists of floats or keep as strings depending on the section\n",
    "                if key in ['EVOLUTIONARY', 'TERTIARY']:\n",
    "                    # Split the line by spaces, convert each item to float\n",
    "                    current_protein[key].append([float(x) for x in line.split()])\n",
    "                else:\n",
    "                    current_protein[key].append(line)\n",
    "                \n",
    "    # Check if last processed protein needs to be added\n",
    "    if count < n and current_protein:\n",
    "        ids.append(current_protein.get('ID', ''))\n",
    "        sequences.append(current_protein.get('PRIMARY', ''))\n",
    "        evolutionary.append(current_protein.get('EVOLUTIONARY', []))\n",
    "        tertiary.append(current_protein.get('TERTIARY', []))\n",
    "        masks.append(current_protein.get('MASK', ''))\n",
    "    \n",
    "    # Create DataFrame\n",
    "    df = pd.DataFrame({\n",
    "        'ID': ids,\n",
    "        'Sequence': sequences,\n",
    "        'Evolutionary': evolutionary,\n",
    "        'Tertiary': tertiary,\n",
    "        'Mask': masks\n",
    "    })\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76d18c6c-35c4-4af4-aed6-a83bab3ce381",
   "metadata": {},
   "source": [
    "### Load the data into a df. We used the training_100 for fine tuning, but loaded the training_30 (which contains 30% of the full dataset) as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d658b9d9-e92e-4c53-a4fa-800342c5cdf9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "834867it [00:29, 28693.81it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Evolutionary</th>\n",
       "      <th>Tertiary</th>\n",
       "      <th>Mask</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[4W8Q_1_A]</td>\n",
       "      <td>[NGIVPDAGHQGPDVSAVNGGTQVINIVTPNNEGISHNQYQDFNVG...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 0.003620273531777957, 0.05267...</td>\n",
       "      <td>[[8667.0, 8744.7, 8693.8, 8561.7, 8507.4, 8541...</td>\n",
       "      <td>[+++++++++++++++++++++++++++++++++++++++++++++...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[2J58_1_A]</td>\n",
       "      <td>[ATIIPGQGLNSLRKNVVELPDSDYDLDKLVNVYPMTPGLIDQLRP...</td>\n",
       "      <td>[[0.05, 0.0, 0.0, 0.013847675568743818, 0.0, 0...</td>\n",
       "      <td>[[2309.3, 2397.7, 2539.4, 2627.5, 2769.0, 2861...</td>\n",
       "      <td>[+++++++++++++++++++++++++++++++++++++++++++++...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                           Sequence  \\\n",
       "0  [4W8Q_1_A]  [NGIVPDAGHQGPDVSAVNGGTQVINIVTPNNEGISHNQYQDFNVG...   \n",
       "1  [2J58_1_A]  [ATIIPGQGLNSLRKNVVELPDSDYDLDKLVNVYPMTPGLIDQLRP...   \n",
       "\n",
       "                                        Evolutionary  \\\n",
       "0  [[0.0, 0.0, 0.0, 0.003620273531777957, 0.05267...   \n",
       "1  [[0.05, 0.0, 0.0, 0.013847675568743818, 0.0, 0...   \n",
       "\n",
       "                                            Tertiary  \\\n",
       "0  [[8667.0, 8744.7, 8693.8, 8561.7, 8507.4, 8541...   \n",
       "1  [[2309.3, 2397.7, 2539.4, 2627.5, 2769.0, 2861...   \n",
       "\n",
       "                                                Mask  \n",
       "0  [+++++++++++++++++++++++++++++++++++++++++++++...  \n",
       "1  [+++++++++++++++++++++++++++++++++++++++++++++...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = './data/training_30'\n",
    "number_of_proteins = 50000  # Set how many proteins you want to load\n",
    "protein_df = parse_protein_data(file_path, number_of_proteins)\n",
    "protein_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "26dfc59c-16db-48ff-a39f-906e12b8e06e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins loaded: 25299\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of proteins loaded:\",len(protein_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "891cf7e2-2ed4-4bfb-b2bd-50bca1286655",
   "metadata": {},
   "source": [
    "### The secondary parser updates the types and dimensions of several variables to allow for easier processing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7c4627db-4a1c-4f8d-b56c-0cf992d370c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_dataframe(df):\n",
    "    # Convert ID, Sequence, and Mask to string directly\n",
    "    df['ID'] = df['ID'].apply(lambda x: ''.join(x))\n",
    "    df['Sequence'] = df['Sequence'].apply(lambda x: ''.join(x))\n",
    "    df['Mask'] = df['Mask'].apply(lambda x: ''.join(x).replace('+', '1').replace('-', '0'))\n",
    "\n",
    "    # Initialize columns for coordinates and amino acids\n",
    "    coord_columns = ['x', 'y', 'z']\n",
    "    aa_columns = ['A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','Y','Info']\n",
    "\n",
    "    # Expand tertiary coordinates\n",
    "    for i, col in enumerate(coord_columns):\n",
    "        df[col] = df['Tertiary'].apply(lambda x: x[i] if len(x) > i else None)\n",
    "\n",
    "    # Expand evolutionary data into separate columns for each amino acid position\n",
    "    for i, col in enumerate(aa_columns):\n",
    "        df[col] = df['Evolutionary'].apply(lambda x: x[i] if len(x) > i else None)\n",
    "\n",
    "    # Drop the original Tertiary and Evolutionary columns\n",
    "    df.drop(['Tertiary', 'Evolutionary'], axis=1, inplace=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e1c3fe4f-345f-4441-817a-6c408acd6631",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Mask</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>...</th>\n",
       "      <th>N</th>\n",
       "      <th>P</th>\n",
       "      <th>Q</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "      <th>Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4W8Q_1_A</td>\n",
       "      <td>NGIVPDAGHQGPDVSAVNGGTQVINIVTPNNEGISHNQYQDFNVGK...</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "      <td>[8667.0, 8744.7, 8693.8, 8561.7, 8507.4, 8541....</td>\n",
       "      <td>[220.7, 313.5, 320.5, 325.9, 345.5, 481.0, 583...</td>\n",
       "      <td>[2297.8, 2216.1, 2071.6, 2056.4, 1922.2, 1860....</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.003620273531777957, 0.052679...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.007...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0778237933244589, 0.159...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.015286944949013542...</td>\n",
       "      <td>...</td>\n",
       "      <td>[0.9999999999999999, 0.0, 0.0, 0.0, 0.03705073...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.6938815757037914, 0.072...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.006507913030616773, 0.0...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.001870402676114598...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.005743726273233743, 0.0...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.004424778761061947, 0.000640...</td>\n",
       "      <td>[0.0, 0.0, 0.2848732624693377, 0.7300884955752...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.001...</td>\n",
       "      <td>[0.9999999988691709, 0.9999999988691709, 0.683...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2J58_1_A</td>\n",
       "      <td>ATIIPGQGLNSLRKNVVELPDSDYDLDKLVNVYPMTPGLIDQLRPE...</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "      <td>[2309.3, 2397.7, 2539.4, 2627.5, 2769.0, 2861....</td>\n",
       "      <td>[12053.4, 11937.9, 11979.8, 11789.8, 11789.8, ...</td>\n",
       "      <td>[1185.1, 1201.8, 1162.6, 1307.8, 1353.3, 1234....</td>\n",
       "      <td>[0.05, 0.0, 0.0, 0.013847675568743818, 0.0, 0....</td>\n",
       "      <td>[0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "      <td>[0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0190545...</td>\n",
       "      <td>[0.05, 0.0, 0.0, 0.010880316518298716, 0.0, 0....</td>\n",
       "      <td>...</td>\n",
       "      <td>[0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.012617083741...</td>\n",
       "      <td>[0.05, 0.0, 0.0, 0.0, 0.987119353442055, 0.0, ...</td>\n",
       "      <td>[0.05, 0.0, 0.0, 0.003297065611605671, 0.0, 0....</td>\n",
       "      <td>[0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0272455...</td>\n",
       "      <td>[0.05, 0.0, 0.0, 0.005793415288964251, 0.0, 0....</td>\n",
       "      <td>[0.05, 1.0, 0.0, 0.00923178371249588, 0.005159...</td>\n",
       "      <td>[0.05, 0.0, 0.4501113275814083, 0.357590316047...</td>\n",
       "      <td>[0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "      <td>[0.05, 0.0, 0.0, 0.0012246243700249637, 0.0, 0...</td>\n",
       "      <td>[0.0, 0.9999999988691709, 0.6918268592985294, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID                                           Sequence  \\\n",
       "0  4W8Q_1_A  NGIVPDAGHQGPDVSAVNGGTQVINIVTPNNEGISHNQYQDFNVGK...   \n",
       "1  2J58_1_A  ATIIPGQGLNSLRKNVVELPDSDYDLDKLVNVYPMTPGLIDQLRPE...   \n",
       "\n",
       "                                                Mask  \\\n",
       "0  1111111111111111111111111111111111111111111111...   \n",
       "1  1111111111111111111111111111111111111111111111...   \n",
       "\n",
       "                                                   x  \\\n",
       "0  [8667.0, 8744.7, 8693.8, 8561.7, 8507.4, 8541....   \n",
       "1  [2309.3, 2397.7, 2539.4, 2627.5, 2769.0, 2861....   \n",
       "\n",
       "                                                   y  \\\n",
       "0  [220.7, 313.5, 320.5, 325.9, 345.5, 481.0, 583...   \n",
       "1  [12053.4, 11937.9, 11979.8, 11789.8, 11789.8, ...   \n",
       "\n",
       "                                                   z  \\\n",
       "0  [2297.8, 2216.1, 2071.6, 2056.4, 1922.2, 1860....   \n",
       "1  [1185.1, 1201.8, 1162.6, 1307.8, 1353.3, 1234....   \n",
       "\n",
       "                                                   A  \\\n",
       "0  [0.0, 0.0, 0.0, 0.003620273531777957, 0.052679...   \n",
       "1  [0.05, 0.0, 0.0, 0.013847675568743818, 0.0, 0....   \n",
       "\n",
       "                                                   C  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.007...   \n",
       "1  [0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...   \n",
       "\n",
       "                                                   D  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0778237933244589, 0.159...   \n",
       "1  [0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0190545...   \n",
       "\n",
       "                                                   E  ...  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.015286944949013542...  ...   \n",
       "1  [0.05, 0.0, 0.0, 0.010880316518298716, 0.0, 0....  ...   \n",
       "\n",
       "                                                   N  \\\n",
       "0  [0.9999999999999999, 0.0, 0.0, 0.0, 0.03705073...   \n",
       "1  [0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.012617083741...   \n",
       "\n",
       "                                                   P  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.6938815757037914, 0.072...   \n",
       "1  [0.05, 0.0, 0.0, 0.0, 0.987119353442055, 0.0, ...   \n",
       "\n",
       "                                                   Q  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.006507913030616773, 0.0...   \n",
       "1  [0.05, 0.0, 0.0, 0.003297065611605671, 0.0, 0....   \n",
       "\n",
       "                                                   R  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.001870402676114598...   \n",
       "1  [0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0272455...   \n",
       "\n",
       "                                                   S  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.005743726273233743, 0.0...   \n",
       "1  [0.05, 0.0, 0.0, 0.005793415288964251, 0.0, 0....   \n",
       "\n",
       "                                                   T  \\\n",
       "0  [0.0, 0.0, 0.0, 0.004424778761061947, 0.000640...   \n",
       "1  [0.05, 1.0, 0.0, 0.00923178371249588, 0.005159...   \n",
       "\n",
       "                                                   V  \\\n",
       "0  [0.0, 0.0, 0.2848732624693377, 0.7300884955752...   \n",
       "1  [0.05, 0.0, 0.4501113275814083, 0.357590316047...   \n",
       "\n",
       "                                                   W  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...   \n",
       "\n",
       "                                                   Y  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.001...   \n",
       "1  [0.05, 0.0, 0.0, 0.0012246243700249637, 0.0, 0...   \n",
       "\n",
       "                                                Info  \n",
       "0  [0.9999999988691709, 0.9999999988691709, 0.683...  \n",
       "1  [0.0, 0.9999999988691709, 0.6918268592985294, ...  \n",
       "\n",
       "[2 rows x 27 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example usage:\n",
    "processed_df = process_dataframe(protein_df)\n",
    "processed_df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da087be6-acdc-41c8-b422-c1fc03c203a1",
   "metadata": {},
   "source": [
    "## Data description:\n",
    "- ID is the protein name\n",
    "- Sequence is the amino acid sequence using the single letter code\n",
    "- Mask (same length as sequence) is a binary representation of how importatnt each amino acid is to the structure (that is, amino acids marked `0` can be changed without affecting the structure much, whereas amino acids marked `1` would significantly alter structure/function).\n",
    "- XYZ are the coordinates of each atom (note there are 3 atoms per amino acid, Nitrogen, alpha-Carbon, and Carbon, so there are 3 times as many coordinates as amino acids in the sequence)\n",
    "- The 20 columns A-Y are normalized PSSM information for each amino acid. This is essentially a sparse matrix where there is a row for each position in the protein sequence, and a column for each of the 20 amino acids.\n",
    "- Info contains the information content for that residue (unclear what this actually is)\n",
    "## Note:\n",
    "Proteins are chains of amino acids, and each amino acid has a backbone of three atoms, N-C-C, which are always linked in the N to C direction: \n",
    "\n",
    "`Beginning of protein sequence amino acids:       1       2       3    ... etc`\n",
    "\n",
    "`Beginning of protein sequence backbone atoms: (N-C-C)-(N-C-C)-(N-C-C)-... etc`\n",
    "\n",
    "So in our final output, we need to have xyz coordinates for each of the three atoms, in order, for every amino acid."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "417a2857-aef0-4fa1-b5da-c8e8b02d5701",
   "metadata": {},
   "source": [
    "## The third parsing step turns a row of our dataframe into a PDB file that can be viewed with NGL. Note that our dataset doesn't have info on the secondary structure, so these structures will not appear bolded in the output. \n",
    "\n",
    "The PDB specification is available here: https://www.biostat.jhsph.edu/~iruczins/teaching/260.655/links/pdbformat.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8f4d1567-519d-41d4-a427-bead0e15e95c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_pdb_from_df(row):\n",
    "    pdb_str = \"\"\n",
    "    atom_count = 1  # Starting index for ATOM records\n",
    "    chain_id = 'A'  # Assuming a single chain for simplicity\n",
    "    res_num = 1     # Residue number\n",
    "    occupancy = 1.00\n",
    "    t_factor = 50.00\n",
    "\n",
    "    atom_types = ['N', 'CA', 'C']  # Backbone atoms\n",
    "    element_types = {'N': 'N', 'CA': 'C', 'C': 'C'}  # Elements for atoms\n",
    "    residue_mapping = {'A': 'ALA','R': 'ARG','N': 'ASN','D': 'ASP','C': 'CYS','E': 'GLU','Q': 'GLN','G': 'GLY',\n",
    "                       'H': 'HIS','I': 'ILE','L': 'LEU','K': 'LYS','M': 'MET','F': 'PHE','P': 'PRO','S': 'SER',\n",
    "                       'T': 'THR','W': 'TRP','Y': 'TYR','V': 'VAL' } # PDB requires the 3 letter codes for amino acids\n",
    "\n",
    "    for i in range(len(row['Sequence'])):\n",
    "        residue = residue_mapping[row['Sequence'][i]]\n",
    "        # Loop over each backbone atom type\n",
    "        for j, atom_type in enumerate(atom_types):\n",
    "            # Index to pull the correct coordinates from flattened list\n",
    "            idx = 3 * i + j\n",
    "            x = row['x'][idx]/100\n",
    "            y = row['y'][idx]/100\n",
    "            z = row['z'][idx]/100\n",
    "            element = element_types[atom_type]\n",
    "\n",
    "            pdb_str += f\"ATOM  {atom_count:>5}  {atom_type:<2}  {residue:>3} {chain_id}{res_num:>4}    {x:>8.3f}{y:>8.3f}{z:>8.3f}{occupancy:>6.2f}{t_factor:>6.2f}          {element:>2}  \\n\"\n",
    "            atom_count += 1\n",
    "\n",
    "        res_num += 1  # Increment residue number for each new amino acid\n",
    "\n",
    "    return pdb_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c1dfffb0-3e2f-4313-a6bd-962a1da2f662",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATOM      1  N   SER A   1       0.000   0.000   0.000  1.00 50.00           N  \n",
      "ATOM      2  CA  SER A   1       0.000   0.000   0.000  1.00 50.00           C  \n",
      "ATOM      3  C   SER A   1       0.000   0.000   0.000  1.00 50.00           C  \n",
      "ATOM      4  N   ASN A   2       0.000   0.000   0.000  1.00 50.00           N  \n",
      "ATOM      5  CA  ASN A   2       0.000   0.000   0.000  1.00 50.00           C  \n",
      "ATOM      6  C   ASN A   2       0.000   0.000   0.000  1.00 50.00           C  \n",
      "ATOM      7  N   ALA A   3       0.000   0.000   0.000  1.00 50.00           N  \n",
      "ATOM      8  CA  ALA A   3       0.000   0.000   0.000  1.00 50.00           C  \n",
      "ATOM      9  C   ALA A   3       0.000   0.000   0.000  1.00 50.00           C  \n",
      "ATOM     10  N   MET A   4      48.047  73.486  30.296  1.00 50.00           N  \n",
      "ATOM     11  CA  MET A   4      47.844  72.216  30.986  1.00 50.00           C  \n",
      "ATOM     12  C   MET A   4      46.701  71.429  30.361  1.00 50.00           C  \n",
      "ATOM     13  N   TYR A   5      47.012  70.220  29.912  1.00 50.00           N  \n",
      "ATOM     14  CA  TYR A   5      46.012  69.355  29.306  1.00 50.00           C  \n",
      "ATOM     15  C   TYR A   5      45.507  68.332  30.301  1.00 50.00           C  \n",
      "ATOM     16  N   LYS A   6      44.236  67.981  30.160  1.00 50.00           N  \n",
      "ATOM     17  CA  LYS A   6      43.703  66.872  30.908  1.00 50.00           C  \n",
      "ATOM     18  C   LYS A   6      43.695  65.642  30.000  1.00 50.00           C  \n",
      "ATOM     19  N   ILE A   7      44.608  64.714  30.277  1.00 50.00           N  \n",
      "ATOM     20  CA  ILE A   7      44.712  63.468  29.515  1.00 50.00           C  \n",
      "ATOM     21  C   ILE A   7      43.684  62.448  30.016  1.00 50.00           C  \n",
      "ATOM     22  N   VAL A   8      42.918  61.878  29.087  1.00 50.00           N  \n",
      "ATOM     23  CA  VAL A   8      41.915  60.873  29.410  1.00 50.00           C  \n",
      "ATOM     24  C   VAL A   8      42.171  59.607  28.575  1.00 50.00           C  \n",
      "ATOM     25  N   SER A   9      41.883  58.443  29.145  1.00 50.00           N  \n",
      "ATOM     26  CA  SER A   9      41.941  57.191  28.381  1.00 50.00           C  \n",
      "ATOM     27  C   SER A   9      40.817  56.290  28.840  1.00 50.00           C  \n",
      "ATOM     28  N   ASP A  10      40.404  55.353  27.983  1.00 50.00           N  \n",
      "ATOM     29  CA  ASP A  10      39.502  54.301  28.421  1.00 50.00           C  \n",
      "ATOM     30  C   ASP A  10      40.325  53.290  29.202  1.00 50.00           C  \n",
      "ATOM     31  N   SER A  11      39.651  52.425  29.943  1.00 50.00           N  \n",
      "ATOM     32  CA  SER A  11      40.357  51.562  30.891  1.00 50.00           C  \n",
      "ATOM     33  C   SER A  11      41.171  50.442  30.245  1.00 50.00           C  \n",
      "ATOM     34  N   ALA A  12      41.163  50.363  28.917  1.00 50.00           N  \n",
      "ATOM     35  CA  ALA A  12      41.926  49.313  28.241  1.00 50.00           C  \n",
      "ATOM     36  C   ALA A  12      43.430  49.533  28.361  1.00 50.00           C  \n",
      "ATOM     37  N   CYS A  13      43.848  50.745  28.732  1.00 50.00           N  \n",
      "ATOM     38  CA  CYS A  13      45.291  51.074  28.777  1.00 50.00           C  \n",
      "ATOM     39  C   CYS A  13      46.022  50.393  29.947  1.00 50.00           C  \n",
      "ATOM     40  N   ASP A  14      45.276  50.026  30.984  1.00 50.00           N  \n",
      "ATOM     41  CA  ASP A  14      45.843  49.261  32.109  1.00 50.00           C  \n",
      "ATOM     42  C   ASP A  14      46.959  49.991  32.891  1.00 50.00           C  \n",
      "ATOM     43  N   LEU A  15      47.005  51.313  32.798  1.00 50.00           N  \n",
      "ATOM     44  CA  LEU A  15      47.989  52.087  33.548  1.00 50.00           C  \n",
      "ATOM     45  C   LEU A  15      47.787  51.912  35.046  1.00 50.00           C  \n",
      "ATOM     46  N   SER A  16      48.885  51.816  35.791  1.00 50.00           N  \n",
      "ATOM     47  CA  SER A  16      48.801  51.595  37.237  1.00 50.00           C  \n",
      "ATOM     48  C   SER A  16      48.116  52.738  37.992  1.00 50.00           C  \n",
      "ATOM     49  N   LYS A  17      47.538  52.422  39.149  1.00 50.00           N  \n",
      "ATOM     50  CA  LYS A  17      46.979  53.447  40.033  1.00 50.00           C  \n",
      "ATOM     51  C   LYS A  17      47.991  54.559  40.288  1.00 50.00           C  \n",
      "ATOM     52  N   GLU A  18      49.227  54.164  40.588  1.00 50.00           N  \n",
      "ATOM     53  CA  GLU A  18      50.318  55.111  40.829  1.00 50.00           C  \n",
      "ATOM     54  C   GLU A  18      50.543  56.054  39.655  1.00 50.00           C  \n",
      "ATOM     55  N   TYR A  19      50.547  55.500  38.445  1.00 50.00           N  \n",
      "ATOM     56  CA  TYR A  19      50.842  56.294  37.263  1.00 50.00           C  \n",
      "ATOM     57  C   TYR A  19      49.754  57.340  37.071  1.00 50.00           C  \n",
      "ATOM     58  N   LEU A  20      48.504  56.922  37.261  1.00 50.00           N  \n",
      "ATOM     59  CA  LEU A  20      47.362  57.808  37.067  1.00 50.00           C  \n",
      "ATOM     60  C   LEU A  20      47.437  59.019  37.996  1.00 50.00           C  \n",
      "ATOM     61  N   GLU A  21      47.750  58.776  39.262  1.00 50.00           N  \n",
      "ATOM     62  CA  GLU A  21      47.868  59.865  40.222  1.00 50.00           C  \n",
      "ATOM     63  C   GLU A  21      49.060  60.761  39.896  1.00 50.00           C  \n",
      "ATOM     64  N   LYS A  22      50.203  60.144  39.633  1.00 50.00           N  \n",
      "ATOM     65  CA  LYS A  22      51.422  60.889  39.354  1.00 50.00           C  \n",
      "ATOM     66  C   LYS A  22      51.291  61.830  38.166  1.00 50.00           C  \n",
      "ATOM     67  N   HIS A  23      50.675  61.352  37.085  1.00 50.00           N  \n",
      "ATOM     68  CA  HIS A  23      50.627  62.116  35.841  1.00 50.00           C  \n",
      "ATOM     69  C   HIS A  23      49.258  62.712  35.504  1.00 50.00           C  \n",
      "ATOM     70  N   ASP A  24      48.331  62.659  36.456  1.00 50.00           N  \n",
      "ATOM     71  CA  ASP A  24      47.007  63.251  36.296  1.00 50.00           C  \n",
      "ATOM     72  C   ASP A  24      46.249  62.726  35.063  1.00 50.00           C  \n",
      "ATOM     73  N   VAL A  25      46.185  61.405  34.921  1.00 50.00           N  \n",
      "ATOM     74  CA  VAL A  25      45.438  60.779  33.829  1.00 50.00           C  \n",
      "ATOM     75  C   VAL A  25      44.107  60.259  34.365  1.00 50.00           C  \n",
      "ATOM     76  N   THR A  26      43.033  60.562  33.650  1.00 50.00           N  \n",
      "ATOM     77  CA  THR A  26      41.689  60.176  34.054  1.00 50.00           C  \n",
      "ATOM     78  C   THR A  26      41.234  58.994  33.202  1.00 50.00           C  \n",
      "ATOM     79  N   ILE A  27      40.643  57.991  33.843  1.00 50.00           N  \n",
      "ATOM     80  CA  ILE A  27      40.190  56.790  33.139  1.00 50.00           C  \n",
      "ATOM     81  C   ILE A  27      38.680  56.729  32.975  1.00 50.00           C  \n",
      "ATOM     82  N   VAL A  28      38.234  56.440  31.752  1.00 50.00           N  \n",
      "ATOM     83  CA  VAL A  28      36.824  56.129  31.490  1.00 50.00           C  \n",
      "ATOM     84  C   VAL A  28      36.692  54.613  31.344  1.00 50.00           C  \n",
      "ATOM     85  N   PRO A  29      35.966  53.983  32.272  1.00 50.00           N  \n",
      "ATOM     86  CA  PRO A  29      35.923  52.512  32.327  1.00 50.00           C  \n",
      "ATOM     87  C   PRO A  29      35.069  51.825  31.257  1.00 50.00           C  \n",
      "ATOM     88  N   LEU A  30      35.573  50.706  30.741  1.00 50.00           N  \n",
      "ATOM     89  CA  LEU A  30      34.711  49.751  30.048  1.00 50.00           C  \n",
      "ATOM     90  C   LEU A  30      33.990  48.923  31.109  1.00 50.00           C  \n",
      "ATOM     91  N   SER A  31      33.053  48.075  30.693  1.00 50.00           N  \n",
      "ATOM     92  CA  SER A  31      32.356  47.208  31.643  1.00 50.00           C  \n",
      "ATOM     93  C   SER A  31      32.525  45.741  31.284  1.00 50.00           C  \n",
      "ATOM     94  N   VAL A  32      32.352  44.887  32.283  1.00 50.00           N  \n",
      "ATOM     95  CA  VAL A  32      32.359  43.447  32.072  1.00 50.00           C  \n",
      "ATOM     96  C   VAL A  32      31.092  42.817  32.664  1.00 50.00           C  \n",
      "ATOM     97  N   SER A  33      30.611  41.740  32.042  1.00 50.00           N  \n",
      "ATOM     98  CA  SER A  33      29.459  40.999  32.553  1.00 50.00           C  \n",
      "ATOM     99  C   SER A  33      29.762  39.512  32.524  1.00 50.00           C  \n",
      "ATOM    100  N   PHE A  34      29.348  38.809  33.571  1.00 50.00           N  \n",
      "ATOM    101  CA  PHE A  34      29.537  37.368  33.635  1.00 50.00           C  \n",
      "ATOM    102  C   PHE A  34      28.305  36.588  33.199  1.00 50.00           C  \n",
      "ATOM    103  N   ASP A  35      27.133  37.197  33.334  1.00 50.00           N  \n",
      "ATOM    104  CA  ASP A  35      25.894  36.523  32.963  1.00 50.00           C  \n",
      "ATOM    105  C   ASP A  35      25.249  37.118  31.719  1.00 50.00           C  \n",
      "ATOM    106  N   GLY A  36      25.827  38.202  31.209  1.00 50.00           N  \n",
      "ATOM    107  CA  GLY A  36      25.266  38.888  30.059  1.00 50.00           C  \n",
      "ATOM    108  C   GLY A  36      24.110  39.818  30.386  1.00 50.00           C  \n",
      "ATOM    109  N   GLU A  37      23.746  39.904  31.665  1.00 50.00           N  \n",
      "ATOM    110  CA  GLU A  37      22.666  40.788  32.097  1.00 50.00           C  \n",
      "ATOM    111  C   GLU A  37      23.197  41.957  32.913  1.00 50.00           C  \n",
      "ATOM    112  N   THR A  38      23.909  41.638  33.986  1.00 50.00           N  \n",
      "ATOM    113  CA  THR A  38      24.469  42.660  34.854  1.00 50.00           C  \n",
      "ATOM    114  C   THR A  38      25.908  42.979  34.444  1.00 50.00           C  \n",
      "ATOM    115  N   TYR A  39      26.211  44.273  34.304  1.00 50.00           N  \n",
      "ATOM    116  CA  TYR A  39      27.557  44.740  33.954  1.00 50.00           C  \n",
      "ATOM    117  C   TYR A  39      28.186  45.546  35.084  1.00 50.00           C  \n",
      "ATOM    118  N   TYR A  40      29.510  45.450  35.181  1.00 50.00           N  \n",
      "ATOM    119  CA  TYR A  40      30.311  46.113  36.206  1.00 50.00           C  \n",
      "ATOM    120  C   TYR A  40      31.418  46.920  35.557  1.00 50.00           C  \n",
      "ATOM    121  N   ARG A  41      31.507  48.206  35.905  1.00 50.00           N  \n",
      "ATOM    122  CA  ARG A  41      32.555  49.069  35.390  1.00 50.00           C  \n",
      "ATOM    123  C   ARG A  41      33.938  48.651  35.872  1.00 50.00           C  \n",
      "ATOM    124  N   ASP A  42      34.814  48.445  34.892  1.00 50.00           N  \n",
      "ATOM    125  CA  ASP A  42      36.191  47.995  35.079  1.00 50.00           C  \n",
      "ATOM    126  C   ASP A  42      36.906  48.806  36.156  1.00 50.00           C  \n",
      "ATOM    127  N   GLY A  43      37.354  48.137  37.206  1.00 50.00           N  \n",
      "ATOM    128  CA  GLY A  43      38.129  48.815  38.225  1.00 50.00           C  \n",
      "ATOM    129  C   GLY A  43      37.295  49.616  39.214  1.00 50.00           C  \n",
      "ATOM    130  N   VAL A  44      36.068  49.968  38.833  1.00 50.00           N  \n",
      "ATOM    131  CA  VAL A  44      35.206  50.779  39.694  1.00 50.00           C  \n",
      "ATOM    132  C   VAL A  44      34.200  49.909  40.446  1.00 50.00           C  \n",
      "ATOM    133  N   ASP A  45      33.419  49.119  39.711  1.00 50.00           N  \n",
      "ATOM    134  CA  ASP A  45      32.413  48.241  40.322  1.00 50.00           C  \n",
      "ATOM    135  C   ASP A  45      32.962  46.840  40.591  1.00 50.00           C  \n",
      "ATOM    136  N   ILE A  46      34.173  46.579  40.120  1.00 50.00           N  \n",
      "ATOM    137  CA  ILE A  46      34.787  45.265  40.263  1.00 50.00           C  \n",
      "ATOM    138  C   ILE A  46      36.276  45.428  40.042  1.00 50.00           C  \n",
      "ATOM    139  N   THR A  47      37.078  44.541  40.620  1.00 50.00           N  \n",
      "ATOM    140  CA  THR A  47      38.530  44.675  40.497  1.00 50.00           C  \n",
      "ATOM    141  C   THR A  47      39.116  43.702  39.490  1.00 50.00           C  \n",
      "ATOM    142  N   ARG A  48      40.292  44.060  38.994  1.00 50.00           N  \n",
      "ATOM    143  CA  ARG A  48      40.978  43.260  38.007  1.00 50.00           C  \n",
      "ATOM    144  C   ARG A  48      41.238  41.852  38.541  1.00 50.00           C  \n",
      "ATOM    145  N   ASP A  49      41.740  41.761  39.768  1.00 50.00           N  \n",
      "ATOM    146  CA  ASP A  49      42.040  40.454  40.339  1.00 50.00           C  \n",
      "ATOM    147  C   ASP A  49      40.803  39.555  40.483  1.00 50.00           C  \n",
      "ATOM    148  N   GLU A  50      39.664  40.132  40.875  1.00 50.00           N  \n",
      "ATOM    149  CA  GLU A  50      38.428  39.360  41.032  1.00 50.00           C  \n",
      "ATOM    150  C   GLU A  50      37.995  38.752  39.710  1.00 50.00           C  \n",
      "ATOM    151  N   CYS A  51      38.088  39.544  38.649  1.00 50.00           N  \n",
      "ATOM    152  CA  CYS A  51      37.685  39.073  37.335  1.00 50.00           C  \n",
      "ATOM    153  C   CYS A  51      38.529  37.882  36.889  1.00 50.00           C  \n",
      "ATOM    154  N   TYR A  52      39.846  37.975  37.060  1.00 50.00           N  \n",
      "ATOM    155  CA  TYR A  52      40.711  36.874  36.673  1.00 50.00           C  \n",
      "ATOM    156  C   TYR A  52      40.317  35.620  37.456  1.00 50.00           C  \n",
      "ATOM    157  N   GLN A  53      39.988  35.809  38.730  1.00 50.00           N  \n",
      "ATOM    158  CA  GLN A  53      39.595  34.694  39.592  1.00 50.00           C  \n",
      "ATOM    159  C   GLN A  53      38.365  33.969  39.064  1.00 50.00           C  \n",
      "ATOM    160  N   ARG A  54      37.328  34.725  38.711  1.00 50.00           N  \n",
      "ATOM    161  CA  ARG A  54      36.113  34.125  38.174  1.00 50.00           C  \n",
      "ATOM    162  C   ARG A  54      36.441  33.230  36.988  1.00 50.00           C  \n",
      "ATOM    163  N   MET A  55      37.324  33.696  36.108  1.00 50.00           N  \n",
      "ATOM    164  CA  MET A  55      37.680  32.897  34.936  1.00 50.00           C  \n",
      "ATOM    165  C   MET A  55      38.524  31.684  35.298  1.00 50.00           C  \n",
      "ATOM    166  N   VAL A  56      39.447  31.861  36.240  1.00 50.00           N  \n",
      "ATOM    167  CA  VAL A  56      40.315  30.765  36.659  1.00 50.00           C  \n",
      "ATOM    168  C   VAL A  56      39.519  29.677  37.394  1.00 50.00           C  \n",
      "ATOM    169  N   ASP A  57      38.563  30.101  38.214  1.00 50.00           N  \n",
      "ATOM    170  CA  ASP A  57      37.705  29.167  38.942  1.00 50.00           C  \n",
      "ATOM    171  C   ASP A  57      36.752  28.416  38.016  1.00 50.00           C  \n",
      "ATOM    172  N   ASP A  58      36.478  28.976  36.842  1.00 50.00           N  \n",
      "ATOM    173  CA  ASP A  58      35.489  28.381  35.945  1.00 50.00           C  \n",
      "ATOM    174  C   ASP A  58      35.892  28.491  34.474  1.00 50.00           C  \n",
      "ATOM    175  N   PRO A  59      36.612  27.479  33.972  1.00 50.00           N  \n",
      "ATOM    176  CA  PRO A  59      37.184  27.447  32.620  1.00 50.00           C  \n",
      "ATOM    177  C   PRO A  59      36.131  27.474  31.518  1.00 50.00           C  \n",
      "ATOM    178  N   LYS A  60      34.862  27.361  31.894  1.00 50.00           N  \n",
      "ATOM    179  CA  LYS A  60      33.768  27.357  30.926  1.00 50.00           C  \n",
      "ATOM    180  C   LYS A  60      33.071  28.712  30.863  1.00 50.00           C  \n",
      "ATOM    181  N   LEU A  61      33.461  29.609  31.762  1.00 50.00           N  \n",
      "ATOM    182  CA  LEU A  61      32.872  30.941  31.861  1.00 50.00           C  \n",
      "ATOM    183  C   LEU A  61      33.490  31.871  30.815  1.00 50.00           C  \n",
      "ATOM    184  N   PHE A  62      32.678  32.385  29.897  1.00 50.00           N  \n",
      "ATOM    185  CA  PHE A  62      33.148  33.398  28.950  1.00 50.00           C  \n",
      "ATOM    186  C   PHE A  62      32.440  34.736  29.187  1.00 50.00           C  \n",
      "ATOM    187  N   PRO A  63      33.111  35.656  29.902  1.00 50.00           N  \n",
      "ATOM    188  CA  PRO A  63      32.496  36.955  30.198  1.00 50.00           C  \n",
      "ATOM    189  C   PRO A  63      32.367  37.804  28.946  1.00 50.00           C  \n",
      "ATOM    190  N   LYS A  64      31.602  38.888  29.043  1.00 50.00           N  \n",
      "ATOM    191  CA  LYS A  64      31.460  39.824  27.936  1.00 50.00           C  \n",
      "ATOM    192  C   LYS A  64      31.945  41.200  28.355  1.00 50.00           C  \n",
      "ATOM    193  N   THR A  65      32.544  41.927  27.418  1.00 50.00           N  \n",
      "ATOM    194  CA  THR A  65      32.929  43.310  27.674  1.00 50.00           C  \n",
      "ATOM    195  C   THR A  65      31.919  44.208  27.000  1.00 50.00           C  \n",
      "ATOM    196  N   SER A  66      31.903  45.463  27.421  1.00 50.00           N  \n",
      "ATOM    197  CA  SER A  66      31.066  46.450  26.781  1.00 50.00           C  \n",
      "ATOM    198  C   SER A  66      31.818  47.772  26.753  1.00 50.00           C  \n",
      "ATOM    199  N   LEU A  67      31.763  48.454  25.607  1.00 50.00           N  \n",
      "ATOM    200  CA  LEU A  67      32.370  49.769  25.427  1.00 50.00           C  \n",
      "ATOM    201  C   LEU A  67      31.795  50.790  26.421  1.00 50.00           C  \n",
      "ATOM    202  N   PRO A  68      32.528  51.889  26.699  1.00 50.00           N  \n",
      "ATOM    203  CA  PRO A  68      31.946  52.878  27.626  1.00 50.00           C  \n",
      "ATOM    204  C   PRO A  68      30.799  53.594  26.945  1.00 50.00           C  \n",
      "ATOM    205  N   SER A  69      29.811  54.038  27.717  1.00 50.00           N  \n",
      "ATOM    206  CA  SER A  69      28.670  54.723  27.139  1.00 50.00           C  \n",
      "ATOM    207  C   SER A  69      29.125  56.101  26.686  1.00 50.00           C  \n",
      "ATOM    208  N   VAL A  70      28.475  56.637  25.661  1.00 50.00           N  \n",
      "ATOM    209  CA  VAL A  70      28.665  58.038  25.289  1.00 50.00           C  \n",
      "ATOM    210  C   VAL A  70      28.546  58.993  26.504  1.00 50.00           C  \n",
      "ATOM    211  N   GLU A  71      27.565  58.764  27.373  1.00 50.00           N  \n",
      "ATOM    212  CA  GLU A  71      27.366  59.649  28.530  1.00 50.00           C  \n",
      "ATOM    213  C   GLU A  71      28.541  59.642  29.507  1.00 50.00           C  \n",
      "ATOM    214  N   SER A  72      29.180  58.482  29.672  1.00 50.00           N  \n",
      "ATOM    215  CA  SER A  72      30.355  58.354  30.547  1.00 50.00           C  \n",
      "ATOM    216  C   SER A  72      31.525  59.232  30.102  1.00 50.00           C  \n",
      "ATOM    217  N   TYR A  73      31.841  59.171  28.819  1.00 50.00           N  \n",
      "ATOM    218  CA  TYR A  73      32.826  60.082  28.247  1.00 50.00           C  \n",
      "ATOM    219  C   TYR A  73      32.363  61.533  28.391  1.00 50.00           C  \n",
      "ATOM    220  N   ALA A  74      31.107  61.792  28.033  1.00 50.00           N  \n",
      "ATOM    221  CA  ALA A  74      30.597  63.159  27.988  1.00 50.00           C  \n",
      "ATOM    222  C   ALA A  74      30.697  63.860  29.346  1.00 50.00           C  \n",
      "ATOM    223  N   ASP A  75      30.333  63.162  30.416  1.00 50.00           N  \n",
      "ATOM    224  CA  ASP A  75      30.426  63.738  31.758  1.00 50.00           C  \n",
      "ATOM    225  C   ASP A  75      31.874  64.097  32.140  1.00 50.00           C  \n",
      "ATOM    226  N   VAL A  76      32.831  63.265  31.746  1.00 50.00           N  \n",
      "ATOM    227  CA  VAL A  76      34.237  63.592  31.984  1.00 50.00           C  \n",
      "ATOM    228  C   VAL A  76      34.665  64.794  31.139  1.00 50.00           C  \n",
      "ATOM    229  N   PHE A  77      34.385  64.750  29.836  1.00 50.00           N  \n",
      "ATOM    230  CA  PHE A  77      34.749  65.871  28.970  1.00 50.00           C  \n",
      "ATOM    231  C   PHE A  77      34.146  67.176  29.491  1.00 50.00           C  \n",
      "ATOM    232  N   ARG A  78      32.869  67.124  29.852  1.00 50.00           N  \n",
      "ATOM    233  CA  ARG A  78      32.139  68.307  30.292  1.00 50.00           C  \n",
      "ATOM    234  C   ARG A  78      32.760  68.947  31.535  1.00 50.00           C  \n",
      "ATOM    235  N   SER A  79      33.080  68.131  32.532  1.00 50.00           N  \n",
      "ATOM    236  CA  SER A  79      33.609  68.659  33.785  1.00 50.00           C  \n",
      "ATOM    237  C   SER A  79      34.956  69.360  33.597  1.00 50.00           C  \n",
      "ATOM    238  N   PHE A  80      35.812  68.811  32.739  1.00 50.00           N  \n",
      "ATOM    239  CA  PHE A  80      37.082  69.461  32.440  1.00 50.00           C  \n",
      "ATOM    240  C   PHE A  80      36.874  70.739  31.624  1.00 50.00           C  \n",
      "ATOM    241  N   VAL A  81      36.036  70.654  30.592  1.00 50.00           N  \n",
      "ATOM    242  CA  VAL A  81      35.806  71.776  29.678  1.00 50.00           C  \n",
      "ATOM    243  C   VAL A  81      35.185  73.019  30.345  1.00 50.00           C  \n",
      "ATOM    244  N   GLU A  82      34.276  72.811  31.287  1.00 50.00           N  \n",
      "ATOM    245  CA  GLU A  82      33.679  73.956  31.965  1.00 50.00           C  \n",
      "ATOM    246  C   GLU A  82      34.697  74.618  32.892  1.00 50.00           C  \n",
      "ATOM    247  N   GLN A  83      35.804  73.925  33.136  1.00 50.00           N  \n",
      "ATOM    248  CA  GLN A  83      36.911  74.491  33.900  1.00 50.00           C  \n",
      "ATOM    249  C   GLN A  83      37.936  75.158  32.985  1.00 50.00           C  \n",
      "ATOM    250  N   GLY A  84      37.706  75.085  31.677  1.00 50.00           N  \n",
      "ATOM    251  CA  GLY A  84      38.592  75.714  30.714  1.00 50.00           C  \n",
      "ATOM    252  C   GLY A  84      39.760  74.850  30.272  1.00 50.00           C  \n",
      "ATOM    253  N   PHE A  85      39.800  73.610  30.753  1.00 50.00           N  \n",
      "ATOM    254  CA  PHE A  85      40.873  72.683  30.388  1.00 50.00           C  \n",
      "ATOM    255  C   PHE A  85      40.668  72.107  28.989  1.00 50.00           C  \n",
      "ATOM    256  N   PRO A  86      41.729  72.122  28.163  1.00 50.00           N  \n",
      "ATOM    257  CA  PRO A  86      41.706  71.375  26.904  1.00 50.00           C  \n",
      "ATOM    258  C   PRO A  86      41.798  69.887  27.247  1.00 50.00           C  \n",
      "ATOM    259  N   VAL A  87      41.172  69.032  26.450  1.00 50.00           N  \n",
      "ATOM    260  CA  VAL A  87      41.128  67.606  26.760  1.00 50.00           C  \n",
      "ATOM    261  C   VAL A  87      41.748  66.776  25.645  1.00 50.00           C  \n",
      "ATOM    262  N   VAL A  88      42.567  65.794  26.011  1.00 50.00           N  \n",
      "ATOM    263  CA  VAL A  88      43.037  64.817  25.043  1.00 50.00           C  \n",
      "ATOM    264  C   VAL A  88      42.676  63.419  25.503  1.00 50.00           C  \n",
      "ATOM    265  N   CYS A  89      41.962  62.684  24.657  1.00 50.00           N  \n",
      "ATOM    266  CA  CYS A  89      41.401  61.402  25.058  1.00 50.00           C  \n",
      "ATOM    267  C   CYS A  89      41.864  60.246  24.163  1.00 50.00           C  \n",
      "ATOM    268  N   PHE A  90      42.499  59.232  24.761  1.00 50.00           N  \n",
      "ATOM    269  CA  PHE A  90      42.965  58.046  24.023  1.00 50.00           C  \n",
      "ATOM    270  C   PHE A  90      42.025  56.861  24.221  1.00 50.00           C  \n",
      "ATOM    271  N   THR A  91      41.790  56.102  23.158  1.00 50.00           N  \n",
      "ATOM    272  CA  THR A  91      40.946  54.925  23.275  1.00 50.00           C  \n",
      "ATOM    273  C   THR A  91      41.666  53.729  22.705  1.00 50.00           C  \n",
      "ATOM    274  N   ILE A  92      41.165  52.556  23.066  1.00 50.00           N  \n",
      "ATOM    275  CA  ILE A  92      41.628  51.309  22.476  1.00 50.00           C  \n",
      "ATOM    276  C   ILE A  92      41.418  51.379  20.952  1.00 50.00           C  \n",
      "ATOM    277  N   THR A  93      42.204  50.614  20.194  1.00 50.00           N  \n",
      "ATOM    278  CA  THR A  93      42.044  50.550  18.735  1.00 50.00           C  \n",
      "ATOM    279  C   THR A  93      40.593  50.367  18.280  1.00 50.00           C  \n",
      "ATOM    280  N   THR A  94      40.194  51.128  17.262  1.00 50.00           N  \n",
      "ATOM    281  CA  THR A  94      38.836  51.046  16.731  1.00 50.00           C  \n",
      "ATOM    282  C   THR A  94      38.553  49.682  16.142  1.00 50.00           C  \n",
      "ATOM    283  N   LEU A  95      39.606  48.934  15.825  1.00 50.00           N  \n",
      "ATOM    284  CA  LEU A  95      39.423  47.582  15.284  1.00 50.00           C  \n",
      "ATOM    285  C   LEU A  95      38.994  46.554  16.334  1.00 50.00           C  \n",
      "ATOM    286  N   PHE A  96      39.107  46.882  17.631  1.00 50.00           N  \n",
      "ATOM    287  CA  PHE A  96      38.667  45.970  18.695  1.00 50.00           C  \n",
      "ATOM    288  C   PHE A  96      37.345  46.411  19.331  1.00 50.00           C  \n",
      "ATOM    289  N   SER A  97      37.041  47.703  19.280  1.00 50.00           N  \n",
      "ATOM    290  CA  SER A  97      35.967  48.223  20.126  1.00 50.00           C  \n",
      "ATOM    291  C   SER A  97      35.379  49.488  19.548  1.00 50.00           C  \n",
      "ATOM    292  N   GLY A  98      34.110  49.742  19.846  1.00 50.00           N  \n",
      "ATOM    293  CA  GLY A  98      33.506  51.004  19.474  1.00 50.00           C  \n",
      "ATOM    294  C   GLY A  98      33.751  52.112  20.493  1.00 50.00           C  \n",
      "ATOM    295  N   SER A  99      34.676  51.882  21.423  1.00 50.00           N  \n",
      "ATOM    296  CA  SER A  99      35.037  52.893  22.416  1.00 50.00           C  \n",
      "ATOM    297  C   SER A  99      35.358  54.259  21.794  1.00 50.00           C  \n",
      "ATOM    298  N   TYR A 100      36.212  54.270  20.770  1.00 50.00           N  \n",
      "ATOM    299  CA  TYR A 100      36.567  55.501  20.093  1.00 50.00           C  \n",
      "ATOM    300  C   TYR A 100      35.333  56.246  19.603  1.00 50.00           C  \n",
      "ATOM    301  N   ASN A 101      34.404  55.527  18.992  1.00 50.00           N  \n",
      "ATOM    302  CA  ASN A 101      33.246  56.214  18.447  1.00 50.00           C  \n",
      "ATOM    303  C   ASN A 101      32.262  56.707  19.514  1.00 50.00           C  \n",
      "ATOM    304  N   SER A 102      32.262  56.068  20.680  1.00 50.00           N  \n",
      "ATOM    305  CA  SER A 102      31.509  56.630  21.811  1.00 50.00           C  \n",
      "ATOM    306  C   SER A 102      32.109  57.960  22.217  1.00 50.00           C  \n",
      "ATOM    307  N   ALA A 103      33.442  58.031  22.230  1.00 50.00           N  \n",
      "ATOM    308  CA  ALA A 103      34.097  59.266  22.637  1.00 50.00           C  \n",
      "ATOM    309  C   ALA A 103      33.865  60.370  21.607  1.00 50.00           C  \n",
      "ATOM    310  N   ILE A 104      33.837  60.005  20.329  1.00 50.00           N  \n",
      "ATOM    311  CA  ILE A 104      33.557  60.984  19.284  1.00 50.00           C  \n",
      "ATOM    312  C   ILE A 104      32.144  61.545  19.475  1.00 50.00           C  \n",
      "ATOM    313  N   ASN A 105      31.183  60.657  19.675  1.00 50.00           N  \n",
      "ATOM    314  CA  ASN A 105      29.817  61.121  19.943  1.00 50.00           C  \n",
      "ATOM    315  C   ASN A 105      29.740  61.981  21.201  1.00 50.00           C  \n",
      "ATOM    316  N   ALA A 106      30.552  61.654  22.201  1.00 50.00           N  \n",
      "ATOM    317  CA  ALA A 106      30.532  62.406  23.456  1.00 50.00           C  \n",
      "ATOM    318  C   ALA A 106      31.129  63.800  23.279  1.00 50.00           C  \n",
      "ATOM    319  N   LYS A 107      32.210  63.893  22.503  1.00 50.00           N  \n",
      "ATOM    320  CA  LYS A 107      32.771  65.196  22.160  1.00 50.00           C  \n",
      "ATOM    321  C   LYS A 107      31.703  66.088  21.512  1.00 50.00           C  \n",
      "ATOM    322  N   SER A 108      30.955  65.523  20.569  1.00 50.00           N  \n",
      "ATOM    323  CA  SER A 108      29.903  66.281  19.890  1.00 50.00           C  \n",
      "ATOM    324  C   SER A 108      28.823  66.745  20.861  1.00 50.00           C  \n",
      "ATOM    325  N   LEU A 109      28.393  65.848  21.742  1.00 50.00           N  \n",
      "ATOM    326  CA  LEU A 109      27.431  66.194  22.787  1.00 50.00           C  \n",
      "ATOM    327  C   LEU A 109      27.882  67.414  23.598  1.00 50.00           C  \n",
      "ATOM    328  N   VAL A 110      29.140  67.409  24.038  1.00 50.00           N  \n",
      "ATOM    329  CA  VAL A 110      29.631  68.455  24.932  1.00 50.00           C  \n",
      "ATOM    330  C   VAL A 110      29.739  69.782  24.199  1.00 50.00           C  \n",
      "ATOM    331  N   LEU A 111      30.105  69.718  22.925  1.00 50.00           N  \n",
      "ATOM    332  CA  LEU A 111      30.232  70.915  22.107  1.00 50.00           C  \n",
      "ATOM    333  C   LEU A 111      28.945  71.742  22.104  1.00 50.00           C  \n",
      "ATOM    334  N   GLU A 112      27.803  71.062  22.040  1.00 50.00           N  \n",
      "ATOM    335  CA  GLU A 112      26.510  71.737  22.076  1.00 50.00           C  \n",
      "ATOM    336  C   GLU A 112      26.432  72.715  23.240  1.00 50.00           C  \n",
      "ATOM    337  N   ASP A 113      26.785  72.241  24.430  1.00 50.00           N  \n",
      "ATOM    338  CA  ASP A 113      26.707  73.056  25.636  1.00 50.00           C  \n",
      "ATOM    339  C   ASP A 113      27.939  73.945  25.850  1.00 50.00           C  \n",
      "ATOM    340  N   TYR A 114      29.072  73.570  25.258  1.00 50.00           N  \n",
      "ATOM    341  CA  TYR A 114      30.298  74.362  25.392  1.00 50.00           C  \n",
      "ATOM    342  C   TYR A 114      31.032  74.506  24.060  1.00 50.00           C  \n",
      "ATOM    343  N   PRO A 115      30.513  75.364  23.171  1.00 50.00           N  \n",
      "ATOM    344  CA  PRO A 115      30.964  75.470  21.776  1.00 50.00           C  \n",
      "ATOM    345  C   PRO A 115      32.428  75.897  21.620  1.00 50.00           C  \n",
      "ATOM    346  N   ASP A 116      33.012  76.486  22.656  1.00 50.00           N  \n",
      "ATOM    347  CA  ASP A 116      34.415  76.889  22.578  1.00 50.00           C  \n",
      "ATOM    348  C   ASP A 116      35.323  75.945  23.368  1.00 50.00           C  \n",
      "ATOM    349  N   ALA A 117      34.975  74.662  23.366  1.00 50.00           N  \n",
      "ATOM    350  CA  ALA A 117      35.724  73.663  24.118  1.00 50.00           C  \n",
      "ATOM    351  C   ALA A 117      36.772  72.999  23.239  1.00 50.00           C  \n",
      "ATOM    352  N   ASN A 118      37.917  72.663  23.826  1.00 50.00           N  \n",
      "ATOM    353  CA  ASN A 118      38.953  71.945  23.086  1.00 50.00           C  \n",
      "ATOM    354  C   ASN A 118      39.069  70.479  23.509  1.00 50.00           C  \n",
      "ATOM    355  N   ILE A 119      38.569  69.587  22.664  1.00 50.00           N  \n",
      "ATOM    356  CA  ILE A 119      38.574  68.160  22.952  1.00 50.00           C  \n",
      "ATOM    357  C   ILE A 119      39.154  67.430  21.752  1.00 50.00           C  \n",
      "ATOM    358  N   CYS A 120      40.243  66.703  21.976  1.00 50.00           N  \n",
      "ATOM    359  CA  CYS A 120      40.880  65.910  20.931  1.00 50.00           C  \n",
      "ATOM    360  C   CYS A 120      40.735  64.414  21.257  1.00 50.00           C  \n",
      "ATOM    361  N   VAL A 121      40.094  63.656  20.372  1.00 50.00           N  \n",
      "ATOM    362  CA  VAL A 121      39.880  62.225  20.618  1.00 50.00           C  \n",
      "ATOM    363  C   VAL A 121      40.808  61.450  19.695  1.00 50.00           C  \n",
      "ATOM    364  N   ILE A 122      41.595  60.551  20.276  1.00 50.00           N  \n",
      "ATOM    365  CA  ILE A 122      42.571  59.790  19.517  1.00 50.00           C  \n",
      "ATOM    366  C   ILE A 122      42.314  58.287  19.565  1.00 50.00           C  \n",
      "ATOM    367  N   ASP A 123      42.130  57.692  18.390  1.00 50.00           N  \n",
      "ATOM    368  CA  ASP A 123      42.201  56.235  18.221  1.00 50.00           C  \n",
      "ATOM    369  C   ASP A 123      43.663  55.818  18.389  1.00 50.00           C  \n",
      "ATOM    370  N   SER A 124      44.016  55.298  19.555  1.00 50.00           N  \n",
      "ATOM    371  CA  SER A 124      45.420  54.983  19.826  1.00 50.00           C  \n",
      "ATOM    372  C   SER A 124      46.017  53.950  18.864  1.00 50.00           C  \n",
      "ATOM    373  N   LYS A 125      45.152  53.120  18.278  1.00 50.00           N  \n",
      "ATOM    374  CA  LYS A 125      45.533  51.948  17.487  1.00 50.00           C  \n",
      "ATOM    375  C   LYS A 125      46.270  50.907  18.325  1.00 50.00           C  \n",
      "ATOM    376  N   GLN A 126      46.195  51.043  19.644  1.00 50.00           N  \n",
      "ATOM    377  CA  GLN A 126      46.864  50.114  20.551  1.00 50.00           C  \n",
      "ATOM    378  C   GLN A 126      45.874  49.470  21.531  1.00 50.00           C  \n",
      "ATOM    379  N   ASN A 127      46.431  48.772  22.512  1.00 50.00           N  \n",
      "ATOM    380  CA  ASN A 127      45.649  48.191  23.605  1.00 50.00           C  \n",
      "ATOM    381  C   ASN A 127      46.627  47.937  24.731  1.00 50.00           C  \n",
      "ATOM    382  N   THR A 128      46.102  47.832  25.947  1.00 50.00           N  \n",
      "ATOM    383  CA  THR A 128      46.904  47.476  27.117  1.00 50.00           C  \n",
      "ATOM    384  C   THR A 128      48.323  48.095  27.061  1.00 50.00           C  \n",
      "ATOM    385  N   VAL A 129      49.372  47.274  27.145  1.00 50.00           N  \n",
      "ATOM    386  CA  VAL A 129      50.741  47.810  27.277  1.00 50.00           C  \n",
      "ATOM    387  C   VAL A 129      51.109  48.908  26.240  1.00 50.00           C  \n",
      "ATOM    388  N   THR A 130      50.723  48.742  24.971  1.00 50.00           N  \n",
      "ATOM    389  CA  THR A 130      51.116  49.734  23.971  1.00 50.00           C  \n",
      "ATOM    390  C   THR A 130      50.230  50.965  24.039  1.00 50.00           C  \n",
      "ATOM    391  N   GLN A 131      48.997  50.818  24.523  1.00 50.00           N  \n",
      "ATOM    392  CA  GLN A 131      48.196  51.984  24.764  1.00 50.00           C  \n",
      "ATOM    393  C   GLN A 131      48.813  52.755  25.923  1.00 50.00           C  \n",
      "ATOM    394  N   ALA A 132      49.252  52.048  26.967  1.00 50.00           N  \n",
      "ATOM    395  CA  ALA A 132      49.961  52.705  28.076  1.00 50.00           C  \n",
      "ATOM    396  C   ALA A 132      51.272  53.389  27.610  1.00 50.00           C  \n",
      "ATOM    397  N   LEU A 133      52.036  52.695  26.772  1.00 50.00           N  \n",
      "ATOM    398  CA  LEU A 133      53.282  53.278  26.220  1.00 50.00           C  \n",
      "ATOM    399  C   LEU A 133      53.022  54.542  25.403  1.00 50.00           C  \n",
      "ATOM    400  N   LEU A 134      51.957  54.533  24.596  1.00 50.00           N  \n",
      "ATOM    401  CA  LEU A 134      51.641  55.707  23.793  1.00 50.00           C  \n",
      "ATOM    402  C   LEU A 134      51.221  56.879  24.674  1.00 50.00           C  \n",
      "ATOM    403  N   ILE A 135      50.330  56.627  25.629  1.00 50.00           N  \n",
      "ATOM    404  CA  ILE A 135      49.929  57.680  26.531  1.00 50.00           C  \n",
      "ATOM    405  C   ILE A 135      51.148  58.256  27.242  1.00 50.00           C  \n",
      "ATOM    406  N   ASP A 136      52.013  57.371  27.703  1.00 50.00           N  \n",
      "ATOM    407  CA  ASP A 136      53.230  57.802  28.384  1.00 50.00           C  \n",
      "ATOM    408  C   ASP A 136      54.165  58.635  27.491  1.00 50.00           C  \n",
      "ATOM    409  N   GLN A 137      54.273  58.297  26.211  1.00 50.00           N  \n",
      "ATOM    410  CA  GLN A 137      55.056  59.151  25.316  1.00 50.00           C  \n",
      "ATOM    411  C   GLN A 137      54.423  60.540  25.261  1.00 50.00           C  \n",
      "ATOM    412  N   PHE A 138      53.092  60.597  25.218  1.00 50.00           N  \n",
      "ATOM    413  CA  PHE A 138      52.393  61.875  25.182  1.00 50.00           C  \n",
      "ATOM    414  C   PHE A 138      52.612  62.653  26.478  1.00 50.00           C  \n",
      "ATOM    415  N   VAL A 139      52.483  61.967  27.612  1.00 50.00           N  \n",
      "ATOM    416  CA  VAL A 139      52.754  62.577  28.914  1.00 50.00           C  \n",
      "ATOM    417  C   VAL A 139      54.180  63.132  28.964  1.00 50.00           C  \n",
      "ATOM    418  N   ARG A 140      55.130  62.354  28.461  1.00 50.00           N  \n",
      "ATOM    419  CA  ARG A 140      56.530  62.797  28.393  1.00 50.00           C  \n",
      "ATOM    420  C   ARG A 140      56.697  64.072  27.568  1.00 50.00           C  \n",
      "ATOM    421  N   MET A 141      56.119  64.085  26.369  1.00 50.00           N  \n",
      "ATOM    422  CA  MET A 141      56.264  65.237  25.475  1.00 50.00           C  \n",
      "ATOM    423  C   MET A 141      55.766  66.527  26.098  1.00 50.00           C  \n",
      "ATOM    424  N   LEU A 142      54.581  66.465  26.702  1.00 50.00           N  \n",
      "ATOM    425  CA  LEU A 142      54.003  67.622  27.363  1.00 50.00           C  \n",
      "ATOM    426  C   LEU A 142      54.812  68.053  28.578  1.00 50.00           C  \n",
      "ATOM    427  N   GLU A 143      55.289  67.100  29.370  1.00 50.00           N  \n",
      "ATOM    428  CA  GLU A 143      56.120  67.452  30.519  1.00 50.00           C  \n",
      "ATOM    429  C   GLU A 143      57.477  68.016  30.092  1.00 50.00           C  \n",
      "ATOM    430  N   ASP A 144      57.981  67.561  28.950  1.00 50.00           N  \n",
      "ATOM    431  CA  ASP A 144      59.248  68.080  28.431  1.00 50.00           C  \n",
      "ATOM    432  C   ASP A 144      59.059  69.425  27.719  1.00 50.00           C  \n",
      "ATOM    433  N   GLY A 145      57.826  69.923  27.702  1.00 50.00           N  \n",
      "ATOM    434  CA  GLY A 145      57.575  71.266  27.211  1.00 50.00           C  \n",
      "ATOM    435  C   GLY A 145      57.227  71.409  25.740  1.00 50.00           C  \n",
      "ATOM    436  N   LEU A 146      57.011  70.297  25.046  1.00 50.00           N  \n",
      "ATOM    437  CA  LEU A 146      56.488  70.368  23.684  1.00 50.00           C  \n",
      "ATOM    438  C   LEU A 146      55.068  70.906  23.720  1.00 50.00           C  \n",
      "ATOM    439  N   SER A 147      54.686  71.654  22.688  1.00 50.00           N  \n",
      "ATOM    440  CA  SER A 147      53.305  72.108  22.536  1.00 50.00           C  \n",
      "ATOM    441  C   SER A 147      52.425  70.955  22.072  1.00 50.00           C  \n",
      "ATOM    442  N   PHE A 148      51.107  71.100  22.222  1.00 50.00           N  \n",
      "ATOM    443  CA  PHE A 148      50.199  70.079  21.724  1.00 50.00           C  \n",
      "ATOM    444  C   PHE A 148      50.462  69.785  20.247  1.00 50.00           C  \n",
      "ATOM    445  N   GLU A 149      50.670  70.831  19.456  1.00 50.00           N  \n",
      "ATOM    446  CA  GLU A 149      50.902  70.653  18.025  1.00 50.00           C  \n",
      "ATOM    447  C   GLU A 149      52.178  69.863  17.739  1.00 50.00           C  \n",
      "ATOM    448  N   GLN A 150      53.243  70.172  18.470  1.00 50.00           N  \n",
      "ATOM    449  CA  GLN A 150      54.506  69.461  18.292  1.00 50.00           C  \n",
      "ATOM    450  C   GLN A 150      54.349  67.980  18.624  1.00 50.00           C  \n",
      "ATOM    451  N   ALA A 151      53.552  67.684  19.648  1.00 50.00           N  \n",
      "ATOM    452  CA  ALA A 151      53.361  66.304  20.070  1.00 50.00           C  \n",
      "ATOM    453  C   ALA A 151      52.538  65.554  19.029  1.00 50.00           C  \n",
      "ATOM    454  N   MET A 152      51.504  66.209  18.512  1.00 50.00           N  \n",
      "ATOM    455  CA  MET A 152      50.630  65.568  17.540  1.00 50.00           C  \n",
      "ATOM    456  C   MET A 152      51.383  65.258  16.251  1.00 50.00           C  \n",
      "ATOM    457  N   SER A 153      52.375  66.075  15.914  1.00 50.00           N  \n",
      "ATOM    458  CA  SER A 153      53.173  65.829  14.713  1.00 50.00           C  \n",
      "ATOM    459  C   SER A 153      54.053  64.581  14.834  1.00 50.00           C  \n",
      "ATOM    460  N   LYS A 154      54.157  64.027  16.036  1.00 50.00           N  \n",
      "ATOM    461  CA  LYS A 154      54.932  62.795  16.229  1.00 50.00           C  \n",
      "ATOM    462  C   LYS A 154      54.041  61.582  16.442  1.00 50.00           C  \n",
      "ATOM    463  N   LEU A 155      52.750  61.830  16.634  1.00 50.00           N  \n",
      "ATOM    464  CA  LEU A 155      51.827  60.757  16.985  1.00 50.00           C  \n",
      "ATOM    465  C   LEU A 155      51.774  59.599  15.976  1.00 50.00           C  \n",
      "ATOM    466  N   ASP A 156      51.676  59.900  14.684  1.00 50.00           N  \n",
      "ATOM    467  CA  ASP A 156      51.577  58.838  13.681  1.00 50.00           C  \n",
      "ATOM    468  C   ASP A 156      52.760  57.878  13.734  1.00 50.00           C  \n",
      "ATOM    469  N   ALA A 157      53.965  58.434  13.750  1.00 50.00           N  \n",
      "ATOM    470  CA  ALA A 157      55.169  57.625  13.796  1.00 50.00           C  \n",
      "ATOM    471  C   ALA A 157      55.235  56.794  15.083  1.00 50.00           C  \n",
      "ATOM    472  N   LEU A 158      54.835  57.393  16.202  1.00 50.00           N  \n",
      "ATOM    473  CA  LEU A 158      54.809  56.657  17.467  1.00 50.00           C  \n",
      "ATOM    474  C   LEU A 158      53.875  55.453  17.368  1.00 50.00           C  \n",
      "ATOM    475  N   MET A 159      52.660  55.687  16.884  1.00 50.00           N  \n",
      "ATOM    476  CA  MET A 159      51.692  54.603  16.774  1.00 50.00           C  \n",
      "ATOM    477  C   MET A 159      52.151  53.493  15.828  1.00 50.00           C  \n",
      "ATOM    478  N   ALA A 160      52.737  53.884  14.699  1.00 50.00           N  \n",
      "ATOM    479  CA  ALA A 160      53.246  52.907  13.730  1.00 50.00           C  \n",
      "ATOM    480  C   ALA A 160      54.376  52.031  14.303  1.00 50.00           C  \n",
      "ATOM    481  N   SER A 161      55.125  52.574  15.261  1.00 50.00           N  \n",
      "ATOM    482  CA  SER A 161      56.292  51.883  15.805  1.00 50.00           C  \n",
      "ATOM    483  C   SER A 161      55.906  50.922  16.915  1.00 50.00           C  \n",
      "ATOM    484  N   ALA A 162      54.718  51.108  17.503  1.00 50.00           N  \n",
      "ATOM    485  CA  ALA A 162      54.322  50.293  18.637  1.00 50.00           C  \n",
      "ATOM    486  C   ALA A 162      54.115  48.831  18.263  1.00 50.00           C  \n",
      "ATOM    487  N   ARG A 163      54.526  47.922  19.146  1.00 50.00           N  \n",
      "ATOM    488  CA  ARG A 163      54.255  46.491  18.953  1.00 50.00           C  \n",
      "ATOM    489  C   ARG A 163      53.928  45.832  20.282  1.00 50.00           C  \n",
      "ATOM    490  N   ILE A 164      53.020  44.859  20.268  1.00 50.00           N  \n",
      "ATOM    491  CA  ILE A 164      52.935  43.930  21.383  1.00 50.00           C  \n",
      "ATOM    492  C   ILE A 164      53.332  42.564  20.876  1.00 50.00           C  \n",
      "ATOM    493  N   PHE A 165      54.161  41.871  21.645  1.00 50.00           N  \n",
      "ATOM    494  CA  PHE A 165      54.529  40.497  21.351  1.00 50.00           C  \n",
      "ATOM    495  C   PHE A 165      54.087  39.658  22.535  1.00 50.00           C  \n",
      "ATOM    496  N   PHE A 166      53.229  38.676  22.306  1.00 50.00           N  \n",
      "ATOM    497  CA  PHE A 166      52.646  37.978  23.450  1.00 50.00           C  \n",
      "ATOM    498  C   PHE A 166      52.129  36.603  23.131  1.00 50.00           C  \n",
      "ATOM    499  N   THR A 167      51.776  35.876  24.188  1.00 50.00           N  \n",
      "ATOM    500  CA  THR A 167      51.124  34.592  24.025  1.00 50.00           C  \n",
      "ATOM    501  C   THR A 167      49.954  34.486  25.006  1.00 50.00           C  \n",
      "ATOM    502  N   VAL A 168      49.106  33.481  24.813  1.00 50.00           N  \n",
      "ATOM    503  CA  VAL A 168      47.934  33.279  25.667  1.00 50.00           C  \n",
      "ATOM    504  C   VAL A 168      47.784  31.789  25.977  1.00 50.00           C  \n",
      "ATOM    505  N   GLY A 169      46.843  31.426  26.847  1.00 50.00           N  \n",
      "ATOM    506  CA  GLY A 169      46.717  30.037  27.273  1.00 50.00           C  \n",
      "ATOM    507  C   GLY A 169      45.934  29.131  26.334  1.00 50.00           C  \n",
      "ATOM    508  N   SER A 170      45.153  29.730  25.443  1.00 50.00           N  \n",
      "ATOM    509  CA  SER A 170      44.289  28.982  24.533  1.00 50.00           C  \n",
      "ATOM    510  C   SER A 170      43.703  29.968  23.544  1.00 50.00           C  \n",
      "ATOM    511  N   LEU A 171      43.008  29.465  22.530  1.00 50.00           N  \n",
      "ATOM    512  CA  LEU A 171      42.331  30.340  21.578  1.00 50.00           C  \n",
      "ATOM    513  C   LEU A 171      40.861  30.552  21.948  1.00 50.00           C  \n",
      "ATOM    514  N   ASP A 172      40.397  29.849  22.980  1.00 50.00           N  \n",
      "ATOM    515  CA  ASP A 172      38.971  29.820  23.330  1.00 50.00           C  \n",
      "ATOM    516  C   ASP A 172      38.356  31.207  23.526  1.00 50.00           C  \n",
      "ATOM    517  N   TYR A 173      39.016  32.044  24.318  1.00 50.00           N  \n",
      "ATOM    518  CA  TYR A 173      38.473  33.371  24.613  1.00 50.00           C  \n",
      "ATOM    519  C   TYR A 173      38.530  34.324  23.425  1.00 50.00           C  \n",
      "ATOM    520  N   LEU A 174      39.644  34.309  22.688  1.00 50.00           N  \n",
      "ATOM    521  CA  LEU A 174      39.768  35.120  21.484  1.00 50.00           C  \n",
      "ATOM    522  C   LEU A 174      38.660  34.807  20.489  1.00 50.00           C  \n",
      "ATOM    523  N   LYS A 175      38.398  33.522  20.252  1.00 50.00           N  \n",
      "ATOM    524  CA  LYS A 175      37.373  33.153  19.283  1.00 50.00           C  \n",
      "ATOM    525  C   LYS A 175      35.987  33.540  19.798  1.00 50.00           C  \n",
      "ATOM    526  N   MET A 176      35.708  33.201  21.048  1.00 50.00           N  \n",
      "ATOM    527  CA  MET A 176      34.383  33.457  21.611  1.00 50.00           C  \n",
      "ATOM    528  C   MET A 176      34.069  34.957  21.639  1.00 50.00           C  \n",
      "ATOM    529  N   GLY A 177      35.081  35.781  21.899  1.00 50.00           N  \n",
      "ATOM    530  CA  GLY A 177      34.871  37.220  21.997  1.00 50.00           C  \n",
      "ATOM    531  C   GLY A 177      34.838  37.956  20.659  1.00 50.00           C  \n",
      "ATOM    532  N   GLY A 178      35.191  37.266  19.577  1.00 50.00           N  \n",
      "ATOM    533  CA  GLY A 178      35.052  37.819  18.235  1.00 50.00           C  \n",
      "ATOM    534  C   GLY A 178      36.227  38.610  17.658  1.00 50.00           C  \n",
      "ATOM    535  N   ARG A 179      37.311  38.757  18.412  1.00 50.00           N  \n",
      "ATOM    536  CA  ARG A 179      38.430  39.584  17.929  1.00 50.00           C  \n",
      "ATOM    537  C   ARG A 179      39.634  38.771  17.455  1.00 50.00           C  \n",
      "ATOM    538  N   ILE A 180      39.460  37.464  17.318  1.00 50.00           N  \n",
      "ATOM    539  CA  ILE A 180      40.588  36.601  17.004  1.00 50.00           C  \n",
      "ATOM    540  C   ILE A 180      41.206  36.945  15.642  1.00 50.00           C  \n",
      "ATOM    541  N   GLY A 181      40.360  37.306  14.685  1.00 50.00           N  \n",
      "ATOM    542  CA  GLY A 181      40.812  37.663  13.355  1.00 50.00           C  \n",
      "ATOM    543  C   GLY A 181      41.852  36.732  12.747  1.00 50.00           C  \n",
      "ATOM    544  N   LYS A 182      42.961  37.320  12.320  1.00 50.00           N  \n",
      "ATOM    545  CA  LYS A 182      43.994  36.571  11.594  1.00 50.00           C  \n",
      "ATOM    546  C   LYS A 182      44.771  35.549  12.434  1.00 50.00           C  \n",
      "ATOM    547  N   VAL A 183      44.558  35.566  13.743  1.00 50.00           N  \n",
      "ATOM    548  CA  VAL A 183      45.195  34.572  14.602  1.00 50.00           C  \n",
      "ATOM    549  C   VAL A 183      44.521  33.194  14.447  1.00 50.00           C  \n",
      "ATOM    550  N   ALA A 184      43.299  33.164  13.915  1.00 50.00           N  \n",
      "ATOM    551  CA  ALA A 184      42.643  31.879  13.628  1.00 50.00           C  \n",
      "ATOM    552  C   ALA A 184      43.110  31.336  12.282  1.00 50.00           C  \n",
      "ATOM    553  N   THR A 185      43.883  30.257  12.311  1.00 50.00           N  \n",
      "ATOM    554  CA  THR A 185      44.319  29.608  11.077  1.00 50.00           C  \n",
      "ATOM    555  C   THR A 185      44.232  28.095  11.245  1.00 50.00           C  \n",
      "ATOM    556  N   ALA A 186      44.506  27.357  10.177  1.00 50.00           N  \n",
      "ATOM    557  CA  ALA A 186      44.459  25.906  10.256  1.00 50.00           C  \n",
      "ATOM    558  C   ALA A 186      45.431  25.434  11.331  1.00 50.00           C  \n",
      "ATOM    559  N   ALA A 187      46.646  25.977  11.304  1.00 50.00           N  \n",
      "ATOM    560  CA  ALA A 187      47.680  25.601  12.262  1.00 50.00           C  \n",
      "ATOM    561  C   ALA A 187      47.301  25.865  13.729  1.00 50.00           C  \n",
      "ATOM    562  N   THR A 188      46.707  27.017  14.016  1.00 50.00           N  \n",
      "ATOM    563  CA  THR A 188      46.393  27.354  15.406  1.00 50.00           C  \n",
      "ATOM    564  C   THR A 188      45.149  26.627  15.907  1.00 50.00           C  \n",
      "ATOM    565  N   GLY A 189      44.385  26.047  14.988  1.00 50.00           N  \n",
      "ATOM    566  CA  GLY A 189      43.151  25.376  15.356  1.00 50.00           C  \n",
      "ATOM    567  C   GLY A 189      43.390  23.957  15.844  1.00 50.00           C  \n",
      "ATOM    568  N   LYS A 190      44.627  23.494  15.716  1.00 50.00           N  \n",
      "ATOM    569  CA  LYS A 190      44.973  22.132  16.107  1.00 50.00           C  \n",
      "ATOM    570  C   LYS A 190      45.026  21.962  17.630  1.00 50.00           C  \n",
      "ATOM    571  N   LEU A 191      44.692  20.766  18.095  1.00 50.00           N  \n",
      "ATOM    572  CA  LEU A 191      44.645  20.488  19.525  1.00 50.00           C  \n",
      "ATOM    573  C   LEU A 191      46.004  20.710  20.196  1.00 50.00           C  \n",
      "ATOM    574  N   GLY A 192      45.999  21.489  21.272  1.00 50.00           N  \n",
      "ATOM    575  CA  GLY A 192      47.197  21.718  22.061  1.00 50.00           C  \n",
      "ATOM    576  C   GLY A 192      48.010  22.949  21.686  1.00 50.00           C  \n",
      "ATOM    577  N   VAL A 193      47.720  23.542  20.533  1.00 50.00           N  \n",
      "ATOM    578  CA  VAL A 193      48.536  24.651  20.043  1.00 50.00           C  \n",
      "ATOM    579  C   VAL A 193      48.113  26.002  20.614  1.00 50.00           C  \n",
      "ATOM    580  N   LYS A 194      49.104  26.848  20.888  1.00 50.00           N  \n",
      "ATOM    581  CA  LYS A 194      48.869  28.241  21.254  1.00 50.00           C  \n",
      "ATOM    582  C   LYS A 194      49.663  29.130  20.312  1.00 50.00           C  \n",
      "ATOM    583  N   PRO A 195      49.140  30.330  20.056  1.00 50.00           N  \n",
      "ATOM    584  CA  PRO A 195      49.783  31.294  19.157  1.00 50.00           C  \n",
      "ATOM    585  C   PRO A 195      50.791  32.189  19.854  1.00 50.00           C  \n",
      "ATOM    586  N   VAL A 196      51.812  32.592  19.102  1.00 50.00           N  \n",
      "ATOM    587  CA  VAL A 196      52.578  33.766  19.458  1.00 50.00           C  \n",
      "ATOM    588  C   VAL A 196      51.916  34.845  18.620  1.00 50.00           C  \n",
      "ATOM    589  N   ILE A 197      51.525  35.937  19.260  1.00 50.00           N  \n",
      "ATOM    590  CA  ILE A 197      50.709  36.943  18.598  1.00 50.00           C  \n",
      "ATOM    591  C   ILE A 197      51.444  38.273  18.545  1.00 50.00           C  \n",
      "ATOM    592  N   ILE A 198      51.246  39.012  17.459  1.00 50.00           N  \n",
      "ATOM    593  CA  ILE A 198      51.823  40.331  17.330  1.00 50.00           C  \n",
      "ATOM    594  C   ILE A 198      50.711  41.331  17.126  1.00 50.00           C  \n",
      "ATOM    595  N   MET A 199      50.707  42.396  17.920  1.00 50.00           N  \n",
      "ATOM    596  CA  MET A 199      49.802  43.506  17.656  1.00 50.00           C  \n",
      "ATOM    597  C   MET A 199      50.550  44.620  16.956  1.00 50.00           C  \n",
      "ATOM    598  N   LYS A 200      50.062  45.006  15.785  1.00 50.00           N  \n",
      "ATOM    599  CA  LYS A 200      50.674  46.112  15.080  1.00 50.00           C  \n",
      "ATOM    600  C   LYS A 200      49.659  46.712  14.150  1.00 50.00           C  \n",
      "ATOM    601  N   ASP A 201      49.798  48.011  13.924  1.00 50.00           N  \n",
      "ATOM    602  CA  ASP A 201      48.917  48.694  12.993  1.00 50.00           C  \n",
      "ATOM    603  C   ASP A 201      47.447  48.571  13.423  1.00 50.00           C  \n",
      "ATOM    604  N   GLY A 202      47.202  48.472  14.727  1.00 50.00           N  \n",
      "ATOM    605  CA  GLY A 202      45.830  48.478  15.231  1.00 50.00           C  \n",
      "ATOM    606  C   GLY A 202      45.112  47.148  15.355  1.00 50.00           C  \n",
      "ATOM    607  N   ASP A 203      45.792  46.054  15.013  1.00 50.00           N  \n",
      "ATOM    608  CA  ASP A 203      45.151  44.746  15.015  1.00 50.00           C  \n",
      "ATOM    609  C   ASP A 203      46.143  43.634  15.397  1.00 50.00           C  \n",
      "ATOM    610  N   ILE A 204      45.654  42.400  15.539  1.00 50.00           N  \n",
      "ATOM    611  CA  ILE A 204      46.533  41.256  15.839  1.00 50.00           C  \n",
      "ATOM    612  C   ILE A 204      46.766  40.317  14.658  1.00 50.00           C  \n",
      "ATOM    613  N   GLY A 205      47.947  39.707  14.639  1.00 50.00           N  \n",
      "ATOM    614  CA  GLY A 205      48.319  38.750  13.615  1.00 50.00           C  \n",
      "ATOM    615  C   GLY A 205      49.183  37.662  14.225  1.00 50.00           C  \n",
      "ATOM    616  N   LEU A 206      49.392  36.581  13.480  1.00 50.00           N  \n",
      "ATOM    617  CA  LEU A 206      50.107  35.430  14.014  1.00 50.00           C  \n",
      "ATOM    618  C   LEU A 206      51.611  35.596  13.818  1.00 50.00           C  \n",
      "ATOM    619  N   GLY A 207      52.381  35.457  14.895  1.00 50.00           N  \n",
      "ATOM    620  CA  GLY A 207      53.830  35.551  14.799  1.00 50.00           C  \n",
      "ATOM    621  C   GLY A 207      54.575  34.242  15.017  1.00 50.00           C  \n",
      "ATOM    622  N   GLY A 208      53.842  33.164  15.280  1.00 50.00           N  \n",
      "ATOM    623  CA  GLY A 208      54.447  31.867  15.557  1.00 50.00           C  \n",
      "ATOM    624  C   GLY A 208      53.459  30.965  16.290  1.00 50.00           C  \n",
      "ATOM    625  N   ILE A 209      53.849  29.719  16.551  1.00 50.00           N  \n",
      "ATOM    626  CA  ILE A 209      53.015  28.798  17.336  1.00 50.00           C  \n",
      "ATOM    627  C   ILE A 209      53.865  27.919  18.255  1.00 50.00           C  \n",
      "ATOM    628  N   GLY A 210      53.203  27.191  19.146  1.00 50.00           N  \n",
      "ATOM    629  CA  GLY A 210      53.893  26.268  20.035  1.00 50.00           C  \n",
      "ATOM    630  C   GLY A 210      52.920  25.609  20.994  1.00 50.00           C  \n",
      "ATOM    631  N   ARG A 211      53.447  24.842  21.944  1.00 50.00           N  \n",
      "ATOM    632  CA  ARG A 211      52.609  24.046  22.847  1.00 50.00           C  \n",
      "ATOM    633  C   ARG A 211      52.991  24.234  24.309  1.00 50.00           C  \n",
      "ATOM    634  N   ASN A 212      54.286  24.374  24.559  1.00 50.00           N  \n",
      "ATOM    635  CA  ASN A 212      54.794  24.501  25.915  1.00 50.00           C  \n",
      "ATOM    636  C   ASN A 212      54.880  25.971  26.327  1.00 50.00           C  \n",
      "ATOM    637  N   ARG A 213      54.348  26.295  27.502  1.00 50.00           N  \n",
      "ATOM    638  CA  ARG A 213      54.353  27.668  27.999  1.00 50.00           C  \n",
      "ATOM    639  C   ARG A 213      55.740  28.316  27.941  1.00 50.00           C  \n",
      "ATOM    640  N   ASN A 214      56.746  27.628  28.472  1.00 50.00           N  \n",
      "ATOM    641  CA  ASN A 214      58.091  28.193  28.503  1.00 50.00           C  \n",
      "ATOM    642  C   ASN A 214      58.680  28.404  27.114  1.00 50.00           C  \n",
      "ATOM    643  N   LYS A 215      58.443  27.455  26.215  1.00 50.00           N  \n",
      "ATOM    644  CA  LYS A 215      58.914  27.594  24.839  1.00 50.00           C  \n",
      "ATOM    645  C   LYS A 215      58.184  28.732  24.117  1.00 50.00           C  \n",
      "ATOM    646  N   LEU A 216      56.895  28.878  24.389  1.00 50.00           N  \n",
      "ATOM    647  CA  LEU A 216      56.137  29.966  23.782  1.00 50.00           C  \n",
      "ATOM    648  C   LEU A 216      56.670  31.316  24.238  1.00 50.00           C  \n",
      "ATOM    649  N   LYS A 217      57.001  31.441  25.523  1.00 50.00           N  \n",
      "ATOM    650  CA  LYS A 217      57.539  32.702  26.025  1.00 50.00           C  \n",
      "ATOM    651  C   LYS A 217      58.906  32.974  25.400  1.00 50.00           C  \n",
      "ATOM    652  N   ASN A 218      59.680  31.926  25.164  1.00 50.00           N  \n",
      "ATOM    653  CA  ASN A 218      60.954  32.117  24.478  1.00 50.00           C  \n",
      "ATOM    654  C   ASN A 218      60.753  32.577  23.036  1.00 50.00           C  \n",
      "ATOM    655  N   SER A 219      59.710  32.075  22.384  1.00 50.00           N  \n",
      "ATOM    656  CA  SER A 219      59.422  32.489  21.014  1.00 50.00           C  \n",
      "ATOM    657  C   SER A 219      58.952  33.933  20.963  1.00 50.00           C  \n",
      "ATOM    658  N   VAL A 220      58.114  34.330  21.927  1.00 50.00           N  \n",
      "ATOM    659  CA  VAL A 220      57.730  35.735  22.058  1.00 50.00           C  \n",
      "ATOM    660  C   VAL A 220      58.975  36.623  22.073  1.00 50.00           C  \n",
      "ATOM    661  N   LEU A 221      59.957  36.262  22.899  1.00 50.00           N  \n",
      "ATOM    662  CA  LEU A 221      61.189  37.044  23.025  1.00 50.00           C  \n",
      "ATOM    663  C   LEU A 221      61.983  37.076  21.721  1.00 50.00           C  \n",
      "ATOM    664  N   GLN A 222      62.016  35.959  21.011  1.00 50.00           N  \n",
      "ATOM    665  CA  GLN A 222      62.772  35.928  19.764  1.00 50.00           C  \n",
      "ATOM    666  C   GLN A 222      62.118  36.799  18.694  1.00 50.00           C  \n",
      "ATOM    667  N   VAL A 223      60.788  36.797  18.640  1.00 50.00           N  \n",
      "ATOM    668  CA  VAL A 223      60.096  37.664  17.700  1.00 50.00           C  \n",
      "ATOM    669  C   VAL A 223      60.316  39.151  18.041  1.00 50.00           C  \n",
      "ATOM    670  N   ALA A 224      60.296  39.483  19.326  1.00 50.00           N  \n",
      "ATOM    671  CA  ALA A 224      60.511  40.877  19.710  1.00 50.00           C  \n",
      "ATOM    672  C   ALA A 224      61.951  41.275  19.371  1.00 50.00           C  \n",
      "ATOM    673  N   LYS A 225      62.881  40.363  19.618  1.00 50.00           N  \n",
      "ATOM    674  CA  LYS A 225      64.306  40.660  19.440  1.00 50.00           C  \n",
      "ATOM    675  C   LYS A 225      64.616  41.009  17.987  1.00 50.00           C  \n",
      "ATOM    676  N   LYS A 226      64.034  40.241  17.072  1.00 50.00           N  \n",
      "ATOM    677  CA  LYS A 226      64.153  40.500  15.646  1.00 50.00           C  \n",
      "ATOM    678  C   LYS A 226      63.650  41.898  15.274  1.00 50.00           C  \n",
      "ATOM    679  N   TYR A 227      62.485  42.278  15.787  1.00 50.00           N  \n",
      "ATOM    680  CA  TYR A 227      61.950  43.593  15.506  1.00 50.00           C  \n",
      "ATOM    681  C   TYR A 227      62.827  44.709  16.090  1.00 50.00           C  \n",
      "ATOM    682  N   LEU A 228      63.258  44.542  17.336  1.00 50.00           N  \n",
      "ATOM    683  CA  LEU A 228      64.088  45.567  17.961  1.00 50.00           C  \n",
      "ATOM    684  C   LEU A 228      65.439  45.696  17.260  1.00 50.00           C  \n",
      "ATOM    685  N   ASP A 229      66.038  44.566  16.908  1.00 50.00           N  \n",
      "ATOM    686  CA  ASP A 229      67.322  44.591  16.212  1.00 50.00           C  \n",
      "ATOM    687  C   ASP A 229      67.227  45.274  14.854  1.00 50.00           C  \n",
      "ATOM    688  N   GLU A 230      66.139  45.018  14.135  1.00 50.00           N  \n",
      "ATOM    689  CA  GLU A 230      65.917  45.659  12.846  1.00 50.00           C  \n",
      "ATOM    690  C   GLU A 230      65.809  47.170  12.977  1.00 50.00           C  \n",
      "ATOM    691  N   ASN A 231      65.178  47.634  14.055  1.00 50.00           N  \n",
      "ATOM    692  CA  ASN A 231      64.758  49.027  14.153  1.00 50.00           C  \n",
      "ATOM    693  C   ASN A 231      65.556  49.914  15.118  1.00 50.00           C  \n",
      "ATOM    694  N   ASN A 232      66.556  49.320  15.753  1.00 50.00           N  \n",
      "ATOM    695  CA  ASN A 232      67.487  50.042  16.611  1.00 50.00           C  \n",
      "ATOM    696  C   ASN A 232      66.946  49.999  18.027  1.00 50.00           C  \n",
      "ATOM    697  N   LYS A 233      67.332  48.957  18.742  1.00 50.00           N  \n",
      "ATOM    698  CA  LYS A 233      66.822  48.690  20.071  1.00 50.00           C  \n",
      "ATOM    699  C   LYS A 233      66.982  49.879  21.032  1.00 50.00           C  \n",
      "ATOM    700  N   ASP A 234      68.065  50.638  20.866  1.00 50.00           N  \n",
      "ATOM    701  CA  ASP A 234      68.359  51.755  21.764  1.00 50.00           C  \n",
      "ATOM    702  C   ASP A 234      67.296  52.838  21.661  1.00 50.00           C  \n",
      "ATOM    703  N   ASN A 235      66.553  52.842  20.560  1.00 50.00           N  \n",
      "ATOM    704  CA  ASN A 235      65.585  53.900  20.327  1.00 50.00           C  \n",
      "ATOM    705  C   ASN A 235      64.175  53.574  20.809  1.00 50.00           C  \n",
      "ATOM    706  N   PHE A 236      64.025  52.460  21.521  1.00 50.00           N  \n",
      "ATOM    707  CA  PHE A 236      62.701  52.010  22.007  1.00 50.00           C  \n",
      "ATOM    708  C   PHE A 236      62.574  51.967  23.528  1.00 50.00           C  \n",
      "ATOM    709  N   ILE A 237      61.353  52.196  24.005  1.00 50.00           N  \n",
      "ATOM    710  CA  ILE A 237      60.947  51.917  25.380  1.00 50.00           C  \n",
      "ATOM    711  C   ILE A 237      60.351  50.512  25.419  1.00 50.00           C  \n",
      "ATOM    712  N   VAL A 238      60.650  49.734  26.452  1.00 50.00           N  \n",
      "ATOM    713  CA  VAL A 238      60.131  48.373  26.530  1.00 50.00           C  \n",
      "ATOM    714  C   VAL A 238      59.454  48.134  27.890  1.00 50.00           C  \n",
      "ATOM    715  N   SER A 239      58.232  47.602  27.872  1.00 50.00           N  \n",
      "ATOM    716  CA  SER A 239      57.525  47.255  29.116  1.00 50.00           C  \n",
      "ATOM    717  C   SER A 239      56.852  45.900  28.988  1.00 50.00           C  \n",
      "ATOM    718  N   VAL A 240      56.520  45.298  30.130  1.00 50.00           N  \n",
      "ATOM    719  CA  VAL A 240      55.932  43.958  30.162  1.00 50.00           C  \n",
      "ATOM    720  C   VAL A 240      54.486  43.988  30.678  1.00 50.00           C  \n",
      "ATOM    721  N   GLY A 241      53.654  43.084  30.157  1.00 50.00           N  \n",
      "ATOM    722  CA  GLY A 241      52.312  42.852  30.693  1.00 50.00           C  \n",
      "ATOM    723  C   GLY A 241      52.142  41.402  31.134  1.00 50.00           C  \n",
      "ATOM    724  N   TYR A 242      51.459  41.197  32.255  1.00 50.00           N  \n",
      "ATOM    725  CA  TYR A 242      51.249  39.853  32.782  1.00 50.00           C  \n",
      "ATOM    726  C   TYR A 242      49.796  39.648  33.193  1.00 50.00           C  \n",
      "ATOM    727  N   GLY A 243      49.388  38.387  33.255  1.00 50.00           N  \n",
      "ATOM    728  CA  GLY A 243      48.009  38.039  33.534  1.00 50.00           C  \n",
      "ATOM    729  C   GLY A 243      47.827  37.600  34.971  1.00 50.00           C  \n",
      "ATOM    730  N   TYR A 244      47.283  36.403  35.144  1.00 50.00           N  \n",
      "ATOM    731  CA  TYR A 244      46.878  35.928  36.461  1.00 50.00           C  \n",
      "ATOM    732  C   TYR A 244      48.063  35.716  37.398  1.00 50.00           C  \n",
      "ATOM    733  N   ASP A 245      49.137  35.155  36.858  1.00 50.00           N  \n",
      "ATOM    734  CA  ASP A 245      50.292  34.779  37.666  1.00 50.00           C  \n",
      "ATOM    735  C   ASP A 245      51.335  35.892  37.704  1.00 50.00           C  \n",
      "ATOM    736  N   LYS A 246      51.307  36.686  38.772  1.00 50.00           N  \n",
      "ATOM    737  CA  LYS A 246      52.239  37.798  38.940  1.00 50.00           C  \n",
      "ATOM    738  C   LYS A 246      53.705  37.357  39.009  1.00 50.00           C  \n",
      "ATOM    739  N   GLU A 247      53.960  36.219  39.656  1.00 50.00           N  \n",
      "ATOM    740  CA  GLU A 247      55.322  35.710  39.813  1.00 50.00           C  \n",
      "ATOM    741  C   GLU A 247      55.943  35.389  38.469  1.00 50.00           C  \n",
      "ATOM    742  N   GLU A 248      55.144  34.768  37.609  1.00 50.00           N  \n",
      "ATOM    743  CA  GLU A 248      55.576  34.459  36.257  1.00 50.00           C  \n",
      "ATOM    744  C   GLU A 248      55.916  35.762  35.561  1.00 50.00           C  \n",
      "ATOM    745  N   GLY A 249      55.065  36.761  35.756  1.00 50.00           N  \n",
      "ATOM    746  CA  GLY A 249      55.314  38.079  35.198  1.00 50.00           C  \n",
      "ATOM    747  C   GLY A 249      56.669  38.651  35.591  1.00 50.00           C  \n",
      "ATOM    748  N   PHE A 250      56.959  38.667  36.890  1.00 50.00           N  \n",
      "ATOM    749  CA  PHE A 250      58.228  39.227  37.361  1.00 50.00           C  \n",
      "ATOM    750  C   PHE A 250      59.412  38.445  36.795  1.00 50.00           C  \n",
      "ATOM    751  N   GLU A 251      59.267  37.127  36.736  1.00 50.00           N  \n",
      "ATOM    752  CA  GLU A 251      60.312  36.265  36.207  1.00 50.00           C  \n",
      "ATOM    753  C   GLU A 251      60.468  36.515  34.712  1.00 50.00           C  \n",
      "ATOM    754  N   PHE A 252      59.347  36.710  34.032  1.00 50.00           N  \n",
      "ATOM    755  CA  PHE A 252      59.399  37.019  32.611  1.00 50.00           C  \n",
      "ATOM    756  C   PHE A 252      60.164  38.319  32.380  1.00 50.00           C  \n",
      "ATOM    757  N   MET A 253      59.949  39.306  33.249  1.00 50.00           N  \n",
      "ATOM    758  CA  MET A 253      60.633  40.585  33.100  1.00 50.00           C  \n",
      "ATOM    759  C   MET A 253      62.138  40.393  33.133  1.00 50.00           C  \n",
      "ATOM    760  N   LYS A 254      62.595  39.409  33.902  1.00 50.00           N  \n",
      "ATOM    761  CA  LYS A 254      64.027  39.154  34.000  1.00 50.00           C  \n",
      "ATOM    762  C   LYS A 254      64.537  38.554  32.697  1.00 50.00           C  \n",
      "ATOM    763  N   GLU A 255      63.759  37.633  32.134  1.00 50.00           N  \n",
      "ATOM    764  CA  GLU A 255      64.099  37.003  30.861  1.00 50.00           C  \n",
      "ATOM    765  C   GLU A 255      64.135  38.027  29.736  1.00 50.00           C  \n",
      "ATOM    766  N   VAL A 256      63.201  38.975  29.764  1.00 50.00           N  \n",
      "ATOM    767  CA  VAL A 256      63.167  40.030  28.752  1.00 50.00           C  \n",
      "ATOM    768  C   VAL A 256      64.437  40.875  28.812  1.00 50.00           C  \n",
      "ATOM    769  N   GLU A 257      64.795  41.318  30.014  1.00 50.00           N  \n",
      "ATOM    770  CA  GLU A 257      65.940  42.201  30.168  1.00 50.00           C  \n",
      "ATOM    771  C   GLU A 257      67.212  41.491  29.729  1.00 50.00           C  \n",
      "ATOM    772  N   SER A 258      67.324  40.227  30.124  1.00 50.00           N  \n",
      "ATOM    773  CA  SER A 258      68.470  39.394  29.810  1.00 50.00           C  \n",
      "ATOM    774  C   SER A 258      68.552  39.102  28.312  1.00 50.00           C  \n",
      "ATOM    775  N   THR A 259      67.438  38.675  27.732  1.00 50.00           N  \n",
      "ATOM    776  CA  THR A 259      67.440  38.269  26.329  1.00 50.00           C  \n",
      "ATOM    777  C   THR A 259      67.630  39.449  25.378  1.00 50.00           C  \n",
      "ATOM    778  N   LEU A 260      66.974  40.558  25.685  1.00 50.00           N  \n",
      "ATOM    779  CA  LEU A 260      67.011  41.727  24.820  1.00 50.00           C  \n",
      "ATOM    780  C   LEU A 260      68.137  42.690  25.184  1.00 50.00           C  \n",
      "ATOM    781  N   ASP A 261      68.734  42.500  26.360  1.00 50.00           N  \n",
      "ATOM    782  CA  ASP A 261      69.761  43.414  26.859  1.00 50.00           C  \n",
      "ATOM    783  C   ASP A 261      69.203  44.833  26.964  1.00 50.00           C  \n",
      "ATOM    784  N   VAL A 262      68.073  44.963  27.651  1.00 50.00           N  \n",
      "ATOM    785  CA  VAL A 262      67.505  46.268  27.919  1.00 50.00           C  \n",
      "ATOM    786  C   VAL A 262      67.098  46.345  29.378  1.00 50.00           C  \n",
      "ATOM    787  N   LYS A 263      66.775  47.546  29.825  1.00 50.00           N  \n",
      "ATOM    788  CA  LYS A 263      66.149  47.718  31.122  1.00 50.00           C  \n",
      "ATOM    789  C   LYS A 263      64.716  48.170  30.876  1.00 50.00           C  \n",
      "ATOM    790  N   LEU A 264      63.758  47.469  31.468  1.00 50.00           N  \n",
      "ATOM    791  CA  LEU A 264      62.354  47.817  31.268  1.00 50.00           C  \n",
      "ATOM    792  C   LEU A 264      62.039  49.206  31.795  1.00 50.00           C  \n",
      "ATOM    793  N   ASP A 265      61.104  49.895  31.152  1.00 50.00           N  \n",
      "ATOM    794  CA  ASP A 265      60.562  51.103  31.750  1.00 50.00           C  \n",
      "ATOM    795  C   ASP A 265      59.409  50.697  32.647  1.00 50.00           C  \n",
      "ATOM    796  N   SER A 266      59.673  50.644  33.948  1.00 50.00           N  \n",
      "ATOM    797  CA  SER A 266      58.755  49.997  34.874  1.00 50.00           C  \n",
      "ATOM    798  C   SER A 266      57.453  50.759  35.063  1.00 50.00           C  \n",
      "ATOM    799  N   GLU A 267      57.474  52.063  34.806  1.00 50.00           N  \n",
      "ATOM    800  CA  GLU A 267      56.265  52.863  34.971  1.00 50.00           C  \n",
      "ATOM    801  C   GLU A 267      55.124  52.341  34.104  1.00 50.00           C  \n",
      "ATOM    802  N   THR A 268      55.475  51.789  32.945  1.00 50.00           N  \n",
      "ATOM    803  CA  THR A 268      54.470  51.328  31.981  1.00 50.00           C  \n",
      "ATOM    804  C   THR A 268      54.351  49.809  31.865  1.00 50.00           C  \n",
      "ATOM    805  N   ASN A 269      54.977  49.079  32.782  1.00 50.00           N  \n",
      "ATOM    806  CA  ASN A 269      54.645  47.673  32.953  1.00 50.00           C  \n",
      "ATOM    807  C   ASN A 269      53.175  47.641  33.369  1.00 50.00           C  \n",
      "ATOM    808  N   VAL A 270      52.433  46.627  32.955  1.00 50.00           N  \n",
      "ATOM    809  CA  VAL A 270      51.023  46.571  33.345  1.00 50.00           C  \n",
      "ATOM    810  C   VAL A 270      50.591  45.204  33.834  1.00 50.00           C  \n",
      "ATOM    811  N   ALA A 271      49.679  45.203  34.800  1.00 50.00           N  \n",
      "ATOM    812  CA  ALA A 271      48.909  44.014  35.094  1.00 50.00           C  \n",
      "ATOM    813  C   ALA A 271      47.779  44.048  34.072  1.00 50.00           C  \n",
      "ATOM    814  N   ILE A 272      47.823  43.137  33.110  1.00 50.00           N  \n",
      "ATOM    815  CA  ILE A 272      46.891  43.189  31.985  1.00 50.00           C  \n",
      "ATOM    816  C   ILE A 272      45.466  43.218  32.536  1.00 50.00           C  \n",
      "ATOM    817  N   GLY A 273      44.672  44.164  32.057  1.00 50.00           N  \n",
      "ATOM    818  CA  GLY A 273      43.370  44.402  32.658  1.00 50.00           C  \n",
      "ATOM    819  C   GLY A 273      42.242  43.547  32.139  1.00 50.00           C  \n",
      "ATOM    820  N   ILE A 274      41.080  43.746  32.759  1.00 50.00           N  \n",
      "ATOM    821  CA  ILE A 274      39.841  43.080  32.410  1.00 50.00           C  \n",
      "ATOM    822  C   ILE A 274      39.482  43.164  30.934  1.00 50.00           C  \n",
      "ATOM    823  N   VAL A 275      39.602  44.349  30.341  1.00 50.00           N  \n",
      "ATOM    824  CA  VAL A 275      39.250  44.502  28.918  1.00 50.00           C  \n",
      "ATOM    825  C   VAL A 275      39.969  43.453  28.034  1.00 50.00           C  \n",
      "ATOM    826  N   SER A 276      41.277  43.313  28.200  1.00 50.00           N  \n",
      "ATOM    827  CA  SER A 276      42.007  42.322  27.417  1.00 50.00           C  \n",
      "ATOM    828  C   SER A 276      41.782  40.901  27.902  1.00 50.00           C  \n",
      "ATOM    829  N   ALA A 277      41.764  40.722  29.219  1.00 50.00           N  \n",
      "ATOM    830  CA  ALA A 277      41.693  39.375  29.773  1.00 50.00           C  \n",
      "ATOM    831  C   ALA A 277      40.402  38.644  29.410  1.00 50.00           C  \n",
      "ATOM    832  N   VAL A 278      39.298  39.374  29.297  1.00 50.00           N  \n",
      "ATOM    833  CA  VAL A 278      38.051  38.734  28.903  1.00 50.00           C  \n",
      "ATOM    834  C   VAL A 278      38.246  38.074  27.538  1.00 50.00           C  \n",
      "ATOM    835  N   HIS A 279      39.095  38.682  26.717  1.00 50.00           N  \n",
      "ATOM    836  CA  HIS A 279      39.293  38.216  25.357  1.00 50.00           C  \n",
      "ATOM    837  C   HIS A 279      40.505  37.299  25.170  1.00 50.00           C  \n",
      "ATOM    838  N   THR A 280      41.499  37.392  26.055  1.00 50.00           N  \n",
      "ATOM    839  CA  THR A 280      42.669  36.513  25.951  1.00 50.00           C  \n",
      "ATOM    840  C   THR A 280      42.558  35.265  26.804  1.00 50.00           C  \n",
      "ATOM    841  N   GLY A 281      41.704  35.318  27.828  1.00 50.00           N  \n",
      "ATOM    842  CA  GLY A 281      41.749  34.346  28.896  1.00 50.00           C  \n",
      "ATOM    843  C   GLY A 281      42.650  34.878  29.990  1.00 50.00           C  \n",
      "ATOM    844  N   PRO A 282      42.736  34.159  31.117  1.00 50.00           N  \n",
      "ATOM    845  CA  PRO A 282      43.486  34.675  32.263  1.00 50.00           C  \n",
      "ATOM    846  C   PRO A 282      45.002  34.488  32.156  1.00 50.00           C  \n",
      "ATOM    847  N   TYR A 283      45.519  33.923  31.065  1.00 50.00           N  \n",
      "ATOM    848  CA  TYR A 283      46.970  33.671  31.004  1.00 50.00           C  \n",
      "ATOM    849  C   TYR A 283      47.769  34.386  29.903  1.00 50.00           C  \n",
      "ATOM    850  N   PRO A 284      47.461  35.664  29.622  1.00 50.00           N  \n",
      "ATOM    851  CA  PRO A 284      48.296  36.349  28.627  1.00 50.00           C  \n",
      "ATOM    852  C   PRO A 284      49.626  36.777  29.240  1.00 50.00           C  \n",
      "ATOM    853  N   ILE A 285      50.686  36.828  28.439  1.00 50.00           N  \n",
      "ATOM    854  CA  ILE A 285      51.960  37.343  28.929  1.00 50.00           C  \n",
      "ATOM    855  C   ILE A 285      52.782  37.812  27.739  1.00 50.00           C  \n",
      "ATOM    856  N   GLY A 286      53.465  38.934  27.888  1.00 50.00           N  \n",
      "ATOM    857  CA  GLY A 286      54.163  39.485  26.742  1.00 50.00           C  \n",
      "ATOM    858  C   GLY A 286      54.806  40.808  27.041  1.00 50.00           C  \n",
      "ATOM    859  N   LEU A 287      55.241  41.481  25.985  1.00 50.00           N  \n",
      "ATOM    860  CA  LEU A 287      55.911  42.758  26.132  1.00 50.00           C  \n",
      "ATOM    861  C   LEU A 287      55.447  43.692  25.034  1.00 50.00           C  \n",
      "ATOM    862  N   GLY A 288      55.502  44.988  25.326  1.00 50.00           N  \n",
      "ATOM    863  CA  GLY A 288      55.192  46.025  24.363  1.00 50.00           C  \n",
      "ATOM    864  C   GLY A 288      56.392  46.933  24.182  1.00 50.00           C  \n",
      "ATOM    865  N   VAL A 289      56.514  47.509  22.995  1.00 50.00           N  \n",
      "ATOM    866  CA  VAL A 289      57.610  48.431  22.735  1.00 50.00           C  \n",
      "ATOM    867  C   VAL A 289      57.059  49.613  21.970  1.00 50.00           C  \n",
      "ATOM    868  N   ILE A 290      57.738  50.742  22.075  1.00 50.00           N  \n",
      "ATOM    869  CA  ILE A 290      57.373  51.910  21.288  1.00 50.00           C  \n",
      "ATOM    870  C   ILE A 290      58.602  52.775  21.123  1.00 50.00           C  \n",
      "ATOM    871  N   ARG A 291      58.644  53.561  20.049  1.00 50.00           N  \n",
      "ATOM    872  CA  ARG A 291      59.791  54.433  19.801  1.00 50.00           C  \n",
      "ATOM    873  C   ARG A 291      59.806  55.552  20.830  1.00 50.00           C  \n",
      "ATOM    874  N   LYS A 292      60.998  55.966  21.253  1.00 50.00           N  \n",
      "ATOM    875  CA  LYS A 292      61.112  57.175  22.073  1.00 50.00           C  \n",
      "ATOM    876  C   LYS A 292      60.887  58.386  21.181  1.00 50.00           C  \n",
      "ATOM    877  N   TYR A 293      59.935  59.243  21.544  1.00 50.00           N  \n",
      "ATOM    878  CA  TYR A 293      59.625  60.395  20.690  1.00 50.00           C  \n",
      "ATOM    879  C   TYR A 293      60.865  61.228  20.325  1.00 50.00           C  \n",
      "ATOM    880  N   GLU A 294      61.830  61.310  21.241  1.00 50.00           N  \n",
      "ATOM    881  CA  GLU A 294      63.047  62.121  21.045  1.00 50.00           C  \n",
      "ATOM    882  C   GLU A 294      63.828  61.715  19.808  1.00 50.00           C  \n",
      "ATOM    883  N   THR A 295      63.640  60.471  19.389  1.00 50.00           N  \n",
      "ATOM    884  CA  THR A 295      64.444  59.892  18.325  1.00 50.00           C  \n",
      "ATOM    885  C   THR A 295      63.779  60.037  16.958  1.00 50.00           C  \n",
      "ATOM    886  N   LEU A 296      62.556  60.556  16.944  1.00 50.00           N  \n",
      "ATOM    887  CA  LEU A 296      61.786  60.603  15.706  1.00 50.00           C  \n",
      "ATOM    888  C   LEU A 296      62.182  61.769  14.812  1.00 50.00           C  \n",
      "ATOM    889  N   ALA A 297       0.000   0.000   0.000  1.00 50.00           N  \n",
      "ATOM    890  CA  ALA A 297       0.000   0.000   0.000  1.00 50.00           C  \n",
      "ATOM    891  C   ALA A 297       0.000   0.000   0.000  1.00 50.00           C  \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# pick a random protein to visualize\n",
    "protein_row = processed_df.iloc[20]\n",
    "\n",
    "# print out the PDB content so we can see it\n",
    "pdb_content = format_pdb_from_df(protein_row)\n",
    "print(pdb_content)\n",
    "\n",
    "# Write PDB to a file\n",
    "with open(\"./data/sample_protein.pdb\", \"w\") as file:\n",
    "    file.write(pdb_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e04a023-6545-416c-8848-525c40591627",
   "metadata": {},
   "source": [
    "## Now we can visualize our protein structure with NGLview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "413fba38-fba4-4306-b8f2-be50c64f0d56",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "59265b4d55ca4c1195154d9712f39036",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load your PDB file\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure('Sample', './data/sample_protein.pdb')\n",
    "\n",
    "# Create NGLView widget for the structure\n",
    "view = nv.show_biopython(structure)\n",
    "view.clear_representations()\n",
    "\n",
    "# Add molecular graphics: e.g., cartoon for secondary structures and ball+stick for atoms\n",
    "view.add_representation('cartoon', selection='protein', color='blue')\n",
    "view.add_representation('ball+stick', selection='ligand')\n",
    "\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fd4e42a-7736-4f04-969c-e7821c6137ce",
   "metadata": {},
   "source": [
    "## Export data to JSON for model input:\n",
    "### The following code exports the data to the JSONL format required by Struct2Seq."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3d42164-5bec-45d7-8748-0bcebd2e656c",
   "metadata": {},
   "source": [
    "#### First we need to zip the coordinates to (x,y,z) format and divide by 100 to convert to Angstroms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d00ec7a6-33c8-4194-8efa-43c5a851791e",
   "metadata": {},
   "outputs": [],
   "source": [
    "backup=processed_df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f82499e1-3da9-4b46-8539-3c880a1b5d42",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating x coords: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 25299/25299 [00:03<00:00, 6719.45it/s]\n",
      "Calculating y coords: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 25299/25299 [00:03<00:00, 6714.57it/s]\n",
      "Calculating z coords: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 25299/25299 [00:03<00:00, 6737.50it/s]\n",
      "Calculating zipped coords: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 25299/25299 [00:00<00:00, 28983.60it/s]\n"
     ]
    }
   ],
   "source": [
    "df_features = processed_df\n",
    "df_features.rename(columns={#\"Sequence\": \"Node_Sequence\",\n",
    "                            \"Mask\": \"Node_Mask\",\n",
    "                            \"ID\": \"Protein_Name\",}, inplace=True)\n",
    "# rescale distances to Angstroms\n",
    "tqdm.pandas(desc=\"Calculating x coords\")\n",
    "df_features['x'] = df_features['x'].progress_apply(lambda x: [round(val / 1, 3) for val in x])\n",
    "tqdm.pandas(desc=\"Calculating y coords\")\n",
    "df_features['y'] = df_features['y'].progress_apply(lambda x: [round(val / 1, 3) for val in x])\n",
    "tqdm.pandas(desc=\"Calculating z coords\")\n",
    "df_features['z'] = df_features['z'].progress_apply(lambda x: [round(val / 1, 3) for val in x])\n",
    "tqdm.pandas(desc=\"Calculating zipped coords\")\n",
    "df_features['Atom_Coordinates'] = df_features.progress_apply(lambda row: list(zip(row['x'], row['y'], row['z'])), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "89a687ec-bacb-4c8e-8411-2c9df6da4612",
   "metadata": {},
   "outputs": [],
   "source": [
    "def protein_to_jsonl(row):\n",
    "    # Split atom coordinates into groups by atom type: N, CA, C\n",
    "    n_coords = row['Atom_Coordinates'][0::3]  # Assuming every third starting from 0 is N\n",
    "    ca_coords = row['Atom_Coordinates'][1::3]  # Assuming every third starting from 1 is CA\n",
    "    c_coords = row['Atom_Coordinates'][2::3]  # Assuming every third starting from 2 is C\n",
    "\n",
    "    # Create a dictionary to hold the protein data\n",
    "    protein_data = {\n",
    "        \"name\": row['Protein_Name'],\n",
    "        \"conserved\": row['Node_Mask'],\n",
    "        \"num_chains\": 1,  # Assuming all entries have one chain; adjust as necessary\n",
    "        \"seq\": row['Sequence'],\n",
    "        \"coords\": {\n",
    "            \"N\": n_coords,\n",
    "            \"CA\": ca_coords,\n",
    "            \"C\": c_coords\n",
    "        }\n",
    "    }\n",
    "    return json.dumps(protein_data)  # Convert dictionary to JSON string"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9f3ef9b-bc71-4ea7-a7e3-1adbd9235b80",
   "metadata": {},
   "source": [
    "#### Then export to JSON at the path expected by Struct2Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8741aba4-fbe4-4bdd-909e-d0ceb8bdaf80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply the function to each row and save the output as a JSONL file\n",
    "jsonl_data = df_features.apply(protein_to_jsonl, axis=1)\n",
    "\n",
    "# Write to a JSONL file\n",
    "with open('./data/cath/chain_set.jsonl', 'w') as outfile:\n",
    "    for entry in jsonl_data:\n",
    "        outfile.write(entry + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb5b9931-de67-49e7-b269-e8e05de142db",
   "metadata": {},
   "source": [
    "#### Finally, create a train-test split JSON at the path required by Struct2Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7574cabc-4b1f-4792-beca-4a3df1ad8a9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shuffle the DataFrame\n",
    "df_shuffled = df_features.sample(frac=1).reset_index(drop=True)\n",
    "\n",
    "# Calculate split indices\n",
    "train_end = int(len(df_shuffled) * 0.85)\n",
    "val_end = train_end + int(len(df_shuffled) * 0.10)\n",
    "\n",
    "# Split the DataFrame into train, validation, and test\n",
    "train_proteins = df_shuffled.iloc[:train_end]['Protein_Name'].tolist()\n",
    "val_proteins = df_shuffled.iloc[train_end:val_end]['Protein_Name'].tolist()\n",
    "test_proteins = df_shuffled.iloc[val_end:]['Protein_Name'].tolist()\n",
    "\n",
    "# Create dictionary to be written as JSON\n",
    "split_dict = {\n",
    "    'train': train_proteins,\n",
    "    'validation': val_proteins,\n",
    "    'test': test_proteins\n",
    "}\n",
    "\n",
    "# Write to JSON file\n",
    "with open('./data/cath/chain_set_splits.json', 'w') as json_file:\n",
    "    json.dump(split_dict, json_file, indent=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3780b7ba-0763-4e42-86c3-c6f4cae2e364",
   "metadata": {},
   "source": [
    "# Struct2Seq model training:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65413859-b9de-4492-8cd5-2d3701455ce4",
   "metadata": {},
   "source": [
    "#### To train the model, navigate to the `/experiments` directory and run the following command:\n",
    "`python3 ./experiments/train_s2s.py --mpnn --cuda`\n",
    "#### The model will train for 100 epochs by default and periodically log progress.\n",
    "#### To test the model, from the same directory, run:\n",
    "`python3 test_redesign.py --mpnn --restore ./log/best_checkpoint_epoch94.pt`\n",
    "#### Make sure you adjust the filename for the best checkpoint. The model will save [FASTA](https://en.wikipedia.org/wiki/FASTA_format) files containing the generated multiple sequence alignments to the logs folder."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "843e67cc-ff58-474c-bbac-f964922942ea",
   "metadata": {},
   "source": [
    "## `Struct2Seq` Model architecture:\n",
    "\n",
    "### 1. Featurization Layers\n",
    "- **ProteinFeatures**: Responsible for processing the raw protein data into usable features.\n",
    "  - **Node Features**: Number specified at initialization.\n",
    "  - **Edge Features**: Number specified at initialization.\n",
    "  - **Top-k Neighbors**: Determines the top k-nearest neighbors for each node for graph message passing.\n",
    "  - **Dropout**: Regularization parameter to prevent overfitting.\n",
    "\n",
    "### 2. Embedding Layers\n",
    "- **Node Embedding (W_v)**: Linear transformation layer for node features.\n",
    "  - **Input Features**: Node features dimension.\n",
    "  - **Output Features**: Hidden dimension size. (Default 128)\n",
    "      \n",
    "- **Edge Embedding (W_e)**: Linear transformation layer for edge features.\n",
    "  - **Input Features**: Edge features dimension.\n",
    "  - **Output Features**: Hidden dimension size. (Default 128)\n",
    "  \n",
    "- **Sequence Embedding (W_s)**: Embedding layer for sequence data.\n",
    "  - **Vocabulary Size**: 20 (for the 20 standard amino acids).\n",
    "  - **Embedding Dimension**: Hidden dimension size. \n",
    "\n",
    "### 3. Transformer/MPNN Layers\n",
    "- **Encoder Layers**: Multiple layers (number specified at initialization) to encode context.\n",
    "  - **Input/Output Features**: Both set to hidden dimension size multiplied by 2 for Transformers, consistent for MPNN.\n",
    "  - **Dropout**: Used within each layer to manage overfitting.\n",
    "  \n",
    "- **Decoder Layers**: Handles sequence prediction with autoregressive properties.\n",
    "  - **Input Features**: Hidden dimension size.\n",
    "  - **Output Features**: Hidden dimension size multiplied by 3 for Transformers, consistent for MPNN.\n",
    "  - **Dropout**: Applied within each layer.\n",
    "\n",
    "### 4. Output Layer\n",
    "- **Linear Transformation (W_out)**: Converts hidden states from the decoder to logits.\n",
    "  - **Input Features**: Hidden dimension size.\n",
    "  - **Output Features**: Number of letters (potential output classes, 20 for amino acids).\n",
    "\n",
    "### 5. Additional Mechanisms\n",
    "- **Autoregressive Masking**: Used in the decoder to ensure that the prediction for a position can only depend on known previous positions.\n",
    "- **Forward Attention Decoder**: Optional mechanism that can be toggled to modify attention behavior in the decoder. Unused when MPNN is active.\n",
    "\n",
    "## Extensions:\n",
    "#### The following features were added to `/struct2seq/protein_features.py` which are computed at runtime to enhance the GNN with additional features for message passing (MPNN).\n",
    "- `Solubility Interactions`: Every amino acid (node) is classified as hydrophobic (-1), hydrophilic (+1), or neutral (0) and similarly, edges between nodes are classified by the outer product of these interactions.\n",
    "- `Charge Interactions`: Every amino acid (node) is classified as positivly charged (+1), negatively charged (-1), or neutral (0) and similarly, edges between nodes are classified by the outer product of these interactions.\n",
    "- `Conservation Interactions`: Every amino acid (node) is classified as highly conserved (1), or not highly conserved (0) and similarly, edges between nodes are classified by the outer product of these interactions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48ba1fe0-8034-4cf2-bcf9-9db66b7be0a9",
   "metadata": {},
   "source": [
    "# Results:\n",
    "- Training parameters:\n",
    "  - Hidden dimensions 128\n",
    "  - K Neighbors for message passing 30\n",
    "  - Vocab size 20\n",
    "  - Batch tokens 2500\n",
    "  - Epochs 100\n",
    "  - Dropout 0.1\n",
    "  - Loss smoothing 0.1\n",
    "- ### Perplexity Scores:\n",
    "  - ### Train: 5.07\n",
    "  - ### Validation: 5.34\n",
    "  - ### Test: 5.31\n",
    "### Interpretation:\n",
    "Every time the model generates a sequence, there are 5 candidate sequences with equal probablility which could be provided as output. Since we are trying to produce novel proteins, this seems like an acceptable amount of entropy. It would be interesting to see if more training lowers the uncertainty although the loss curve  doesn't suggest that this is the case."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d4be2606-85f0-4422-b634-09f2e7be56a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcYAAAE6CAYAAACMMvJLAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABPhElEQVR4nO3dd3hUZdrH8e+0zGTSCYQkEEKoAaSKsBQFFKkWREURBBRQBFREBTvYQN0VeS2Ly+5SFFTWxroWlC4KQgRBFKRoCCVEICE9mcnMPO8fJzNkSAJpMGFyf65rriSn3nMY8svznOeco1NKKYQQQggBgN7XBQghhBC1iQSjEEIIUYIEoxBCCFGCBKMQQghRggSjEEIIUYIEoxBCCFGCBKMQQghRggSjEEIIUYIEoxBCCFGCBKMol06nq9Brw4YN1drP7Nmz0el0VVp3w4YNNVJDbTdu3DiaNm1a7vyTJ08SEBDA7bffXu4y2dnZWK1Wbrjhhgrvd8mSJeh0Og4dOlThWkrS6XTMnj27wvtzS01NZfbs2ezcubPUvOp8XqqradOmXHfddT7Zt7h4jL4uQNReW7Zs8fr5+eefZ/369axbt85retu2bau1nwkTJjBo0KAqrdulSxe2bNlS7RoudQ0aNOCGG25g5cqVnD59moiIiFLLfPDBBxQUFDB+/Phq7evpp5/mwQcfrNY2zic1NZVnn32Wpk2b0qlTJ6951fm8CFEREoyiXH/5y1+8fm7QoAF6vb7U9LPl5+djtVorvJ/GjRvTuHHjKtUYGhp63nrqivHjx/Pxxx+zfPlypk6dWmr+okWLaNiwIUOHDq3Wfpo3b16t9aurOp8XISpCulJFtfTt25fLLruMb7/9lp49e2K1Wrn77rsBWLFiBQMGDCAmJobAwEDatGnDY489Rl5entc2yuoac3dZrVq1ii5duhAYGEhiYiKLFi3yWq6srtRx48YRHBzMwYMHGTJkCMHBwcTFxfHwww9js9m81j969Ci33HILISEhhIeHM2rUKJKSktDpdCxZsuSc7/3kyZNMnjyZtm3bEhwcTFRUFFdffTWbNm3yWu7QoUPodDr+9re/MW/ePBISEggODqZHjx788MMPpba7ZMkSWrdujdlspk2bNrzzzjvnrMNt4MCBNG7cmMWLF5eat3fvXrZu3cqYMWMwGo2sXr2aG2+8kcaNG2OxWGjRogX33nsvp06dOu9+yupKzc7OZuLEiURGRhIcHMygQYPYv39/qXUPHjzIXXfdRcuWLbFarTRq1Ijrr7+e3bt3e5bZsGEDV1xxBQB33XWXp8ve3SVb1ufF5XLxyiuvkJiYiNlsJioqijFjxnD06FGv5dyf16SkJK688kqsVivNmjXjpZdewuVynfe9V0RhYSGPP/44CQkJBAQE0KhRI6ZMmUJmZqbXcuvWraNv375ERkYSGBhIkyZNuPnmm8nPz/css2DBAjp27EhwcDAhISEkJibyxBNP1EidonzSYhTVdvz4cUaPHs2MGTOYM2cOer3299aBAwcYMmQI06ZNIygoiN9++42XX36Zbdu2leqOLcuuXbt4+OGHeeyxx2jYsCH/+te/GD9+PC1atOCqq64657pFRUXccMMNjB8/nocffphvv/2W559/nrCwMJ555hkA8vLy6NevHxkZGbz88su0aNGCVatWcdttt1XofWdkZAAwa9YsoqOjyc3N5dNPP6Vv376sXbuWvn37ei3/1ltvkZiYyPz58wGtS3LIkCEkJycTFhYGaKF41113ceONN/Lqq6+SlZXF7NmzsdlsnuNaHr1ez7hx43jhhRfYtWsXHTt29Mxzh6X7j5bff/+dHj16MGHCBMLCwjh06BDz5s2jd+/e7N69G5PJVKFjAKCUYtiwYWzevJlnnnmGK664gu+//57BgweXWjY1NZXIyEheeuklGjRoQEZGBkuXLqV79+789NNPtG7dmi5durB48WLuuusunnrqKU8L91ytxPvuu4+FCxcydepUrrvuOg4dOsTTTz/Nhg0b2LFjB/Xr1/csm5aWxqhRo3j44YeZNWsWn376KY8//jixsbGMGTOmwu/7XMdi7dq1PP7441x55ZX8/PPPzJo1iy1btrBlyxbMZjOHDh1i6NChXHnllSxatIjw8HCOHTvGqlWrsNvtWK1WPvjgAyZPnsz999/P3/72N/R6PQcPHmTPnj3VqlFUgBKigsaOHauCgoK8pvXp00cBau3atedc1+VyqaKiIrVx40YFqF27dnnmzZo1S539UYyPj1cWi0WlpKR4phUUFKh69eqpe++91zNt/fr1ClDr16/3qhNQ//nPf7y2OWTIENW6dWvPz2+99ZYC1FdffeW13L333qsAtXjx4nO+p7M5HA5VVFSkrrnmGnXTTTd5picnJytAtW/fXjkcDs/0bdu2KUC9//77SimlnE6nio2NVV26dFEul8uz3KFDh5TJZFLx8fHnreGPP/5QOp1OPfDAA55pRUVFKjo6WvXq1avMddz/NikpKQpQ//3vfz3zFi9erACVnJzsmTZ27FivWr766isFqP/7v//z2u6LL76oADVr1qxy63U4HMput6uWLVuqhx56yDM9KSmp3H+Dsz8ve/fuVYCaPHmy13Jbt25VgHriiSc809yf161bt3ot27ZtWzVw4MBy63SLj49XQ4cOLXf+qlWrFKBeeeUVr+krVqxQgFq4cKFSSqmPPvpIAWrnzp3lbmvq1KkqPDz8vDWJmiddqaLaIiIiuPrqq0tN/+OPP7jjjjuIjo7GYDBgMpno06cPoHXtnU+nTp1o0qSJ52eLxUKrVq1ISUk577o6nY7rr7/ea1qHDh281t24cSMhISGlBnKMHDnyvNt3e/vtt+nSpQsWiwWj0YjJZGLt2rVlvr+hQ4diMBi86gE8Ne3bt4/U1FTuuOMOr67C+Ph4evbsWaF6EhIS6NevH8uXL8dutwPw1VdfkZaW5mktApw4cYJJkyYRFxfnqTs+Ph6o2L9NSevXrwdg1KhRXtPvuOOOUss6HA7mzJlD27ZtCQgIwGg0EhAQwIEDByq937P3P27cOK/p3bp1o02bNqxdu9ZrenR0NN26dfOadvZno6rcPSFn13LrrbcSFBTkqaVTp04EBARwzz33sHTpUv74449S2+rWrRuZmZmMHDmS//73vxXq5hY1Q4JRVFtMTEypabm5uVx55ZVs3bqVF154gQ0bNpCUlMQnn3wCQEFBwXm3GxkZWWqa2Wyu0LpWqxWLxVJq3cLCQs/P6enpNGzYsNS6ZU0ry7x587jvvvvo3r07H3/8MT/88ANJSUkMGjSozBrPfj9msxk4cyzS09MB7Rf32cqaVp7x48eTnp7OZ599BmjdqMHBwYwYMQLQzscNGDCATz75hBkzZrB27Vq2bdvmOd9ZkeNbUnp6OkajsdT7K6vm6dOn8/TTTzNs2DD+97//sXXrVpKSkujYsWOl91ty/1D25zA2NtYz3606n6uK1GI0GmnQoIHXdJ1OR3R0tKeW5s2bs2bNGqKiopgyZQrNmzenefPm/N///Z9nnTvvvJNFixaRkpLCzTffTFRUFN27d2f16tXVrlOcm5xjFNVW1jVl69atIzU1lQ0bNnhaiUCpAQi+FBkZybZt20pNT0tLq9D6y5Yto2/fvixYsMBrek5OTpXrKW//Fa0JYPjw4URERLBo0SL69OnD559/zpgxYwgODgbgl19+YdeuXSxZsoSxY8d61jt48GCV63Y4HKSnp3uFTlk1L1u2jDFjxjBnzhyv6adOnSI8PLzK+wftXPfZ5yFTU1O9zi9eaO5jcfLkSa9wVEqRlpbmGVQEcOWVV3LllVfidDr58ccfeeONN5g2bRoNGzb0XI961113cdddd5GXl8e3337LrFmzuO6669i/f7+nhS9qnrQYxQXhDkt3q8jtH//4hy/KKVOfPn3Iycnhq6++8pr+wQcfVGh9nU5X6v39/PPPpa7/rKjWrVsTExPD+++/j1LKMz0lJYXNmzdXeDsWi4U77riDb775hpdffpmioiKvbtSa/rfp168fAMuXL/ea/t5775Vatqxj9sUXX3Ds2DGvaWe3ps/F3Y2/bNkyr+lJSUns3buXa6655rzbqCnufZ1dy8cff0xeXl6ZtRgMBrp3785bb70FwI4dO0otExQUxODBg3nyySex2+38+uuvF6B64SYtRnFB9OzZk4iICCZNmsSsWbMwmUwsX76cXbt2+bo0j7Fjx/Laa68xevRoXnjhBVq0aMFXX33F119/DXDeUaDXXXcdzz//PLNmzaJPnz7s27eP5557joSEBBwOR6Xr0ev1PP/880yYMIGbbrqJiRMnkpmZyezZsyvVlQpad+pbb73FvHnzSExM9DpHmZiYSPPmzXnsscdQSlGvXj3+97//VbmLbsCAAVx11VXMmDGDvLw8unbtyvfff8+7775batnrrruOJUuWkJiYSIcOHdi+fTt//etfS7X0mjdvTmBgIMuXL6dNmzYEBwcTGxtLbGxsqW22bt2ae+65hzfeeAO9Xs/gwYM9o1Lj4uJ46KGHqvS+ypOWlsZHH31UanrTpk259tprGThwIDNnziQ7O5tevXp5RqV27tyZO++8E9DOTa9bt46hQ4fSpEkTCgsLPZci9e/fH4CJEycSGBhIr169iImJIS0tjblz5xIWFubV8hQXgI8H/4hLSHmjUtu1a1fm8ps3b1Y9evRQVqtVNWjQQE2YMEHt2LGj1GjD8kalljX6r0+fPqpPnz6en8sblXp2neXt5/Dhw2r48OEqODhYhYSEqJtvvll9+eWXpUZnlsVms6lHHnlENWrUSFksFtWlSxe1cuXKUqM23aNS//rXv5baBmWM2vzXv/6lWrZsqQICAlSrVq3UokWLSm2zIjp37lzmCEmllNqzZ4+69tprVUhIiIqIiFC33nqrOnz4cKl6KjIqVSmlMjMz1d13363Cw8OV1WpV1157rfrtt99Kbe/06dNq/PjxKioqSlmtVtW7d2+1adOmUv+uSin1/vvvq8TERGUymby2U9a/o9PpVC+//LJq1aqVMplMqn79+mr06NHqyJEjXsuV93mt6PGNj49XQJmvsWPHKqW00dMzZ85U8fHxymQyqZiYGHXfffep06dPe7azZcsWddNNN6n4+HhlNptVZGSk6tOnj/rss888yyxdulT169dPNWzYUAUEBKjY2Fg1YsQI9fPPP5+3TlE9OqVK9NkIIZgzZw5PPfUUhw8fljusCFEHSVeqqNPefPNNQOteLCoqYt26dbz++uuMHj1aQlGIOkqCUdRpVquV1157jUOHDmGz2WjSpAkzZ87kqaee8nVpQggfka5UIYQQogS5XEMIIYQoQYJRCCGEKEGCUQghhCjB7wffuFwuUlNTCQkJKfPWZUIIIeoGpRQ5OTnExsae8wYefh+MqampxMXF+boMIYQQtcSRI0fOeTmW3wdjSEgIoB2I0NBQH1cjhBDCV7Kzs4mLi/PkQnn8Phjd3aehoaESjEIIIc57Wk0G3wghhBAl+DQYv/32W66//npiY2PR6XSsXLnSa/4nn3zCwIEDqV+/Pjqdjp07d/qkTiGEEHWHT4MxLy+Pjh07eu5XWdb8Xr168dJLL13kyoQQQtRVPj3HOHjwYAYPHlzufPezyw4dOnSRKhJCXGhKKRwOB06n09elCD9jMBgwGo3VvjTP7wbf2Gw2bDab5+fs7GwfViOEKMlut3P8+HHy8/N9XYrwU1arlZiYGAICAqq8Db8Lxrlz5/Lss8/6ugwhxFlcLhfJyckYDAZiY2MJCAiQm26IGqOUwm63c/LkSZKTk2nZsuU5L+I/F78Lxscff5zp06d7fnZft1Idn/+cypvrDtKjeSSzrm9X3RKFqJPsdjsul4u4uDisVquvyxF+KDAwEJPJREpKCna7HYvFUqXt+F0wms1mzGZzjW4zp9DBb2k5NI6Q/8xCVFdV/4oXoiJq4vMln9AKsJi0w2RzyGABIYTwdz5tMebm5nLw4EHPz8nJyezcuZN69erRpEkTMjIyOHz4MKmpqQDs27cPgOjoaKKjoy9anWajAQBbkeui7VMIIYRv+LTF+OOPP9K5c2c6d+4MwPTp0+ncuTPPPPMMAJ999hmdO3dm6NChANx+++107tyZt99++6LW6W4xFkqLUQhRQ/r27cu0adN8XYYog09bjH379kUpVe78cePGMW7cuItXUDmkxShE3XW+kbNjx45lyZIlld7uJ598gslkqmJVmnHjxpGZmVnqrmGievxu8M2FIC1GIequ48ePe75fsWIFzzzzjOe0DmgjIUsqKiqqUODVq1ev5ooUNUoG31SAu8VYWCTBKERNUkqRb3dc9Ne5eqrO5h7TEB0dTVhYGDqdzvNzYWEh4eHh/Oc//6Fv375YLBaWLVtGeno6I0eOpHHjxlitVtq3b8/777/vtd2zu1KbNm3KnDlzuPvuuwkJCaFJkyYsXLiwWsd348aNdOvWDbPZTExMDI899hgOh8Mz/6OPPqJ9+/YEBgYSGRlJ//79ycvLA2DDhg1069aNoKAgwsPD6dWrFykpKdWq51IhLcYKODMqVbpShahJBUVO2j7z9UXf757nBmINqLlffzNnzuTVV19l8eLFmM1mCgsLufzyy5k5cyahoaF88cUX3HnnnTRr1ozu3buXu51XX32V559/nieeeIKPPvqI++67j6uuuorExMRK13Ts2DGGDBnCuHHjeOedd/jtt9+YOHEiFouF2bNnc/z4cUaOHMkrr7zCTTfdRE5ODps2bfLcsm/YsGFMnDiR999/H7vdzrZt2+rMDRkkGCtAWoxCiHOZNm0aw4cP95r2yCOPeL6///77WbVqFR9++OE5g3HIkCFMnjwZ0ML2tddeY8OGDVUKxr///e/ExcXx5ptvotPpSExMJDU1lZkzZ/LMM89w/PhxHA4Hw4cPJz4+HoD27dsDkJGRQVZWFtdddx3NmzcHoE2bNpWu4VIlwVgB5hItRqVUnfmrSYgLLdBkYM9zA32y35rUtWtXr5+dTicvvfQSK1as4NixY557OAcFBZ1zOx06dPB87+6yPXHiRJVq2rt3Lz169PD6fdWrVy9yc3M5evQoHTt25JprrqF9+/YMHDiQAQMGcMsttxAREUG9evUYN24cAwcO5Nprr6V///6MGDGCmJiYKtVyqZFzjBVgKf5PpBTYndKdKkRN0el0WAOMF/1V03/cnh14r776Kq+99hozZsxg3bp17Ny5k4EDB2K328+5nbMH7eh0Olyuqv3OKeuPePe5VZ1Oh8FgYPXq1Xz11Ve0bduWN954g9atW5OcnAzA4sWL2bJlCz179mTFihW0atWKH374oUq1XGokGCvAbDxzmOQ8oxDifDZt2sSNN97I6NGj6dixI82aNePAgQMXtYa2bduyefNmr4FGmzdvJiQkhEaNGgFaQPbq1Ytnn32Wn376iYCAAD799FPP8p07d+bxxx9n8+bNXHbZZbz33nsX9T34inSlVkCAQY9Op7UYC4uchFqqd+2REMK/tWjRgo8//pjNmzcTERHBvHnzSEtLuyDn6bKysti5c6fXtHr16jF58mTmz5/P/fffz9SpU9m3bx+zZs1i+vTp6PV6tm7dytq1axkwYABRUVFs3bqVkydP0qZNG5KTk1m4cCE33HADsbGx7Nu3j/379zNmzJgar782kmCsAJ1Oh9mop7DIJRf5CyHO6+mnnyY5OZmBAwditVq55557GDZsGFlZWTW+rw0bNnjuHubmvunAl19+yaOPPkrHjh2pV68e48eP56mnngIgNDSUb7/9lvnz55OdnU18fDyvvvoqgwcP5s8//+S3335j6dKlpKenExMTw9SpU7n33ntrvP7aSKcqc0HPJSg7O5uwsDCysrIIDQ2t8nY6PfcNmflFrJl+FS2iQmqwQiHqhsLCQpKTk0lISKjy44CEOJ9zfc4qmgdyjrGC3OcZC6XFKIQQfk2CsYLcI1Pl0VNCCOHfJBgryN1ilHOMQgjh3yQYK8jdYpQbiQshhH+TYKwgaTEKIUTdIMFYQdJiFEKIukGCsYKkxSiEEHWDBGMFmU3yhA0hhKgLJBgryHMdo9wrVQgh/JoEYwV5rmOUrlQhRBX07duXadOmeX5u2rQp8+fPP+c6Op2OlStXVnvfNbWdukKCsYIsRhl8I0RddP3119O/f/8y523ZsgWdTseOHTsqvd2kpCTuueee6pbnZfbs2XTq1KnU9OPHjzN48OAa3dfZlixZQnh4+AXdx8UiwVhBnocVS4tRiDpl/PjxrFu3jpSUlFLzFi1aRKdOnejSpUult9ugQQOsVmtNlHhe0dHRmM3mi7IvfyDBWEHSYhTiAlAK7HkX/1WJZydcd911REVFsWTJEq/p+fn5rFixgvHjx5Oens7IkSNp3LgxVquV9u3b8/77759zu2d3pR44cICrrroKi8VC27ZtWb16dal1Zs6cSatWrbBarTRr1oynn36aoqIiQGuxPfvss+zatQudTodOp/PUfHZX6u7du7n66qsJDAwkMjKSe+65h9zcXM/8cePGMWzYMP72t78RExNDZGQkU6ZM8eyrKg4fPsyNN95IcHAwoaGhjBgxgj///NMzf9euXfTr14+QkBBCQ0O5/PLL+fHHHwFISUnh+uuvJyIigqCgINq1a8eXX35Z5VrORx47VUHSYhTiAijKhzmxF3+/T6RCQFCFFjUajYwZM4YlS5bwzDPPoNPpAPjwww+x2+2MGjWK/Px8Lr/8cmbOnEloaChffPEFd955J82aNaN79+7n3YfL5WL48OHUr1+fH374gezsbK/zkW4hISEsWbKE2NhYdu/ezcSJEwkJCWHGjBncdttt/PLLL6xatYo1a9YAEBYWVmob+fn5DBo0iL/85S8kJSVx4sQJJkyYwNSpU73Cf/369cTExLB+/XoOHjzIbbfdRqdOnZg4cWKFjltJSimGDRtGUFAQGzduxOFwMHnyZG677TY2bNgAwKhRo+jcuTMLFizAYDCwc+dOTCbt2bdTpkzBbrfz7bffEhQUxJ49ewgODq50HRUlwVhBFs+oVGkxClHX3H333fz1r39lw4YN9OvXD9C6UYcPH05ERAQRERE88sgjnuXvv/9+Vq1axYcfflihYFyzZg179+7l0KFDNG7cGIA5c+aUOi/ofpYiaC3Ohx9+mBUrVjBjxgwCAwMJDg7GaDQSHR1d7r6WL19OQUEB77zzDkFB2h8Hb775Jtdffz0vv/wyDRs2BCAiIoI333wTg8FAYmIiQ4cOZe3atVUKxjVr1vDzzz+TnJxMXFwcAO+++y7t2rUjKSmJK664gsOHD/Poo4+SmJgIQMuWLT3rHz58mJtvvpn27dsD0KxZs0rXUBk+DcZvv/2Wv/71r2zfvp3jx4/z6aefMmzYMM98pRTPPvssCxcu5PTp03Tv3p233nqLdu3aXfRazTIqVYiaZ7JqrTdf7LcSEhMT6dmzJ4sWLaJfv378/vvvbNq0iW+++QYAp9PJSy+9xIoVKzh27Bg2mw2bzeYJnvPZu3cvTZo08YQiQI8ePUot99FHHzF//nwOHjxIbm4uDoej0s+Z3bt3Lx07dvSqrVevXrhcLvbt2+cJxnbt2mEwGDzLxMTEsHv37krtq+Q+4+LiPKEI0LZtW8LDw9m7dy9XXHEF06dPZ8KECbz77rv079+fW2+9lebNmwPwwAMPcN999/HNN9/Qv39/br75Zjp06FClWirCp+cY8/Ly6NixI2+++WaZ81955RXmzZvHm2++SVJSEtHR0Vx77bXk5ORc5ErB4u5KlRajEDVHp9O6NC/2q7g7tDLGjx/Pxx9/THZ2NosXLyY+Pp5rrrkGgFdffZXXXnuNGTNmsG7dOnbu3MnAgQOx2+0V2nZZz4vXnVXjDz/8wO23387gwYP5/PPP+emnn3jyyScrvI+S+zp722Xt092NWXKey1W1hkF5+yw5ffbs2fz6668MHTqUdevW0bZtWz799FMAJkyYwB9//MGdd97J7t276dq1K2+88UaVaqkInwbj4MGDeeGFFxg+fHipeUop5s+fz5NPPsnw4cO57LLLWLp0Kfn5+bz33nsXvVazUVqMQtRlI0aMwGAw8N5777F06VLuuusuzy/1TZs2ceONNzJ69Gg6duxIs2bNOHDgQIW33bZtWw4fPkxq6pnW85YtW7yW+f7774mPj+fJJ5+ka9eutGzZstRI2YCAAJzOc//x3rZtW3bu3EleXp7XtvV6Pa1atapwzZXhfn9HjhzxTNuzZw9ZWVm0adPGM61Vq1Y89NBDfPPNNwwfPpzFixd75sXFxTFp0iQ++eQTHn74Yf75z39ekFqhFo9KTU5OJi0tjQEDBnimmc1m+vTpw+bNm8tdz2azkZ2d7fWqCe4Wo5xjFKJuCg4O5rbbbuOJJ54gNTWVcePGeea1aNGC1atXs3nzZvbu3cu9995LWlpahbfdv39/WrduzZgxY9i1axebNm3iySef9FqmRYsWHD58mA8++IDff/+d119/3dOicmvatCnJycns3LmTU6dOYbPZSu1r1KhRWCwWxo4dyy+//ML69eu5//77ufPOOz3dqFXldDrZuXOn12vPnj3079+fDh06MGrUKHbs2MG2bdsYM2YMffr0oWvXrhQUFDB16lQ2bNhASkoK33//PUlJSZ7QnDZtGl9//TXJycns2LGDdevWeQVqTau1wej+UJ39D9WwYcNzfuDmzp1LWFiY51WyT7s6pMUohBg/fjynT5+mf//+NGnSxDP96aefpkuXLgwcOJC+ffsSHR3tNV7ifPR6PZ9++ik2m41u3boxYcIEXnzxRa9lbrzxRh566CGmTp1Kp06d2Lx5M08//bTXMjfffDODBg2iX79+NGjQoMxLRqxWK19//TUZGRlcccUV3HLLLVxzzTXlntKqjNzcXDp37uz1GjJkiOdykYiICK666ir69+9Ps2bNWLFiBQAGg4H09HTGjBlDq1atGDFiBIMHD+bZZ58FtMCdMmUKbdq0YdCgQbRu3Zq///3v1a63PDpVVue2D+h0Oq/BN5s3b6ZXr16kpqYSExPjWW7ixIkcOXKEVatWlbkd90lvt+zsbOLi4sjKyqr0SeqStqdkcPOCLcRHWtn4aL8qb0eIuqqwsJDk5GQSEhKwWCy+Lkf4qXN9zrKzswkLCztvHtTayzXcw43T0tK8gvHEiRPnbO6bzeYLcocHaTEKIUTdUGu7UhMSEoiOjva6+4Pdbmfjxo307Nnzotcj5xiFEKJu8GmLMTc3l4MHD3p+dp80rlevHk2aNGHatGnMmTOHli1b0rJlS+bMmYPVauWOO+646LW6W4zyPEYhhPBvPg3GH3/80XMXCYDp06cDMHbsWJYsWcKMGTMoKChg8uTJngv8v/nmG0JCQi56rZ5bwjlc57wOSAghxKXNp8HYt2/fMi9sddPpdMyePZvZs2dfvKLK4X4eo1Jgd7o8LUghROXUkvF+wk/VxOer1p5jrG3MxjOHyuaQAThCVJb7Tir5+fk+rkT4M/fn6+w791RGrR2VWtsEGPTodFqLsbDISail6gddiLrIYDAQHh7OiRMnAO16OjklIWqKUor8/HxOnDhBeHi4131eK0uCsYJ0Oh1mo57CIpdcsiFEFbkvw3KHoxA1LTw8/JxPF6kICcZKsJgMWjDKJRtCVIlOpyMmJoaoqKhqPfRWiLKYTKZqtRTdJBgrwX2esVBajEJUi8FgqJFfYEJcCDL4phLcI1OlxSiEEP5LgrES3C1GOccohBD+S4KxEtwtRrktnBBC+C8JxkqQFqMQQvg/CcZKkBajEEL4PwnGSpAWoxBC+D8Jxkowm+QJG0II4e8kGCvB02KUe6UKIYTfkmCsBM85RulKFUIIvyXBWAkWowy+EUIIfyfBWAmehxVLi1EIIfyWBGMlSItRCCH8nwRjJUiLUQgh/J8EYyVY3E/XkBajEEL4LQnGSnBfxygtRiGE8F8SjJVgcXelSotRCCH8lgRjJZiN0mIUQgh/J8FYCe4Wo5xjFEII/yXBWAnSYhRCCP8nwVgJ0mIUQgj/V+uDMScnh2nTphEfH09gYCA9e/YkKSnJJ7VIi1EIIfxfrQ/GCRMmsHr1at599112797NgAED6N+/P8eOHbvotUiLUQgh/F+tDsaCggI+/vhjXnnlFa666ipatGjB7NmzSUhIYMGCBRe9HmkxCiGE/zP6uoBzcTgcOJ1OLBaL1/TAwEC+++67Mtex2WzYbDbPz9nZ2TVWj7lEi1EphU6nq7FtCyGEqB1qdYsxJCSEHj168Pzzz5OamorT6WTZsmVs3bqV48ePl7nO3LlzCQsL87zi4uJqrB738xiVgiKnqrHtCiGEqD1qdTACvPvuuyilaNSoEWazmddff5077rgDg8FQ5vKPP/44WVlZnteRI0dqrBaz8czhkvOMQgjhn2p1VypA8+bN2bhxI3l5eWRnZxMTE8Ntt91GQkJCmcubzWbMZvMFqSXAoEen01qMhUVOQi2mC7IfIYQQvlPrW4xuQUFBxMTEcPr0ab7++mtuvPHGi16DTqfztBplAI4QQvinWt9i/Prrr1FK0bp1aw4ePMijjz5K69atueuuu3xSj8VkoLDIJTcSF0IIP1XrW4xZWVlMmTKFxMRExowZQ+/evfnmm28wmXzTjeluMRZKi1EIIfxSrW8xjhgxghEjRvi6DA/3yFRpMQohhH+q9S3G2kbOMQohhH+TYKwkd4tRLtcQQgj/JMFYSdJiFEII/ybBWEnSYhRCCP8mwVhJ0mIUQgj/JsFYSWZ3i7FIWoxCCOGPJBgrydNidEiLUQgh/JEEYyV5zjFKV6oQQvglCcZKshjlAn8hhPBntf7ON7VC1lFI+wWs9TCbQgFpMQohhL+SFmNF/PYFvH8bbHnT02KUyzWEEMI/STBWhDVS+5qXjtkkl2sIIYQ/k2CsCHcw5qdjcT9dQ1qMQgjhlyQYKyKovvY1/5TnOkZpMQohhH+SYKwIqzsYM7AUD1eSUalCCOGfJBgrwlpP+6qcBKs8QFqMQgjhryQYK8JoBrN2mUaIMxOQc4xCCOGvqhSMR44c4ejRo56ft23bxrRp01i4cGGNFVbrFLcarY4sQFqMQgjhr6oUjHfccQfr168HIC0tjWuvvZZt27bxxBNP8Nxzz9VogbVG8XlGqyMTkBajEEL4qyoF4y+//EK3bt0A+M9//sNll13G5s2bee+991iyZElN1ld7FI9MDSw6DUiLUQgh/FWVgrGoqAiz2QzAmjVruOGGGwBITEzk+PHjNVddbVJ8LaPFnglIi1EIIfxVlYKxXbt2vP3222zatInVq1czaNAgAFJTU4mMjKzRAmuN4mAMsGUA0mIUQgh/VaVgfPnll/nHP/5B3759GTlyJB07dgTgs88+83Sx+p3irlRTcTAWOpwopXxZkRBCiAugSk/X6Nu3L6dOnSI7O5uIiAjP9HvuuQer1VpjxdUqxS1Go007x6gUFDkVAUadL6sSQghRw6rUYiwoKMBms3lCMSUlhfnz57Nv3z6ioqJqrDiHw8FTTz1FQkICgYGBNGvWjOeeew6XywfdmMWjUvUF6Z5Jcp5RCCH8T5VajDfeeCPDhw9n0qRJZGZm0r17d0wmE6dOnWLevHncd999NVLcyy+/zNtvv83SpUtp164dP/74I3fddRdhYWE8+OCDNbKPCivuStXln0Kn01qMtiIXWC5uGUIIIS6sKrUYd+zYwZVXXgnARx99RMOGDUlJSeGdd97h9ddfr7HitmzZwo033sjQoUNp2rQpt9xyCwMGDODHH3+ssX1UWPEF/rr8DMzuJ2wUSYtRCCH8TZWCMT8/n5CQEAC++eYbhg8fjl6v5y9/+QspKSk1Vlzv3r1Zu3Yt+/fvB2DXrl189913DBkypNx1bDYb2dnZXq8a4b6ReFE+4UaHti/pShVCCL9TpWBs0aIFK1eu5MiRI3z99dcMGDAAgBMnThAaGlpjxc2cOZORI0eSmJiIyWSic+fOTJs2jZEjR5a7zty5cwkLC/O84uLiaqYYcwgYAgBoaMwBoFAu2RBCCL9TpWB85plneOSRR2jatCndunWjR48egNZ67Ny5c40Vt2LFCpYtW8Z7773Hjh07WLp0KX/7299YunRpues8/vjjZGVleV5HjhypmWJ0Os/I1ChD8RM2pMUohBB+p0qDb2655RZ69+7N8ePHPdcwAlxzzTXcdNNNNVbco48+ymOPPcbtt98OQPv27UlJSWHu3LmMHTu2zHXMZrPnrjw1zlofco7TyJwPQHqu/cLsRwghhM9UKRgBoqOjiY6O5ujRo+h0Oho1alTjF/fn5+ej13s3ag0Gg28u1wDPAJx4SwEAqZkFvqlDCCHEBVOlrlSXy8Vzzz1HWFgY8fHxNGnShPDwcJ5//vkaDa3rr7+eF198kS+++IJDhw7x6aefMm/evBptlVZK8SUbjUxaV2pqVqFv6hBCCHHBVKnF+OSTT/Lvf/+bl156iV69eqGU4vvvv2f27NkUFhby4osv1khxb7zxBk8//TSTJ0/mxIkTxMbGcu+99/LMM8/UyPYrrXhkapQxF4Bj0mIUQgi/U6VgXLp0Kf/61788T9UA6NixI40aNWLy5Mk1FowhISHMnz+f+fPn18j2qq148E2kThuVKl2pQgjhf6rUlZqRkUFiYmKp6YmJiWRkZFS7qForSAvGUJd2baQEoxBC+J8qBWPHjh158803S01/88036dChQ7WLqrWKu1KtDu1G4idybNgdci2jEEL4kyp1pb7yyisMHTqUNWvW0KNHD3Q6HZs3b+bIkSN8+eWXNV1j7eF+wkbhaQKMeuwOF39mFxJXz0+fKCKEEHVQlVqMffr0Yf/+/dx0001kZmaSkZHB8OHD+fXXX1m8eHFN11h7lLiReKPwQEAG4AghhL+p8nWMsbGxpQbZ7Nq1i6VLl7Jo0aJqF1Yrue+XWnCaRrEmkk/B8SwJRiGE8CdVajHWWYFnHsrcIqQIgNRMuZZRCCH8iQRjZRiMnnBMsGotRelKFUII/yLBWFnF3alxxfdLlUs2hBDCv1TqHOPw4cPPOT8zM7M6tVwarJGQfoAYUx5gkWAUQgg/U6lgDAsLO+/8MWPGVKugWq94ZGp9fQ4QybHTBSil0Ol0vq1LCCFEjahUMPr1pRgVVfyEjXCl3RYuz+4ku9BBWKDJl1UJIYSoIXKOsbKKzzGaCjOoFxQAyHlGIYTwJxKMlVXclUr+KWLCLIAEoxBC+BMJxsoqvi0c+enEFt/9Rp7LKIQQ/kOCsbLcd7/JS/fcFk5ajEII4T8kGCsryN1iPEVsuHSlCiGEv5FgrKySXalyjlEIIfyOBGNlubtSnXYaBzkBuV+qEEL4EwnGygqwgkl7/mLjAO22cGnZhTic8sBiIYTwBxKMVVHcnVpPZWIy6HC6FCdybD4uSgghRE2QYKyKBokA6FO3Ey3nGYUQwq9IMFZFwpXa1+RNxIRpl2zI46eEEMI/SDBWRcJV2teU74kL0+6RKgNwhBDCP0gwVkV0B7CEgS2bTqbDABzPkhajEEL4AwnGqtAbIL4XAJfZdwFyjlEIIfxFrQ/Gpk2botPpSr2mTJni28KKu1Pjs7cDcPBEri+rEUIIUUMq9TxGX0hKSsLpdHp+/uWXX7j22mu59dZbfVgV0FQbgBNxagcmnYND6fn8mV1Iw1CLb+sSQghRLbW+xdigQQOio6M9r88//5zmzZvTp08f3xYW1RaskeiK8rixwZ8AbPk93bc1CSGEqLZaH4wl2e12li1bxt13341OpytzGZvNRnZ2ttfrgtDroWlvAK4LOQDAD39IMAohxKXukgrGlStXkpmZybhx48pdZu7cuYSFhXlecXFxF66g4u7UDkU/A7BFglEIIS55l1Qw/vvf/2bw4MHExsaWu8zjjz9OVlaW53XkyJELV1CC1p0bkf4TFp2dlPR8GZ0qhBCXuEsmGFNSUlizZg0TJkw453Jms5nQ0FCv1wVTvyUEN0TntHFzg+OAdKcKIcSl7pIJxsWLFxMVFcXQoUN9XcoZOp3nso3BwQcBCUYhhLjUXRLB6HK5WLx4MWPHjsVorGVXmBSfZ2xv3wHIeUYhhLjUXRLBuGbNGg4fPszdd9/t61JKa9EfgNBTO4nRn+ZIRgFHT+f7uCghhBBVdUkE44ABA1BK0apVK1+XUlpYI2jcDR2Ku+r9AsAPf2T4uCghhBBVdUkEY63XbhgAg3Q/AHKhvxBCXMokGGtCmxsAiMvZSQMyZQCOEEJcwiQYa0J4HDS6HB2KIcYkjmUWcCRDzjMKIcSlSIKxprQdBsAtgdrTNlbv+dOHxQghhKgqCcaa0lbrTm1X9AuRZLHshxRcLuXjooQQQlSWBGNNiWgKsZ3R4+IG8w7+OJXH97+f8nVVQgghKkmCsSa1vRGA0SE/AfDOlhRfViOEEKIKJBhrUnEwNsv7iXpks3bvnzIIRwghLjESjDWpXjOI7oBOOZnWcBcuBcu3HvZ1VUIIISpBgrGmdRkDwM3Or9DhYkXSYQqLnD4uSgghREVJMNa0jiPBHEpQ7iFuDvmN0/lFfP7zcV9XJYQQooIkGGuaORg63wnAA8FrAVj0XbJcuiGEEJcICcYLodtEQEeT01u4LCCNPcez+WjHUV9XJYQQogIkGC+EegnQejAAL8dtAeCVVb+RXVjky6qEEEJUgATjhdJ9EgBtT3xJ+/qKU7l25q8+4OOihBBCnI8E44WScBVEtUVXlMf8ltpzGpduOcT+P3N8XJgQQohzkWC8UHQ66H4vAM33/4vRLew4XYpZ//0VpWQgjhBC1FYSjBdSh9sgqh3kneTZ0zNpbUxjyx/prNx5zNeVCSGEKIcE44VkCoQx/4Wothjy/uRj6xwSdMd58tNfOCBdqkIIUStJMF5owQ1gzGcQ1ZZg+yk+CXyRyKJU7l22nVybw9fVCSGEOIsE48VQIhwjXBm8bfk7h09mMeOjXXK+UQghahkJxosluAGM+hAsYbRTB3jQtJIvd6fx7++SfV2ZEEKIEiQYL6awxnDdawBMMayki24/L365l4+2y11xhBCitpBgvNguuxk63IYeF/8KWYhVFfDoR7v4YJs8nkoIIWoDCUZfGPJXCIujnj2VD6LfI0jl89gnu3n3hxRfVyaEEHVerQ/GY8eOMXr0aCIjI7FarXTq1Int27f7uqzqsYTBTf8AdLTPXEtS0ENMMazk5ZXbeH3tAXkShxBC+FCtDsbTp0/Tq1cvTCYTX331FXv27OHVV18lPDzc16VVX9NeMPJ9qN+KQGcOj5r+wybzNI6t+wcTliaRlS83HBdCCF/QqVp8vcBjjz3G999/z6ZNm6q8jezsbMLCwsjKyiI0NLQGq6shLif88glsfBnStZuMf+O8nDeC72fu6Ku5rFGYjwsUQgj/UNE8qNUtxs8++4yuXbty6623EhUVRefOnfnnP/95znVsNhvZ2dler1pNb4AOt8KUrXDtc7j0JgYYtrMo/0Hmv/0WS75PlmsdhRDiIqrVwfjHH3+wYMECWrZsyddff82kSZN44IEHeOedd8pdZ+7cuYSFhXlecXFxF7HiatAboNeD6Ceuw1k/kQa6LP5leBnbV08ycfEWTuXafF2hEELUCbW6KzUgIICuXbuyefNmz7QHHniApKQktmzZUuY6NpsNm+1MiGRnZxMXF1d7u1LLUlSIWv0Mum3/AGCHqwXPmB5m6k39GNguGp1O5+MChRDi0uMXXakxMTG0bdvWa1qbNm04fLj8a/7MZjOhoaFer0uOyYJuyCtw23KcAWF00R9kmeMRtr3/IpMXfs2+NLkBuRBCXCi1Ohh79erFvn37vKbt37+f+Ph4H1V0kbW5DsN9m3DFdiFcl8czpnd5I3Ukx98ayn8Wv0ZaepavKxRCCL9Tq7tSk5KS6NmzJ88++ywjRoxg27ZtTJw4kYULFzJq1KgKbaPWj0qtCIcddizFtn055j9/8kzOUCH8EnU9LQdPJaZZOx8WKIQQtV9F86BWByPA559/zuOPP86BAwdISEhg+vTpTJw4scLr+0UwlnTqIEc3LsHy6/vUd53yTM4wRRMYEU1geAyEN4ErxkOD1j4sVAghahe/Ccbq8rtgdHM62PfdxxRs+ScdCn5Er/P+Z1Q6PboOt0GfmVAvwUdFCiFE7SHBWMxvg7GEX/bt56tNP3Aw+Q8iyaavficDDNpt85TeiK7NDdBqIDS/GoKjfFytEEL4hgRjsboQjG4p6Xn8a1MyH+84Soui/Txs/JA+hp+9F4rpCC36Q/NrIK4bGEy+KVYIIS4yCcZidSkY3XJtDr74OZUVSUdwHvmRaw3b6aPfRXv9Ie8FA0K0oLRGQGAEWCMhvjckXAXGAJ/ULoQQF4oEY7G6GIwlHTyRw393pvLfnankZxznKv0urjL8zJX63UTqyrke0hIGrYdCi2sgJBqCGmgvSzjoa/UVPkIIUS4JxmJ1PRjdlFLsOprFl7uPs+63E/x+IpvLdIdoqksjTJdHi2A7l4flkJizGVPBqbI3ojdCUBQEN4DAelo3rN6ovWI7Qcc7IDTmor4vIYSoKAnGYhKMZTuSkc/6fSdYvedPtvyejqP4GZB6XPSxHGRM2C7a6VMId2VismWgK8w8/0Z1emg5ANrfCsENwRQIRgtY60FwtLQ2hRA+JcFYTILx/LLyizwh+d3BU2QVeD8LMtRi5IomwfSKVlxe307r4EIsRdngcmgvex7s/R8cLvv+tQDoTRAaC6GNtK7aACsEBEG9ZtB2mFxSIoS44CQYi0kwVo7Tpdh1NJNv959k6x8Z7DySSUGR02sZg15Hm5gQusbXo0t8BJfHRxAbZkGXfhB2vAOHNoE9HxwFUFQI+emgnOXssVijy7WA1Bsh5zjk/qmtZ8+Hojxw2CCmE3S8XRscpDdo835fBwe+BqdDuxQluCGENYKmV2otVSGEKCbBWEyCsXocThd7j+fwY0oG21NOsyPlNKlZhaWWqx9spl1sKJc1CqVdbBjtYkNpUs+qPQnE6YDcNMg6CtnHwJartTLtuZDyPSR/C8pV8aJCYiC6PSRv0sK3LDoDNOkBiUMgtguYg7UWakAIBIbLZSpC1EESjMUkGGteamYB21NOa0F5+DS/pmbjdJX+GIWYjbSJDaVtTCito0O0V8MQgsxG7wVz/oQ9K7XWn8mqBV9ItHb5SECQNg1g/1fwyydQ8nxnWBNoc502ajb3hNbSPPkbnNhz7jdhCdO2HxytdefWS9C+BjfUplvrQVE+HEmCo0lwfJfWFZxwlfaq1wzk8V9CXFIkGItJMF54BXYne9Oy+TU1m1+PZfFrajb70nKwO8tuBUaHWkioH0RCgyCa1Q/yBGaDEPP5nzXpsMH+r+F0MiT00a7DLGud04dg3yotTDMPF7dQi1upNcEaqQW23qB1/xoCwGjWBhuZrNr9aiObawEa3kQLYGuk9wAkh137WpFrRgtOa61dg/H8ywohyiTBWEyC0TeKnC4O/JnLr6lZ7EvLYd+fOfyWlsPJHFu564RbTTSNDCI+0kp8PStNSnxfodCsCJcTCjK185f5pyA7FTL+KH4la9Py07Vl9EaI6QCNr4DYzlrYJn8LR7aBq+g8OyqDzqC1RJ127fyoexumIO0GC4ER2qUwwQ2186XKBX/+qr1y/9SCsWlvaNYHotrAyf2Q9rM23xysdR03+Yt2LtZp195DYfH7CInWtmswaed9s49B1hFwFmkt9LBG2nWq0goWfkyCsZgEY+2SmW8n+VSe53XwRC770nI4lJ5HGb2xHoEmA/GRVppGBtG0vhaYUSFmGrhfwWaMhhq8HMTl1IKprHOR9nxIP1g8KtepBZzTrrVmHTawZWshm/EHZPyuhW/eKaAW/FezhEFhOc/xNAVBUKR2jaq1HgQEAwqU0o6Fy6kNonI5tQC1RoK1vraOtb72c1B9MIdA3knISdPeO2ijkcMaaeFclA/5GVor2OU403Ud1EBbTm/wrsuWo/0xolzF+2ignSdWLu38tcuh/Ww0X8ADJ/yBBGMxCcZLQ2GRk99P5pKSns/hjPzir3mkpOeTmllwztAEbaRsTJiFuAgrcfUCaRRupVFEILHhFmLCAmkYasYa4MNuSGeRFo75p7Tu1oAg7aVcWsuu4LQWFnkntYFKuSe0X/hRbbWBRg1aQ/rv8McGSN6oBW+D1tq8hpdBQQYc/kG7ZCbzMKDTwsISrm0nJ827lWuyQlhjMJghJ1VrJdcGJqv2fmI6giUUDn0Hx7Zr7+Fc9CZo2FYbaNWwndZlnntSO54orVUcGqsFs9Ou/XFQmKUt53IWh79TuxZXp9da2Tpd8R8+xX8QmEOKz3/HaAFdlK/9EWTL0f4tI1toXecBQdo6uSe0ljk6qN9C+6NE+JQEYzEJxkuf3eHi6GktLA+l53HoVB5HThdwMsfGyRwbp3JtnhsUnEuw2UhUqJnoUAvRYRZiwiw0DLV4Wp5RIRaiQs2YjYbzbqtWs+eBMdD7fKbLVRy+p4pbd/W8u02LCrTWXX56cWsuQ9uOTncmLHQGrTWnM2hBkZ9+Juzz0s90T9tytH2EFocIaAGRnap1CZuCtP0HRmgB5Fn/lBZaZQmP14IyL10LO6+ubB21ojXuZo3U/tg5+xKl4Gio3xLMoVpPhCFAO57u64FdDu3fwX0u3GE7c2cpg0n7o8ESpr3MIcV3njJpX3Vn9Za41zEEnAl53P+W7n93ndbKDqqv/cEQ1AAKsyH7qDaCvCDzzD2Ug+pr21Ku4t4Dx5nz9kX52nbC4yGiqbYOaJ+Dwswzf/TlpxefKw8609tgCdPOm+uN2ntxFJ55/067Nj+w+LMSYK32P40EYzEJRv/ncilO5Ng4cjqfIxn5HD1dwLHTBaRmaV/TsgvJt5/nOsoSGoSYiQ0PJDrUTFigyfNqEGImOixQC9QQCyEWI3q9nJOrMS6n1kV9/Gc4vlP7xdzkL9oo4Ij4M8sppQWI3qD9MtXptPOlx3ZA6g44dUALH/f5WqW0a2OzU7VQNZqLwyVU+yXtDnudHlBnWomoM4OrdHqtnpy04hZ2htbVbA7RXoVZWu0FGWfq1Bm0PwxcxZcr1RXGQHDaKncJVkWYrPD4sWrdQUuCsZgEowDtiSMnsgv5M9vGn9mFHM8q5HhWASeybZzIKeREjo0TOTbsjor/Z9brICzQRLg1gAiriXpBAURYA4gICiDcatK+t2rzI4O06RHWAAwSpv4rP0NrHVsjtVB2ny8tzNYCO+N3rUXkLD4v7XKUuOewQfvl7+5mN1q0+U6HtmxRvtYCKyzuvnUVFZ9jLdLC30MVr1ekvdzz3a09OLO8o6C4y/mE1qIzBWld7GGNtJaau5chL13bjs6gBZPeqC0bYNVqtudBZorWI1CSIeBMq9Maqf1BUlRQ3LuQrr0XV4k6jRbtD46AIG3dwiztjw2XQ9vOzEPV+ueRYCwmwSgqSilFRp6d1MxCjmUWcDKnkKyCIs/rRI6NtKxC0rILycyvwqhUtMZNqMVEuNVEeKCJiKAAGgSbiQo1Uz9Ya6EGm40EW4yEmE2EBhoJsZgIsRgx1eTgIiHO5nJV/37GRQVa69zd7WsKrH5dSmldq7Yc7TxxNUgwFpNgFBeCzeEkK7+IzIIiTufZOZ1fxOl8u/Yq/jkz305Gnp3M/CLS8+yl7kFbWUEBBq1b1xpAWKCRUIuJ0EBT8dczARpq0b73BGxxyFpM+pq55EWIS1RF80CuFhaiCsxGA1GhBqJCLRVex+F0cTq/iKwCLSy1wLR5BhGdzLWRU+gg1+Ygt9BBTqGDnMIi8orPj+bZneTZnWXekq8ijHodwRajJ0jD3KFaIlitAQbMRj1mkwGLyUCgyYA1wEBggIEQs7ZMaKCRQJNBQlb4LQlGIS4So0Hvue6yMhxOFzmFDk+Xbmbx1+yCIrILte9zSgRpTqEWrLk2B9mFReTZHLgUOFzKE8jVZdDrCAowEGQ2aq8AA9YAI0Fm7as7TIMCtPmhxS3cEIsRa4AWrIHFywQWB7DZqJfBTKJWkGAUopYzGvTawJ2gCtw6rgxKKfLtTi0oC4rILtS+uoM2p/DMtMIiJzaHi8IiJwVFTgqKXBTYHZ71cwodOF0Kp0tp6xSe5/rCSirZQtW+GrEWT7MUt2YtxSEaFGDEatbC1xpg8HQdWwOMmI16jAYdJoPea9kAg3Qni/OTYBTCz+l0Ok/LrmElun7L4g7ZnEIHeXYH+TYtMPPtDvLsTvJtWku1wO4kv8hJgVcgF5Fd4PCEbr5d+1pyJHBB8Tzyqvuuy2bQ67AY9QQY9ZiNBswmvaf1ag0wYDFqXcjaS6+Fs0kLaItJW0db98x67qAOMOox6s+EsXsfAUa9jES+xEgwCiEqrGTI1hSnS2FzFAelXfuaZ3d4vs8v/v5Ma9ZFoUNbNs+mBXSeTfs+t/jnIofC4XJhd7godLg84et0Kc+5Wqh+l3JFmQw6LEZD8blbLVTd53DNJr3ne5NBj8mg87R23dMDTQYMep12vwW0gA8MMBJs1rqzzUZtvkGv84Sz0aAjwKAFtLvLWkY2V0ytDsbZs2fz7LPPek1r2LAhaWl16GJZIfycQa8rPi954X4dOZwu8ouc5Nuc2BxaK9Wry7i49VpY/H2hw+UJ4/zigLY5tGC2F69X6HBRaHeSX+TA7nBR5FQUObX5dqfL69JCbZ6DHFvNdj1Xlsmg87Ri3aF5pnWrx1gczAa9HpNeh9l0ZjkttLXWr6nkup4wL/6qPxPKJoMek1GPpXhAV8mucLNRm2fQ6dDrtNA3Foe7r7u7a3UwArRr1441a9Z4fjYYLvHbdQkhLjqjQU+oQU+o5eI8oFophcOlvALY/dUdxrYi15kwLnJSWOSiyOnC4dRC1u50L6+d53UqbbsK7W5PJbuu7Q4XTqVwOLXzv+7WssOlKCxyeu417A5oyn/Ijc/pdGjBa9B7uq8tJj3BFhP/ndLrotRQ64PRaDQSHR3t6zKEEKLCdDodpuLu0CAfP/RDKS1k3V3T7hatFtpnWsE2hwuHUwtVh7v16zwzTwttLfA9LePibbhby+557vWLXApbkRO704Wt6Mz+bEWucp/XqhSebeeWCPDgGuy+P59aH4wHDhwgNjYWs9lM9+7dmTNnDs2aNSt3eZvNhs125mhmZ2dfjDKFEKJW0um07lOz0UB49e/DXWNcLuUJR5dSntHO9uIWs82rle3CdRHvRVOrg7F79+688847tGrVij///JMXXniBnj178uuvvxIZGVnmOnPnzi11XlIIIUTtotfrsJz97M1a4pK6JVxeXh7NmzdnxowZTJ8+vcxlymoxxsXFyS3hhBCijvPLW8IFBQXRvn17Dhw4UO4yZrMZs1me5C2EEKJqLqmLWmw2G3v37iUmJsbXpQghhPBTtToYH3nkETZu3EhycjJbt27llltuITs7m7Fjx/q6NCGEEH6qVnelHj16lJEjR3Lq1CkaNGjAX/7yF3744Qfi4+PPv7IQQghRBbU6GD/44ANflyCEEKKOqdVdqUIIIcTFVqtbjDXBfTWKXOgvhBB1mzsHzneVot8HY05ODgBxcXE+rkQIIURtkJOTQ1hYWLnzL6kL/KvC5XKRmppKSEhIte7Y7r5RwJEjR+RGAWeRY1M+OTblk2NTPjk25avOsVFKkZOTQ2xsLHp9+WcS/b7FqNfrady4cY1tLzQ0VD6o5ZBjUz45NuWTY1M+OTblq+qxOVdL0U0G3wghhBAlSDAKIYQQJUgwVpDZbGbWrFlyH9YyyLEpnxyb8smxKZ8cm/JdjGPj94NvhBBCiMqQFqMQQghRggSjEEIIUYIEoxBCCFGCBKMQQghRggRjBfz9738nISEBi8XC5ZdfzqZNm3xd0kU3d+5crrjiCkJCQoiKimLYsGHs27fPaxmlFLNnzyY2NpbAwED69u3Lr7/+6qOKfWfu3LnodDqmTZvmmVaXj82xY8cYPXo0kZGRWK1WOnXqxPbt2z3z6+qxcTgcPPXUUyQkJBAYGEizZs147rnncLlcnmXq0rH59ttvuf7664mNjUWn07Fy5Uqv+RU5Fjabjfvvv5/69esTFBTEDTfcwNGjRytfjBLn9MEHHyiTyaT++c9/qj179qgHH3xQBQUFqZSUFF+XdlENHDhQLV68WP3yyy9q586daujQoapJkyYqNzfXs8xLL72kQkJC1Mcff6x2796tbrvtNhUTE6Oys7N9WPnFtW3bNtW0aVPVoUMH9eCDD3qm19Vjk5GRoeLj49W4cePU1q1bVXJyslqzZo06ePCgZ5m6emxeeOEFFRkZqT7//HOVnJysPvzwQxUcHKzmz5/vWaYuHZsvv/xSPfnkk+rjjz9WgPr000+95lfkWEyaNEk1atRIrV69Wu3YsUP169dPdezYUTkcjkrVIsF4Ht26dVOTJk3ympaYmKgee+wxH1VUO5w4cUIBauPGjUoppVwul4qOjlYvvfSSZ5nCwkIVFham3n77bV+VeVHl5OSoli1bqtWrV6s+ffp4grEuH5uZM2eq3r17lzu/Lh+boUOHqrvvvttr2vDhw9Xo0aOVUnX72JwdjBU5FpmZmcpkMqkPPvjAs8yxY8eUXq9Xq1atqtT+pSv1HOx2O9u3b2fAgAFe0wcMGMDmzZt9VFXtkJWVBUC9evUASE5OJi0tzetYmc1m+vTpU2eO1ZQpUxg6dCj9+/f3ml6Xj81nn31G165dufXWW4mKiqJz587885//9Myvy8emd+/erF27lv379wOwa9cuvvvuO4YMGQLU7WNztooci+3bt1NUVOS1TGxsLJdddlmlj5ff30S8Ok6dOoXT6aRhw4Ze0xs2bEhaWpqPqvI9pRTTp0+nd+/eXHbZZQCe41HWsUpJSbnoNV5sH3zwATt27CApKanUvLp8bP744w8WLFjA9OnTeeKJJ9i2bRsPPPAAZrOZMWPG1OljM3PmTLKyskhMTMRgMOB0OnnxxRcZOXIkULc/N2eryLFIS0sjICCAiIiIUstU9ve1BGMFnP24KqVUtR5hdambOnUqP//8M999912peXXxWB05coQHH3yQb775BovFUu5ydfHYuFwuunbtypw5cwDo3Lkzv/76KwsWLGDMmDGe5erisVmxYgXLli3jvffeo127duzcuZNp06YRGxvL2LFjPcvVxWNTnqoci6ocL+lKPYf69etjMBhK/bVx4sSJUn+51BX3338/n332GevXr/d6nFd0dDRAnTxW27dv58SJE1x++eUYjUaMRiMbN27k9ddfx2g0et5/XTw2MTExtG3b1mtamzZtOHz4MFC3PzePPvoojz32GLfffjvt27fnzjvv5KGHHmLu3LlA3T42Z6vIsYiOjsZut3P69Olyl6koCcZzCAgI4PLLL2f16tVe01evXk3Pnj19VJVvKKWYOnUqn3zyCevWrSMhIcFrfkJCAtHR0V7Hym63s3HjRr8/Vtdccw27d+9m586dnlfXrl0ZNWoUO3fupFmzZnX22PTq1avUZT379+8nPj4eqNufm/z8/FIPyzUYDJ7LNerysTlbRY7F5Zdfjslk8lrm+PHj/PLLL5U/XlUaMlSHuC/X+Pe//6327Nmjpk2bpoKCgtShQ4d8XdpFdd9996mwsDC1YcMGdfz4cc8rPz/fs8xLL72kwsLC1CeffKJ2796tRo4c6bdDy8+n5KhUperusdm2bZsyGo3qxRdfVAcOHFDLly9XVqtVLVu2zLNMXT02Y8eOVY0aNfJcrvHJJ5+o+vXrqxkzZniWqUvHJicnR/3000/qp59+UoCaN2+e+umnnzyXxlXkWEyaNEk1btxYrVmzRu3YsUNdffXVcrnGhfLWW2+p+Ph4FRAQoLp06eK5RKEuAcp8LV682LOMy+VSs2bNUtHR0cpsNqurrrpK7d6923dF+9DZwViXj83//vc/ddlllymz2awSExPVwoULvebX1WOTnZ2tHnzwQdWkSRNlsVhUs2bN1JNPPqlsNptnmbp0bNavX1/m75ixY8cqpSp2LAoKCtTUqVNVvXr1VGBgoLruuuvU4cOHK12LPHZKCCGEKEHOMQohhBAlSDAKIYQQJUgwCiGEECVIMAohhBAlSDAKIYQQJUgwCiGEECVIMAohhBAlSDAKIYQQJUgwCiHKpdPpWLlypa/LEOKikmAUopYaN24cOp2u1GvQoEG+Lk0IvybPYxSiFhs0aBCLFy/2mmY2m31UjRB1g7QYhajFzGYz0dHRXi/3E8p1Oh0LFixg8ODBBAYGkpCQwIcffui1/u7du7n66qsJDAwkMjKSe+65h9zcXK9lFi1aRLt27TCbzcTExDB16lSv+adOneKmm27CarXSsmVLPvvsswv7poXwMQlGIS5hTz/9NDfffDO7du1i9OjRjBw5kr179wLa8/4GDRpEREQESUlJfPjhh6xZs8Yr+BYsWMCUKVO455572L17N5999hktWrTw2sezzz7LiBEj+PnnnxkyZAijRo0iIyPjor5PIS6q6j8sRAhxIYwdO1YZDAYVFBTk9XruueeUUtqjwCZNmuS1Tvfu3dV9992nlFJq4cKFKiIiQuXm5nrmf/HFF0qv16u0tDSllFKxsbHqySefLLcGQD311FOen3Nzc5VOp1NfffVVjb1PIWobOccoRC3Wr18/FixY4DWtXr16nu979OjhNa9Hjx7s3LkTgL1799KxY0eCgoI883v16oXL5WLfvn3odDpSU1O55pprzllDhw4dPN8HBQUREhLCiRMnqvqWhKj1JBiFqMWCgoJKdW2ej06nA0Ap5fm+rGUCAwMrtD2TyVRqXZfLVamahLiUyDlGIS5hP/zwQ6mfExMTAWjbti07d+4kLy/PM//7779Hr9fTqlUrQkJCaNq0KWvXrr2oNQtR20mLUYhazGazkZaW5jXNaDRSv359AD788EO6du1K7969Wb58Odu2bePf//43AKNGjWLWrFmMHTuW2bNnc/LkSe6//37uvPNOGjZsCMDs2bOZNGkSUVFRDB48mJycHL7//nvuv//+i/tGhahFJBiFqMVWrVpFTEyM17TWrVvz22+/AdqI0Q8++IDJkycTHR3N8uXLadu2LQBWq5Wvv/6aBx98kCuuuAKr1crNN9/MvHnzPNsaO3YshYWFvPbaazzyyCPUr1+fW2655eK9QSFqIZ1SSvm6CCFE5el0Oj799FOGDRvm61KE8CtyjlEIIYQoQYJRCCGEKEHOMQpxiZKzIEJcGNJiFEIIIUqQYBRCCCFKkGAUQgghSpBgFEIIIUqQYBRCCCFKkGAUQgghSpBgFEIIIUqQYBRCCCFK+H9TC6p6YEQDFQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 500x300 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data = pd.read_csv('./experiments/log/log.txt', sep='\\t')\n",
    "plt.figure(figsize=(5, 3))\n",
    "plt.plot(data['Epoch'], data['Train'], label='Train Loss')\n",
    "plt.plot(data['Epoch'], data['Validation'], label='Validation Loss')\n",
    "plt.xlabel('Epoch')\n",
    "plt.ylabel('Loss')\n",
    "plt.title('Training and Validation Loss')\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b50223d-6bf0-40e1-88b8-5adf31ac9717",
   "metadata": {},
   "source": [
    "### Additional metrics (plots below)\n",
    "- The plots below are produced for a specific temperature. Temperature effects were not investigated in this experiment, but it should be noted that higher temperatures introduce more randomness into the model as higher temperature is a key driver of entropy in folding. We used a low temperature for predictions in order to increase accuracy.\n",
    "#### A) Log Probability Difference (Diff):\n",
    "- This plot displays the difference in log probabilities (`Diff`) between the sampled sequences and the native sequences for the temperature used during the sampling process.\n",
    "- Each point in the plot represents a protein's sampled sequence at a specific temperature, with the vertical axis showing the difference in log probabilities (`log P(sample) - log P(native)`).\n",
    "- The y-axis (log probability difference) measures how close the model's predictions are to considering the sampled sequence as probable as the native sequence. A value close to zero indicates that the sampled sequence is nearly as probable as the native sequence according to the model.\n",
    "- We found a median of 0.47 for the (`Diff`) which suggests that, while the native sample is considered more probable than the predicted sequences by the model, it is still reasonably probable under the model's current parameters.\n",
    "\n",
    "#### B) Decoding Temperature vs. Native Sequence Recovery:\n",
    "- This plot displays the \"similarity\" between the sampled sequences and the native sequences, measured as the fraction of amino acids that match the native sequence (`Native sequence recovery`).\n",
    "- The y-axis shows the similarity as a fraction, ranging from 0 (no similarity) to 1 (perfect match), illustrating how well the model can recreate the native sequence under different sampling conditions.\n",
    "- Higher values indicate better recovery of the native sequence, suggesting that the model is effective at predicting the correct amino acids given a protein structure.\n",
    "- We found a median of 0.41 for the recovery ratio, indicating that the model correctly predicts 41% of the amino acids in the native sequence. This is reasonable and expected as, on average, only 20-50% of protein seqeunces are highly conserved, allowing for significant variability in the remaining 50-80% of amino acid positions in a protein sequence.\n",
    "\n",
    "### A) Decoding Temperature vs, Log Probability Difference. &emsp;&emsp; B) Decoding Temp vs. Native Sequence Recovery\n",
    "![Model predictions for protein 2NWM_1_A](./images/decoding_recovery.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "876b2337-ad76-430f-b1c2-1a7e47abfce4",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: './experiments/test_log/results.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[17], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m data \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m./experiments/test_log/results.csv\u001b[39m\u001b[38;5;124m'\u001b[39m)  \n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# Set up the plots\u001b[39;00m\n\u001b[1;32m      4\u001b[0m fig, axes \u001b[38;5;241m=\u001b[39m plt\u001b[38;5;241m.\u001b[39msubplots(nrows\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m, ncols\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m, figsize\u001b[38;5;241m=\u001b[39m(\u001b[38;5;241m20\u001b[39m, \u001b[38;5;241m5\u001b[39m))\n",
      "File \u001b[0;32m/opt/homebrew/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:948\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m    935\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    936\u001b[0m     dialect,\n\u001b[1;32m    937\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    944\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m    945\u001b[0m )\n\u001b[1;32m    946\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 948\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m _read(filepath_or_buffer, kwds)\n",
      "File \u001b[0;32m/opt/homebrew/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:611\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    608\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    610\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 611\u001b[0m parser \u001b[38;5;241m=\u001b[39m TextFileReader(filepath_or_buffer, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n\u001b[1;32m    613\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    614\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m/opt/homebrew/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1448\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1445\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1447\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1448\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_engine(f, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine)\n",
      "File \u001b[0;32m/opt/homebrew/anaconda3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1705\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1703\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1704\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1705\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m get_handle(\n\u001b[1;32m   1706\u001b[0m     f,\n\u001b[1;32m   1707\u001b[0m     mode,\n\u001b[1;32m   1708\u001b[0m     encoding\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mencoding\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m),\n\u001b[1;32m   1709\u001b[0m     compression\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcompression\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m),\n\u001b[1;32m   1710\u001b[0m     memory_map\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmemory_map\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mFalse\u001b[39;00m),\n\u001b[1;32m   1711\u001b[0m     is_text\u001b[38;5;241m=\u001b[39mis_text,\n\u001b[1;32m   1712\u001b[0m     errors\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mencoding_errors\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstrict\u001b[39m\u001b[38;5;124m\"\u001b[39m),\n\u001b[1;32m   1713\u001b[0m     storage_options\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstorage_options\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m),\n\u001b[1;32m   1714\u001b[0m )\n\u001b[1;32m   1715\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1716\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m/opt/homebrew/anaconda3/lib/python3.11/site-packages/pandas/io/common.py:863\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    858\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    859\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    860\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    861\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    862\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 863\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(\n\u001b[1;32m    864\u001b[0m             handle,\n\u001b[1;32m    865\u001b[0m             ioargs\u001b[38;5;241m.\u001b[39mmode,\n\u001b[1;32m    866\u001b[0m             encoding\u001b[38;5;241m=\u001b[39mioargs\u001b[38;5;241m.\u001b[39mencoding,\n\u001b[1;32m    867\u001b[0m             errors\u001b[38;5;241m=\u001b[39merrors,\n\u001b[1;32m    868\u001b[0m             newline\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    869\u001b[0m         )\n\u001b[1;32m    870\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    871\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    872\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: './experiments/test_log/results.csv'"
     ]
    }
   ],
   "source": [
    "data = pd.read_csv('./experiments/test_log/results.csv')  \n",
    "\n",
    "# Set up the plots\n",
    "fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(20, 5))\n",
    "columns = ['score', 'similarity', 'native', 'diff']\n",
    "titles = {'score': \"i) Distribution of log probability \\n of sampled sequence\", \n",
    "          'similarity': \"ii) Distibution of similarity of sampled \\n sequence to native sequence\" , \n",
    "          'native': \"iii) Distribution of log probability of native \\n sequence under the model\", \n",
    "          'diff': \"iv) Distribution of the difference in log probabilities \\n between the sampled and native sequences\",}\n",
    "# Plot each score in its own subplot\n",
    "for ax, column in zip(axes, columns):\n",
    "    data[column].plot(kind='hist', bins=20, alpha=0.75, ax=ax)\n",
    "    ax.set_title(titles[column])\n",
    "    ax.axvline(data[column].median(), color='r', linestyle='dashed', linewidth=1)\n",
    "    ax.set_xlabel(column)\n",
    "    ax.set_ylabel('Frequency')\n",
    "    # Annotate the median value\n",
    "    median_val = data[column].median()\n",
    "    ax.annotate(f'Median: {median_val:.2f}', xy=(median_val, ax.get_ylim()[1]*0.9), xytext=(median_val, ax.get_ylim()[1]*0.9),\n",
    "                arrowprops=dict(facecolor='black', shrink=0.05), horizontalalignment='left')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffdd9068-2e8f-48cb-b552-b4173a5c37ee",
   "metadata": {},
   "source": [
    "### More Distributions (above) n=7161\n",
    "- i) Score: The log probability of the sampled sequence under the model, given the protein structure. A higher score indicates that the model finds the sequence more likely.\n",
    "- ii) Similarity: This measures how similar the sampled sequence is to the native (actual) sequence. It's calculated as the number of matching amino acids divided by the total length of the sequence. Higher values indicate closer resemblance to the true sequence.\n",
    "- iii) Native: This represents the log probability of the native sequence under the model, suggesting how likely the model thinks the true sequence is given the protein's structure.\n",
    "- iv) Diff: This represents the difference in log probabilities between the sampled sequence and the native sequence. A smaller value (closer to zero) indicates that the sampled sequence is nearly as probable as the native sequence according to the model."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d348141-3843-473d-841d-3b85247e28a2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Visualizing sequence predictions for a given protein:\n",
    "The plot below shows the log probabilities of each of the 20 amino acids appearing at each position in the protein sequence. You can see that certain positions are highly conserved, that is there is only one amino acid that is likely appear in that position (yellow pixels). However, there are also regions with no strong prediction (all blue, or several light blue pixels for a sequence position on the x-axis). These regions have significant entropy in the amino acids sequence of a protein and are the source of variation in multiple sequence alignments. These logs are available for all proteins in the logs folder.\n",
    "\n",
    "![Model predictions for protein 2NWM_1_A](./images/train_2NWM_1_A_probs131248.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c7110cc-df1e-424d-b3c6-59d33529964d",
   "metadata": {},
   "source": [
    "### Visualizing protein structure\n",
    "[AlphaFold Server](https://alphafoldserver.com/about) was used to predict the structure of the novel multiple sequence alignments for 5 proteins. First we need to export the data for some proteins which have a predicted MSA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d7f00e8-6300-4480-a903-c01e3dbb9b98",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_list = data['name'].tolist()\n",
    "test_prots = [protein[:-3] for protein in prot_list]\n",
    "test_rows = df_features[df_features['Protein_Name'].isin(test_prots)]\n",
    "indexes = [1,578,1234,2777,3503] #randomly selected\n",
    "for i in indexes:\n",
    "    pdb_content = format_pdb_from_df(test_rows.iloc[i])\n",
    "    # Write PDB to a file\n",
    "    name = test_rows['Protein_Name'].iloc[i]\n",
    "    with open(f\"./data/{name}_test.pdb\", \"w\") as file:\n",
    "        file.write(pdb_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98b547d8-fb95-4a9c-bb90-f1596bb90f0b",
   "metadata": {},
   "source": [
    "The MSA were then uploaded to AlphaFold and the structures computed. Plots of 5 distinct proteins along with 3 MSAs for each protein are visualized below. We can see that for all proteins, the MSA have visually identical structures to the native proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f19d3f80-d8f5-499a-800a-0e3d0bc15b8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "view1 = nv.show_structure_file('./alphafold/fold_1gxr_msa1/fold_1gxr_msa1_model_0.cif')\n",
    "view1.clear_representations()\n",
    "view1.add_representation('cartoon', color='blue')\n",
    "view2 = nv.show_structure_file('./alphafold/fold_1gxr_msa2/fold_1gxr_msa2_model_0.cif')\n",
    "view2.clear_representations()\n",
    "view2.add_representation('cartoon', color='green')\n",
    "view3 = nv.show_structure_file('./alphafold/fold_1gxr_msa3/fold_1gxr_msa3_model_0.cif')\n",
    "view3.clear_representations()\n",
    "view3.add_representation('cartoon', color='red')\n",
    "view4 = nv.show_structure_file('./alphafold/fold_1gxr_msa4/fold_1gxr_msa4_model_0.cif')\n",
    "view4.clear_representations()\n",
    "view4.add_representation('cartoon', color='orange')\n",
    "app_layout1 = widgets.AppLayout(\n",
    "    center=widgets.HBox([view1, view2]),\n",
    "    header=None, footer=None,\n",
    "    left_sidebar=None, right_sidebar=None,\n",
    "    pane_widths=[1, 0, 1], pane_heights=[0, 1, 0])\n",
    "app_layout2 = widgets.AppLayout(\n",
    "    center=widgets.HBox([view3, view4]),\n",
    "    header=None, footer=None,\n",
    "    left_sidebar=None, right_sidebar=None,\n",
    "    pane_widths=[1, 0, 1], pane_heights=[0, 1, 0])\n",
    "print(\"Structures for 1GXR\")\n",
    "ipy_display(app_layout1)\n",
    "ipy_display(app_layout2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e103c5e9-e50e-4d30-bac4-b5ee8d741617",
   "metadata": {},
   "source": [
    "### Note: Keep in mind that AlphaFold structures are predictions, even for known models, so it is always good to check the actual structure with NGL to make sure it is not significantly different from AlphaFold. The native strucutre below is basically identical (don't be fooled by the lack of formatting for the chains)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42ed63de-1317-49f3-b231-2086db8623bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = PDBParser()\n",
    "structure = parser.get_structure('Sample', './data/1GXR_1_A_test.pdb')\n",
    "view = nv.show_biopython(structure)\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', selection='protein', color='blue')\n",
    "view.add_representation('ball+stick', selection='ligand')\n",
    "view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95e3a3e7-4556-4554-be03-3716153c2b56",
   "metadata": {},
   "outputs": [],
   "source": [
    "view1 = nv.show_structure_file('./alphafold/fold_2i6d_msa1/fold_2i6d_msa1_model_0.cif')\n",
    "view1.clear_representations()\n",
    "view1.add_representation('cartoon', color='blue')\n",
    "view2 = nv.show_structure_file('./alphafold/fold_2i6d_msa2/fold_2i6d_msa2_model_0.cif')\n",
    "view2.clear_representations()\n",
    "view2.add_representation('cartoon', color='green')\n",
    "view3 = nv.show_structure_file('./alphafold/fold_2i6d_msa3/fold_2i6d_msa3_model_0.cif')\n",
    "view3.clear_representations()\n",
    "view3.add_representation('cartoon', color='red')\n",
    "view4 = nv.show_structure_file('./alphafold/fold_2i6d_msa4/fold_2i6d_msa4_model_0.cif')\n",
    "view4.clear_representations()\n",
    "view4.add_representation('cartoon', color='orange')\n",
    "app_layout1 = widgets.AppLayout(\n",
    "    center=widgets.HBox([view1, view2]),\n",
    "    header=None, footer=None,\n",
    "    left_sidebar=None, right_sidebar=None,\n",
    "    pane_widths=[1, 0, 1], pane_heights=[0, 1, 0])\n",
    "app_layout2 = widgets.AppLayout(\n",
    "    center=widgets.HBox([view3, view4]),\n",
    "    header=None, footer=None,\n",
    "    left_sidebar=None, right_sidebar=None,\n",
    "    pane_widths=[1, 0, 1], pane_heights=[0, 1, 0])\n",
    "print(\"Structures for 2I6D\")\n",
    "ipy_display(app_layout1)\n",
    "ipy_display(app_layout2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6576ae54-d91b-4b2a-89be-701fe35ca185",
   "metadata": {},
   "outputs": [],
   "source": [
    "view1 = nv.show_structure_file('./alphafold/fold_3tpd_msa1/fold_3tpd_msa1_model_0.cif')\n",
    "view1.clear_representations()\n",
    "view1.add_representation('cartoon', color='blue')\n",
    "view2 = nv.show_structure_file('./alphafold/fold_3tpd_msa2/fold_3tpd_msa2_model_0.cif')\n",
    "view2.clear_representations()\n",
    "view2.add_representation('cartoon', color='green')\n",
    "view3 = nv.show_structure_file('./alphafold/fold_3tpd_msa3/fold_3tpd_msa3_model_0.cif')\n",
    "view3.clear_representations()\n",
    "view3.add_representation('cartoon', color='red')\n",
    "view4 = nv.show_structure_file('./alphafold/fold_3tpd_msa4/fold_3tpd_msa4_model_0.cif')\n",
    "view4.clear_representations()\n",
    "view4.add_representation('cartoon', color='orange')\n",
    "app_layout1 = widgets.AppLayout(\n",
    "    center=widgets.HBox([view1, view2]),\n",
    "    header=None, footer=None,\n",
    "    left_sidebar=None, right_sidebar=None,\n",
    "    pane_widths=[1, 0, 1], pane_heights=[0, 1, 0])\n",
    "app_layout2 = widgets.AppLayout(\n",
    "    center=widgets.HBox([view3, view4]),\n",
    "    header=None, footer=None,\n",
    "    left_sidebar=None, right_sidebar=None,\n",
    "    pane_widths=[1, 0, 1], pane_heights=[0, 1, 0])\n",
    "print(\"Structures for 3TPD - Notice the loop missing in the red structure\")\n",
    "ipy_display(app_layout1)\n",
    "ipy_display(app_layout2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a6481c5-ee3a-4089-b53f-d95339bf0466",
   "metadata": {},
   "outputs": [],
   "source": [
    "view1 = nv.show_structure_file('./alphafold/fold_4mu9_msa1/fold_4mu9_msa1_model_0.cif')\n",
    "view1.clear_representations()\n",
    "view1.add_representation('cartoon', color='blue')\n",
    "view2 = nv.show_structure_file('./alphafold/fold_4mu9_msa2/fold_4mu9_msa2_model_0.cif')\n",
    "view2.clear_representations()\n",
    "view2.add_representation('cartoon', color='green')\n",
    "view3 = nv.show_structure_file('./alphafold/fold_4mu9_msa3/fold_4mu9_msa3_model_0.cif')\n",
    "view3.clear_representations()\n",
    "view3.add_representation('cartoon', color='red')\n",
    "view4 = nv.show_structure_file('./alphafold/fold_4mu9_msa4/fold_4mu9_msa4_model_0.cif')\n",
    "view4.clear_representations()\n",
    "view4.add_representation('cartoon', color='orange')\n",
    "app_layout1 = widgets.AppLayout(\n",
    "    center=widgets.HBox([view1, view2]),\n",
    "    header=None, footer=None,\n",
    "    left_sidebar=None, right_sidebar=None,\n",
    "    pane_widths=[1, 0, 1], pane_heights=[0, 1, 0])\n",
    "app_layout2 = widgets.AppLayout(\n",
    "    center=widgets.HBox([view3, view4]),\n",
    "    header=None, footer=None,\n",
    "    left_sidebar=None, right_sidebar=None,\n",
    "    pane_widths=[1, 0, 1], pane_heights=[0, 1, 0])\n",
    "print(\"Structures for 4MU9|\")\n",
    "ipy_display(app_layout1)\n",
    "ipy_display(app_layout2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae581237-340e-4646-ad87-6b6d1d2cc461",
   "metadata": {},
   "outputs": [],
   "source": [
    "view1 = nv.show_structure_file('./alphafold/fold_1rff_msa1/fold_1rff_msa1_model_0.cif')\n",
    "view1.clear_representations()\n",
    "view1.add_representation('cartoon', color='blue')\n",
    "view2 = nv.show_structure_file('./alphafold/fold_1rff_msa2/fold_1rff_msa2_model_0.cif')\n",
    "view2.clear_representations()\n",
    "view2.add_representation('cartoon', color='green')\n",
    "view3 = nv.show_structure_file('./alphafold/fold_1rff_msa3/fold_1rff_msa3_model_0.cif')\n",
    "view3.clear_representations()\n",
    "view3.add_representation('cartoon', color='red')\n",
    "view4 = nv.show_structure_file('./alphafold/fold_1rff_msa3/fold_1rff_msa3_model_0.cif')\n",
    "view4.clear_representations()\n",
    "view4.add_representation('cartoon', color='orange')\n",
    "app_layout1 = widgets.AppLayout(\n",
    "    center=widgets.HBox([view1, view2]),\n",
    "    header=None, footer=None,\n",
    "    left_sidebar=None, right_sidebar=None,\n",
    "    pane_widths=[1, 0, 1], pane_heights=[0, 1, 0])\n",
    "app_layout2 = widgets.AppLayout(\n",
    "    center=widgets.HBox([view3, view4]),\n",
    "    header=None, footer=None,\n",
    "    left_sidebar=None, right_sidebar=None,\n",
    "    pane_widths=[1, 0, 1], pane_heights=[0, 1, 0])\n",
    "print(\"Structures for 1RFF\")\n",
    "ipy_display(app_layout1)\n",
    "ipy_display(app_layout2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe003307-05af-4f55-ae02-f8d20c8b9fd6",
   "metadata": {},
   "source": [
    "## Conclusions: \n",
    "- This model is effective for generating MSAs with similar structure to a given protein.\n",
    "- In the future it would be useful to use a metric to compute how close the predicted structures of each of the MSAs is to the known native protein structure."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbaded35-ef7f-4cd7-96e2-a6ae64297f56",
   "metadata": {},
   "source": [
    "# PART 2 - Novel sequence "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "038a9b98-7c03-48ca-84e2-b67d0a3749d1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebde7329-7f8b-4c94-87f0-d3f9fed8f724",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d108f35c-55b0-44ed-a0e5-b3771a749716",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e420bc6-0778-4307-8421-21599b28be73",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
