{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e550d507-e54f-46e3-b9ed-da4ff0006a02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install biopython\n",
    "# !pip install nglview\n",
    "# !conda install -c conda-forge nglview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e6f2f82-892d-4b85-9618-271da9e1024b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55d35c76-33c5-469e-8b0f-8f7b5064e8fc",
   "metadata": {},
   "source": [
    "### The CASP12 dataset is just a huge (12gb) ASCII text file with each protein's data in sequence. Open the `test` file if you want to see the structure as it is only a few mb. Here is a parser that loads it into an initial df."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2b52ff76-1806-419d-8ad8-042ae837ed80",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "def parse_protein_data(file_path, n):\n",
    "    # Initialize lists to store data\n",
    "    ids = []\n",
    "    sequences = []\n",
    "    evolutionary = []\n",
    "    tertiary = []\n",
    "    masks = []\n",
    "    \n",
    "    # Initialize counters and temporary storage for current protein data\n",
    "    current_protein = {}\n",
    "    count = 0\n",
    "    \n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in tqdm(file):\n",
    "            line = line.strip()\n",
    "            if line.startswith('[ID]'):\n",
    "                # Save previous protein data if exists\n",
    "                if current_protein:\n",
    "                    ids.append(current_protein.get('ID', ''))\n",
    "                    sequences.append(current_protein.get('PRIMARY', ''))\n",
    "                    evolutionary.append(current_protein.get('EVOLUTIONARY', []))\n",
    "                    tertiary.append(current_protein.get('TERTIARY', []))\n",
    "                    masks.append(current_protein.get('MASK', ''))\n",
    "                    count += 1\n",
    "                    # Break the loop if the required number of proteins have been parsed\n",
    "                    if count >= n:\n",
    "                        break\n",
    "                    current_protein = {}\n",
    "            \n",
    "            # Identify the section and append the data to the current protein\n",
    "            if line.startswith('['):\n",
    "                key = line[1:line.find(']')]\n",
    "                current_protein[key] = []\n",
    "            elif current_protein:\n",
    "                # Split TSV data into lists of floats or keep as strings depending on the section\n",
    "                if key in ['EVOLUTIONARY', 'TERTIARY']:\n",
    "                    # Split the line by spaces, convert each item to float\n",
    "                    current_protein[key].append([float(x) for x in line.split()])\n",
    "                else:\n",
    "                    current_protein[key].append(line)\n",
    "                \n",
    "    # Check if last processed protein needs to be added\n",
    "    if count < n and current_protein:\n",
    "        ids.append(current_protein.get('ID', ''))\n",
    "        sequences.append(current_protein.get('PRIMARY', ''))\n",
    "        evolutionary.append(current_protein.get('EVOLUTIONARY', []))\n",
    "        tertiary.append(current_protein.get('TERTIARY', []))\n",
    "        masks.append(current_protein.get('MASK', ''))\n",
    "    \n",
    "    # Create DataFrame\n",
    "    df = pd.DataFrame({\n",
    "        'ID': ids,\n",
    "        'Sequence': sequences,\n",
    "        'Evolutionary': evolutionary,\n",
    "        'Tertiary': tertiary,\n",
    "        'Mask': masks\n",
    "    })\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d658b9d9-e92e-4c53-a4fa-800342c5cdf9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1680162it [01:07, 24870.41it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Evolutionary</th>\n",
       "      <th>Tertiary</th>\n",
       "      <th>Mask</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[3LAF_1_A]</td>\n",
       "      <td>[DRWGSELESSHHHHHHGGRRSLHFVSEPSDAVTMRGGNVLLNCSA...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.10...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "      <td>[------------------+++++++++++++++++++++++++++...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[2D5F_d2d5fb1]</td>\n",
       "      <td>[NECQLNNLNALEPDHRVESEGGLIETWNSQHPELQCAGVTVSKRT...</td>\n",
       "      <td>[[0.0, 0.03975101015616468, 0.0, 0.00715892820...</td>\n",
       "      <td>[[472.3, 574.6, 714.3, 726.5, 853.6, 963.7, 92...</td>\n",
       "      <td>[+++++++++++++++++++++++++++++++++++++++++++++...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               ID                                           Sequence  \\\n",
       "0      [3LAF_1_A]  [DRWGSELESSHHHHHHGGRRSLHFVSEPSDAVTMRGGNVLLNCSA...   \n",
       "1  [2D5F_d2d5fb1]  [NECQLNNLNALEPDHRVESEGGLIETWNSQHPELQCAGVTVSKRT...   \n",
       "\n",
       "                                        Evolutionary  \\\n",
       "0  [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.10...   \n",
       "1  [[0.0, 0.03975101015616468, 0.0, 0.00715892820...   \n",
       "\n",
       "                                            Tertiary  \\\n",
       "0  [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...   \n",
       "1  [[472.3, 574.6, 714.3, 726.5, 853.6, 963.7, 92...   \n",
       "\n",
       "                                                Mask  \n",
       "0  [------------------+++++++++++++++++++++++++++...  \n",
       "1  [+++++++++++++++++++++++++++++++++++++++++++++...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Usage example\n",
    "file_path = './data/training_95'\n",
    "number_of_proteins = 90000  # Set how many proteins you want to load\n",
    "protein_df = parse_protein_data(file_path, number_of_proteins)\n",
    "protein_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "26dfc59c-16db-48ff-a39f-906e12b8e06e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50914"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(protein_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "891cf7e2-2ed4-4bfb-b2bd-50bca1286655",
   "metadata": {},
   "source": [
    "### We need to do some further extraction to get the right format:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7c4627db-4a1c-4f8d-b56c-0cf992d370c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_dataframe(df):\n",
    "    # Convert ID, Sequence, and Mask to string directly\n",
    "    df['ID'] = df['ID'].apply(lambda x: ''.join(x))\n",
    "    df['Sequence'] = df['Sequence'].apply(lambda x: ''.join(x))\n",
    "    df['Mask'] = df['Mask'].apply(lambda x: ''.join(x).replace('+', '1').replace('-', '0'))\n",
    "\n",
    "    # Initialize columns for coordinates and amino acids\n",
    "    coord_columns = ['x', 'y', 'z']\n",
    "    aa_columns = ['A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','Y','Info']\n",
    "\n",
    "    # Expand tertiary coordinates\n",
    "    for i, col in enumerate(coord_columns):\n",
    "        df[col] = df['Tertiary'].apply(lambda x: x[i] if len(x) > i else None)\n",
    "\n",
    "    # Expand evolutionary data into separate columns for each amino acid position\n",
    "    for i, col in enumerate(aa_columns):\n",
    "        df[col] = df['Evolutionary'].apply(lambda x: x[i] if len(x) > i else None)\n",
    "\n",
    "    # Drop the original Tertiary and Evolutionary columns\n",
    "    df.drop(['Tertiary', 'Evolutionary'], axis=1, inplace=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e1c3fe4f-345f-4441-817a-6c408acd6631",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Mask</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>...</th>\n",
       "      <th>N</th>\n",
       "      <th>P</th>\n",
       "      <th>Q</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "      <th>Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3LAF_1_A</td>\n",
       "      <td>DRWGSELESSHHHHHHGGRRSLHFVSEPSDAVTMRGGNVLLNCSAE...</td>\n",
       "      <td>0000000000000000001111111111111111111111111111...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.105...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.9999999999999999, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.037...</td>\n",
       "      <td>...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.126...</td>\n",
       "      <td>[0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.729...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.9999999988691709, 0.9999999988691709, 0.999...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2D5F_d2d5fb1</td>\n",
       "      <td>NECQLNNLNALEPDHRVESEGGLIETWNSQHPELQCAGVTVSKRTL...</td>\n",
       "      <td>1111111111111111111111111111111111111111111111...</td>\n",
       "      <td>[472.3, 574.6, 714.3, 726.5, 853.6, 963.7, 925...</td>\n",
       "      <td>[7239.7, 7348.0, 7287.6, 7174.9, 7104.6, 7189....</td>\n",
       "      <td>[-5819.6, -5811.4, -5816.0, -5884.9, -5896.6, ...</td>\n",
       "      <td>[0.0, 0.03975101015616468, 0.0, 0.007158928206...</td>\n",
       "      <td>[0.0, 0.0, 0.9742406353976603, 0.0008181632235...</td>\n",
       "      <td>[0.0, 0.021185977940373482, 0.0, 0.06463489466...</td>\n",
       "      <td>[0.0, 0.6226930217320084, 0.0, 0.0075680098179...</td>\n",
       "      <td>...</td>\n",
       "      <td>[0.9999999999999999, 0.0, 0.0, 0.0928615258744...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.001115392415331576, 0.0...</td>\n",
       "      <td>[0.0, 0.20181282079283608, 0.0, 0.416342810390...</td>\n",
       "      <td>[0.0, 0.014961231844490552, 0.0, 0.27091429740...</td>\n",
       "      <td>[0.0, 0.06202904881511412, 0.0, 0.022090407036...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.008488443444467173, 0.006489...</td>\n",
       "      <td>[0.0, 0.01550726220377853, 0.0, 0.0, 0.0484688...</td>\n",
       "      <td>[0.0, 0.0, 0.0074058173231726955, 0.0, 0.0, 0....</td>\n",
       "      <td>[0.0, 0.0, 0.010089084469249759, 0.0, 0.0, 0.0...</td>\n",
       "      <td>[0.9999999988691709, 0.5927681381443177, 0.949...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ID                                           Sequence  \\\n",
       "0      3LAF_1_A  DRWGSELESSHHHHHHGGRRSLHFVSEPSDAVTMRGGNVLLNCSAE...   \n",
       "1  2D5F_d2d5fb1  NECQLNNLNALEPDHRVESEGGLIETWNSQHPELQCAGVTVSKRTL...   \n",
       "\n",
       "                                                Mask  \\\n",
       "0  0000000000000000001111111111111111111111111111...   \n",
       "1  1111111111111111111111111111111111111111111111...   \n",
       "\n",
       "                                                   x  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [472.3, 574.6, 714.3, 726.5, 853.6, 963.7, 925...   \n",
       "\n",
       "                                                   y  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [7239.7, 7348.0, 7287.6, 7174.9, 7104.6, 7189....   \n",
       "\n",
       "                                                   z  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [-5819.6, -5811.4, -5816.0, -5884.9, -5896.6, ...   \n",
       "\n",
       "                                                   A  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.105...   \n",
       "1  [0.0, 0.03975101015616468, 0.0, 0.007158928206...   \n",
       "\n",
       "                                                   C  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [0.0, 0.0, 0.9742406353976603, 0.0008181632235...   \n",
       "\n",
       "                                                   D  \\\n",
       "0  [0.9999999999999999, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [0.0, 0.021185977940373482, 0.0, 0.06463489466...   \n",
       "\n",
       "                                                   E  ...  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.037...  ...   \n",
       "1  [0.0, 0.6226930217320084, 0.0, 0.0075680098179...  ...   \n",
       "\n",
       "                                                   N  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [0.9999999999999999, 0.0, 0.0, 0.0928615258744...   \n",
       "\n",
       "                                                   P  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [0.0, 0.0, 0.0, 0.0, 0.001115392415331576, 0.0...   \n",
       "\n",
       "                                                   Q  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.126...   \n",
       "1  [0.0, 0.20181282079283608, 0.0, 0.416342810390...   \n",
       "\n",
       "                                                   R  \\\n",
       "0  [0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [0.0, 0.014961231844490552, 0.0, 0.27091429740...   \n",
       "\n",
       "                                                   S  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.729...   \n",
       "1  [0.0, 0.06202904881511412, 0.0, 0.022090407036...   \n",
       "\n",
       "                                                   T  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [0.0, 0.0, 0.0, 0.008488443444467173, 0.006489...   \n",
       "\n",
       "                                                   V  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [0.0, 0.01550726220377853, 0.0, 0.0, 0.0484688...   \n",
       "\n",
       "                                                   W  \\\n",
       "0  [0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [0.0, 0.0, 0.0074058173231726955, 0.0, 0.0, 0....   \n",
       "\n",
       "                                                   Y  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "1  [0.0, 0.0, 0.010089084469249759, 0.0, 0.0, 0.0...   \n",
       "\n",
       "                                                Info  \n",
       "0  [0.9999999988691709, 0.9999999988691709, 0.999...  \n",
       "1  [0.9999999988691709, 0.5927681381443177, 0.949...  \n",
       "\n",
       "[2 rows x 27 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assume 'df' is your DataFrame loaded with data\n",
    "# Example usage:\n",
    "processed_df = process_dataframe(protein_df)\n",
    "processed_df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da087be6-acdc-41c8-b422-c1fc03c203a1",
   "metadata": {},
   "source": [
    "## Data description:\n",
    "- ID is the protein name\n",
    "- Sequence is the amino acid sequence using the single letter code\n",
    "- Mask (same length as sequence) is a binary representation of how importatnt each amino acid is to the structure (that is, amino acids marked `0` can be changed without affecting the structure much, whereas amino acids marked `1` would significantly alter structure/function).\n",
    "- XYZ are the coordinates of each atom (note there are 3 atoms per amino acid, Nitrogen, alpha-Carbon, and Carbon, so there are 3 times as many coordinates as amino acids in the sequence)\n",
    "- The 20 columns A-Y are normalized PSSM information for each amino acid. This is essentially a sparse matrix where there is a row for each position in the protein sequence, and a column for each of the 20 amino acids.\n",
    "- Info contains the information content for that residue (unclear what this actually is)\n",
    "## Note:\n",
    "Proteins are chains of amino acids, and each amino acid has a backbone of three atoms, N-C-C, which are always linked in the N to C direction: \n",
    "\n",
    "`Beginning of protein sequence amino acids:       1       2       3    ... etc`\n",
    "\n",
    "`Beginning of protein sequence backbone atoms: (N-C-C)-(N-C-C)-(N-C-C)-... etc`\n",
    "\n",
    "So in our final output, we need to have xyz coordinates for each of the three atoms, in order, for every amino acid."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc528612-8da7-473e-8eea-e6785ab7660e",
   "metadata": {},
   "source": [
    "## NGL - load a sample protein to get the parser working \n",
    "\n",
    "https://files.rcsb.org/download/5YHX.pdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ef9da172-b10f-47ef-a581-ea33e4834c9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "738489ef2a0f40e9b16c2aa50fb3f1d5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e8d074f0b4bd443cbdb018950180567a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "from Bio.PDB import PDBParser\n",
    "\n",
    "# Load your PDB file\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure('Sample', './data/5yhx.pdb')\n",
    "\n",
    "# Create NGLView widget for the structure\n",
    "view = nv.show_biopython(structure)\n",
    "view.clear_representations()\n",
    "\n",
    "# Add molecular graphics: e.g., cartoon for secondary structures and ball+stick for atoms\n",
    "view.add_representation('cartoon', selection='protein', color='blue')\n",
    "view.add_representation('ball+stick', selection='ligand')\n",
    "\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "417a2857-aef0-4fa1-b5da-c8e8b02d5701",
   "metadata": {},
   "source": [
    "## Here is a parser to turn a row of our dataframe into a PDB file that can be viewed with NGL. Note that our dataset doesn't have info on the secondary structure, so these structures will not appear bolded in the output. The PDB specification is available here:\n",
    "\n",
    "https://www.biostat.jhsph.edu/~iruczins/teaching/260.655/links/pdbformat.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8f4d1567-519d-41d4-a427-bead0e15e95c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def format_pdb_from_df(row):\n",
    "    pdb_str = \"\"\n",
    "    atom_count = 1  # Starting index for ATOM records\n",
    "    chain_id = 'A'  # Assuming a single chain for simplicity\n",
    "    res_num = 1     # Residue number\n",
    "    occupancy = 1.00\n",
    "    t_factor = 50.00\n",
    "\n",
    "    atom_types = ['N', 'CA', 'C']  # Backbone atoms\n",
    "    element_types = {'N': 'N', 'CA': 'C', 'C': 'C'}  # Elements for atoms\n",
    "    residue_mapping = {'A': 'ALA','R': 'ARG','N': 'ASN','D': 'ASP','C': 'CYS','E': 'GLU','Q': 'GLN','G': 'GLY',\n",
    "                       'H': 'HIS','I': 'ILE','L': 'LEU','K': 'LYS','M': 'MET','F': 'PHE','P': 'PRO','S': 'SER',\n",
    "                       'T': 'THR','W': 'TRP','Y': 'TYR','V': 'VAL' } # PDB requires the 3 letter codes for amino acids\n",
    "\n",
    "    for i in range(len(row['Sequence'])):\n",
    "        residue = residue_mapping[row['Sequence'][i]]\n",
    "        # Loop over each backbone atom type\n",
    "        for j, atom_type in enumerate(atom_types):\n",
    "            # Index to pull the correct coordinates from flattened list\n",
    "            idx = 3 * i + j\n",
    "            x = row['x'][idx]/100\n",
    "            y = row['y'][idx]/100\n",
    "            z = row['z'][idx]/100\n",
    "            element = element_types[atom_type]\n",
    "\n",
    "            pdb_str += f\"ATOM  {atom_count:>5}  {atom_type:<2}  {residue:>3} {chain_id}{res_num:>4}    {x:>8.3f}{y:>8.3f}{z:>8.3f}{occupancy:>6.2f}{t_factor:>6.2f}          {element:>2}  \\n\"\n",
    "            atom_count += 1\n",
    "\n",
    "        res_num += 1  # Increment residue number for each new amino acid\n",
    "\n",
    "    return pdb_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c1dfffb0-3e2f-4313-a6bd-962a1da2f662",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATOM      1  N   ASN A   1      15.878  29.275  76.731  1.00 50.00           N  \n",
      "ATOM      2  CA  ASN A   1      15.137  28.332  75.915  1.00 50.00           C  \n",
      "ATOM      3  C   ASN A   1      15.539  26.931  76.360  1.00 50.00           C  \n",
      "ATOM      4  N   LYS A   2      14.647  26.274  77.101  1.00 50.00           N  \n",
      "ATOM      5  CA  LYS A   2      14.972  25.005  77.746  1.00 50.00           C  \n",
      "ATOM      6  C   LYS A   2      15.109  23.930  76.692  1.00 50.00           C  \n",
      "ATOM      7  N   THR A   3      16.130  23.093  76.818  1.00 50.00           N  \n",
      "ATOM      8  CA  THR A   3      16.246  21.915  75.966  1.00 50.00           C  \n",
      "ATOM      9  C   THR A   3      15.103  20.938  76.293  1.00 50.00           C  \n",
      "ATOM     10  N   SER A   4      14.511  20.344  75.257  1.00 50.00           N  \n",
      "ATOM     11  CA  SER A   4      13.358  19.451  75.405  1.00 50.00           C  \n",
      "ATOM     12  C   SER A   4      13.070  18.756  74.091  1.00 50.00           C  \n",
      "ATOM     13  N   GLU A   5      13.223  17.438  74.071  1.00 50.00           N  \n",
      "ATOM     14  CA  GLU A   5      12.877  16.656  72.890  1.00 50.00           C  \n",
      "ATOM     15  C   GLU A   5      11.384  16.737  72.646  1.00 50.00           C  \n",
      "ATOM     16  N   THR A   6      11.041  16.862  71.378  1.00 50.00           N  \n",
      "ATOM     17  CA  THR A   6       9.669  16.922  70.938  1.00 50.00           C  \n",
      "ATOM     18  C   THR A   6       9.697  16.361  69.539  1.00 50.00           C  \n",
      "ATOM     19  N   ASN A   7       9.281  15.108  69.407  1.00 50.00           N  \n",
      "ATOM     20  CA  ASN A   7       9.113  14.517  68.098  1.00 50.00           C  \n",
      "ATOM     21  C   ASN A   7       7.693  14.807  67.693  1.00 50.00           C  \n",
      "ATOM     22  N   THR A   8       7.540  15.433  66.535  1.00 50.00           N  \n",
      "ATOM     23  CA  THR A   8       6.237  15.806  66.029  1.00 50.00           C  \n",
      "ATOM     24  C   THR A   8       6.074  15.141  64.663  1.00 50.00           C  \n",
      "ATOM     25  N   PRO A   9       4.918  14.526  64.425  1.00 50.00           N  \n",
      "ATOM     26  CA  PRO A   9       4.715  13.732  63.214  1.00 50.00           C  \n",
      "ATOM     27  C   PRO A   9       4.959  14.532  61.936  1.00 50.00           C  \n",
      "ATOM     28  N   LEU A  10       5.755  13.969  61.042  1.00 50.00           N  \n",
      "ATOM     29  CA  LEU A  10       6.063  14.596  59.778  1.00 50.00           C  \n",
      "ATOM     30  C   LEU A  10       5.632  13.648  58.679  1.00 50.00           C  \n",
      "ATOM     31  N   PHE A  11       4.681  14.094  57.870  1.00 50.00           N  \n",
      "ATOM     32  CA  PHE A  11       4.200  13.327  56.733  1.00 50.00           C  \n",
      "ATOM     33  C   PHE A  11       4.775  13.965  55.482  1.00 50.00           C  \n",
      "ATOM     34  N   VAL A  12       5.312  13.145  54.588  1.00 50.00           N  \n",
      "ATOM     35  CA  VAL A  12       5.797  13.621  53.294  1.00 50.00           C  \n",
      "ATOM     36  C   VAL A  12       5.125  12.822  52.200  1.00 50.00           C  \n",
      "ATOM     37  N   ASN A  13       4.422  13.523  51.317  1.00 50.00           N  \n",
      "ATOM     38  CA  ASN A  13       3.746  12.910  50.185  1.00 50.00           C  \n",
      "ATOM     39  C   ASN A  13       4.354  13.470  48.914  1.00 50.00           C  \n",
      "ATOM     40  N   LYS A  14       4.716  12.582  47.987  1.00 50.00           N  \n",
      "ATOM     41  CA  LYS A  14       5.414  12.959  46.764  1.00 50.00           C  \n",
      "ATOM     42  C   LYS A  14       4.574  12.585  45.539  1.00 50.00           C  \n",
      "ATOM     43  N   VAL A  15       3.881  13.583  45.009  1.00 50.00           N  \n",
      "ATOM     44  CA  VAL A  15       3.026  13.423  43.850  1.00 50.00           C  \n",
      "ATOM     45  C   VAL A  15       3.837  13.600  42.575  1.00 50.00           C  \n",
      "ATOM     46  N   ASN A  16       3.881  12.559  41.750  1.00 50.00           N  \n",
      "ATOM     47  CA  ASN A  16       4.459  12.641  40.416  1.00 50.00           C  \n",
      "ATOM     48  C   ASN A  16       3.430  12.222  39.365  1.00 50.00           C  \n",
      "ATOM     49  N   GLY A  17       2.695  13.194  38.842  1.00 50.00           N  \n",
      "ATOM     50  CA  GLY A  17       1.615  12.904  37.929  1.00 50.00           C  \n",
      "ATOM     51  C   GLY A  17       0.560  12.058  38.617  1.00 50.00           C  \n",
      "ATOM     52  N   GLU A  18       0.510  10.778  38.261  1.00 50.00           N  \n",
      "ATOM     53  CA  GLU A  18      -0.508   9.867  38.779  1.00 50.00           C  \n",
      "ATOM     54  C   GLU A  18      -0.022   9.046  39.963  1.00 50.00           C  \n",
      "ATOM     55  N   ASP A  19       1.283   9.082  40.233  1.00 50.00           N  \n",
      "ATOM     56  CA  ASP A  19       1.875   8.332  41.341  1.00 50.00           C  \n",
      "ATOM     57  C   ASP A  19       1.926   9.166  42.610  1.00 50.00           C  \n",
      "ATOM     58  N   LEU A  20       1.536   8.558  43.723  1.00 50.00           N  \n",
      "ATOM     59  CA  LEU A  20       1.665   9.170  45.034  1.00 50.00           C  \n",
      "ATOM     60  C   LEU A  20       2.583   8.314  45.899  1.00 50.00           C  \n",
      "ATOM     61  N   ASP A  21       3.628   8.935  46.429  1.00 50.00           N  \n",
      "ATOM     62  CA  ASP A  21       4.602   8.243  47.236  1.00 50.00           C  \n",
      "ATOM     63  C   ASP A  21       4.602   8.850  48.630  1.00 50.00           C  \n",
      "ATOM     64  N   ALA A  22       3.802   8.277  49.521  1.00 50.00           N  \n",
      "ATOM     65  CA  ALA A  22       3.518   8.898  50.808  1.00 50.00           C  \n",
      "ATOM     66  C   ALA A  22       4.120   8.125  51.968  1.00 50.00           C  \n",
      "ATOM     67  N   SER A  23       4.884   8.813  52.811  1.00 50.00           N  \n",
      "ATOM     68  CA  SER A  23       5.606   8.164  53.896  1.00 50.00           C  \n",
      "ATOM     69  C   SER A  23       5.462   8.903  55.225  1.00 50.00           C  \n",
      "ATOM     70  N   ILE A  24       5.612   8.162  56.318  1.00 50.00           N  \n",
      "ATOM     71  CA  ILE A  24       5.623   8.733  57.663  1.00 50.00           C  \n",
      "ATOM     72  C   ILE A  24       7.041   8.940  58.170  1.00 50.00           C  \n",
      "ATOM     73  N   ASP A  25       7.208  10.040  58.880  1.00 50.00           N  \n",
      "ATOM     74  CA  ASP A  25       8.497  10.481  59.385  1.00 50.00           C  \n",
      "ATOM     75  C   ASP A  25       8.262  11.200  60.709  1.00 50.00           C  \n",
      "ATOM     76  N   SER A  26       9.342  11.684  61.307  1.00 50.00           N  \n",
      "ATOM     77  CA  SER A  26       9.260  12.533  62.481  1.00 50.00           C  \n",
      "ATOM     78  C   SER A  26      10.161  13.757  62.329  1.00 50.00           C  \n",
      "ATOM     79  N   PHE A  27       9.857  14.809  63.083  1.00 50.00           N  \n",
      "ATOM     80  CA  PHE A  27      10.679  16.014  63.091  1.00 50.00           C  \n",
      "ATOM     81  C   PHE A  27      11.118  16.350  64.503  1.00 50.00           C  \n",
      "ATOM     82  N   LEU A  28      12.410  16.601  64.634  1.00 50.00           N  \n",
      "ATOM     83  CA  LEU A  28      13.071  16.718  65.918  1.00 50.00           C  \n",
      "ATOM     84  C   LEU A  28      13.232  18.180  66.343  1.00 50.00           C  \n",
      "ATOM     85  N   ILE A  29      12.589  18.558  67.444  1.00 50.00           N  \n",
      "ATOM     86  CA  ILE A  29      12.766  19.884  68.039  1.00 50.00           C  \n",
      "ATOM     87  C   ILE A  29      13.353  19.710  69.431  1.00 50.00           C  \n",
      "ATOM     88  N   GLN A  30      14.195  20.643  69.847  1.00 50.00           N  \n",
      "ATOM     89  CA  GLN A  30      14.965  20.469  71.072  1.00 50.00           C  \n",
      "ATOM     90  C   GLN A  30      14.866  21.667  72.012  1.00 50.00           C  \n",
      "ATOM     91  N   LYS A  31      13.734  22.374  71.959  1.00 50.00           N  \n",
      "ATOM     92  CA  LYS A  31      13.531  23.604  72.726  1.00 50.00           C  \n",
      "ATOM     93  C   LYS A  31      12.038  23.903  72.975  1.00 50.00           C  \n",
      "ATOM     94  N   GLU A  32      11.730  24.663  74.029  1.00 50.00           N  \n",
      "ATOM     95  CA  GLU A  32      10.346  25.088  74.318  1.00 50.00           C  \n",
      "ATOM     96  C   GLU A  32       9.810  26.020  73.227  1.00 50.00           C  \n",
      "ATOM     97  N   GLU A  33      10.727  26.741  72.584  1.00 50.00           N  \n",
      "ATOM     98  CA  GLU A  33      10.410  27.703  71.533  1.00 50.00           C  \n",
      "ATOM     99  C   GLU A  33      11.272  27.440  70.299  1.00 50.00           C  \n",
      "ATOM    100  N   ILE A  34      10.705  27.728  69.136  1.00 50.00           N  \n",
      "ATOM    101  CA  ILE A  34      11.409  27.567  67.866  1.00 50.00           C  \n",
      "ATOM    102  C   ILE A  34      10.886  28.615  66.889  1.00 50.00           C  \n",
      "ATOM    103  N   SER A  35      11.804  29.356  66.271  1.00 50.00           N  \n",
      "ATOM    104  CA  SER A  35      11.439  30.327  65.239  1.00 50.00           C  \n",
      "ATOM    105  C   SER A  35      10.942  29.586  64.009  1.00 50.00           C  \n",
      "ATOM    106  N   LEU A  36      10.070  30.230  63.242  1.00 50.00           N  \n",
      "ATOM    107  CA  LEU A  36       9.635  29.696  61.957  1.00 50.00           C  \n",
      "ATOM    108  C   LEU A  36      10.841  29.522  61.044  1.00 50.00           C  \n",
      "ATOM    109  N   LYS A  37      11.786  30.457  61.104  1.00 50.00           N  \n",
      "ATOM    110  CA  LYS A  37      13.044  30.338  60.360  1.00 50.00           C  \n",
      "ATOM    111  C   LYS A  37      13.693  28.983  60.539  1.00 50.00           C  \n",
      "ATOM    112  N   GLU A  38      13.770  28.518  61.781  1.00 50.00           N  \n",
      "ATOM    113  CA  GLU A  38      14.394  27.236  62.096  1.00 50.00           C  \n",
      "ATOM    114  C   GLU A  38      13.535  26.028  61.687  1.00 50.00           C  \n",
      "ATOM    115  N   LEU A  39      12.213  26.126  61.823  1.00 50.00           N  \n",
      "ATOM    116  CA  LEU A  39      11.327  25.061  61.361  1.00 50.00           C  \n",
      "ATOM    117  C   LEU A  39      11.472  24.939  59.861  1.00 50.00           C  \n",
      "ATOM    118  N   ASP A  40      11.160  26.032  59.180  1.00 50.00           N  \n",
      "ATOM    119  CA  ASP A  40      11.303  26.125  57.747  1.00 50.00           C  \n",
      "ATOM    120  C   ASP A  40      12.629  25.552  57.266  1.00 50.00           C  \n",
      "ATOM    121  N   PHE A  41      13.722  25.941  57.905  1.00 50.00           N  \n",
      "ATOM    122  CA  PHE A  41      15.044  25.575  57.415  1.00 50.00           C  \n",
      "ATOM    123  C   PHE A  41      15.334  24.094  57.585  1.00 50.00           C  \n",
      "ATOM    124  N   LYS A  42      15.174  23.592  58.804  1.00 50.00           N  \n",
      "ATOM    125  CA  LYS A  42      15.496  22.204  59.117  1.00 50.00           C  \n",
      "ATOM    126  C   LYS A  42      14.533  21.210  58.466  1.00 50.00           C  \n",
      "ATOM    127  N   ILE A  43      13.259  21.579  58.348  1.00 50.00           N  \n",
      "ATOM    128  CA  ILE A  43      12.293  20.765  57.607  1.00 50.00           C  \n",
      "ATOM    129  C   ILE A  43      12.747  20.615  56.159  1.00 50.00           C  \n",
      "ATOM    130  N   ARG A  44      13.072  21.725  55.506  1.00 50.00           N  \n",
      "ATOM    131  CA  ARG A  44      13.497  21.697  54.101  1.00 50.00           C  \n",
      "ATOM    132  C   ARG A  44      14.812  20.956  53.909  1.00 50.00           C  \n",
      "ATOM    133  N   GLN A  45      15.742  21.138  54.836  1.00 50.00           N  \n",
      "ATOM    134  CA  GLN A  45      16.995  20.390  54.845  1.00 50.00           C  \n",
      "ATOM    135  C   GLN A  45      16.728  18.889  55.026  1.00 50.00           C  \n",
      "ATOM    136  N   GLN A  46      15.701  18.537  55.793  1.00 50.00           N  \n",
      "ATOM    137  CA  GLN A  46      15.306  17.135  55.949  1.00 50.00           C  \n",
      "ATOM    138  C   GLN A  46      14.719  16.591  54.635  1.00 50.00           C  \n",
      "ATOM    139  N   LEU A  47      13.961  17.418  53.926  1.00 50.00           N  \n",
      "ATOM    140  CA  LEU A  47      13.385  17.035  52.644  1.00 50.00           C  \n",
      "ATOM    141  C   LEU A  47      14.452  16.969  51.554  1.00 50.00           C  \n",
      "ATOM    142  N   VAL A  48      15.484  17.792  51.696  1.00 50.00           N  \n",
      "ATOM    143  CA  VAL A  48      16.581  17.818  50.746  1.00 50.00           C  \n",
      "ATOM    144  C   VAL A  48      17.397  16.549  50.896  1.00 50.00           C  \n",
      "ATOM    145  N   ASN A  49      17.574  16.116  52.140  1.00 50.00           N  \n",
      "ATOM    146  CA  ASN A  49      18.501  15.040  52.452  1.00 50.00           C  \n",
      "ATOM    147  C   ASN A  49      17.912  13.654  52.217  1.00 50.00           C  \n",
      "ATOM    148  N   ASN A  50      16.617  13.494  52.483  1.00 50.00           N  \n",
      "ATOM    149  CA  ASN A  50      15.975  12.178  52.452  1.00 50.00           C  \n",
      "ATOM    150  C   ASN A  50      14.848  12.016  51.432  1.00 50.00           C  \n",
      "ATOM    151  N   TYR A  51      14.359  13.118  50.873  1.00 50.00           N  \n",
      "ATOM    152  CA  TYR A  51      13.260  13.054  49.914  1.00 50.00           C  \n",
      "ATOM    153  C   TYR A  51      13.582  13.723  48.574  1.00 50.00           C  \n",
      "ATOM    154  N   GLY A  52      14.871  13.898  48.289  1.00 50.00           N  \n",
      "ATOM    155  CA  GLY A  52      15.316  14.428  47.012  1.00 50.00           C  \n",
      "ATOM    156  C   GLY A  52      14.717  15.768  46.637  1.00 50.00           C  \n",
      "ATOM    157  N   LEU A  53      14.573  16.653  47.618  1.00 50.00           N  \n",
      "ATOM    158  CA  LEU A  53      14.026  17.979  47.359  1.00 50.00           C  \n",
      "ATOM    159  C   LEU A  53      15.103  18.871  46.745  1.00 50.00           C  \n",
      "ATOM    160  N   TYR A  54      14.680  19.685  45.782  1.00 50.00           N  \n",
      "ATOM    161  CA  TYR A  54      15.538  20.595  45.031  1.00 50.00           C  \n",
      "ATOM    162  C   TYR A  54      16.692  19.886  44.308  1.00 50.00           C  \n",
      "ATOM    163  N   LYS A  55      16.420  18.707  43.742  1.00 50.00           N  \n",
      "ATOM    164  CA  LYS A  55      17.392  18.039  42.866  1.00 50.00           C  \n",
      "ATOM    165  C   LYS A  55      16.774  17.654  41.507  1.00 50.00           C  \n",
      "ATOM    166  N   GLY A  56      15.693  16.877  41.509  1.00 50.00           N  \n",
      "ATOM    167  CA  GLY A  56      15.125  16.364  40.270  1.00 50.00           C  \n",
      "ATOM    168  C   GLY A  56      14.160  17.356  39.650  1.00 50.00           C  \n",
      "ATOM    169  N   THR A  57      12.881  16.982  39.623  1.00 50.00           N  \n",
      "ATOM    170  CA  THR A  57      11.786  17.913  39.332  1.00 50.00           C  \n",
      "ATOM    171  C   THR A  57      11.161  18.410  40.648  1.00 50.00           C  \n",
      "ATOM    172  N   SER A  58      11.700  17.953  41.781  1.00 50.00           N  \n",
      "ATOM    173  CA  SER A  58      11.172  18.320  43.097  1.00 50.00           C  \n",
      "ATOM    174  C   SER A  58      11.542  19.763  43.458  1.00 50.00           C  \n",
      "ATOM    175  N   LYS A  59      10.535  20.629  43.518  1.00 50.00           N  \n",
      "ATOM    176  CA  LYS A  59      10.737  22.049  43.795  1.00 50.00           C  \n",
      "ATOM    177  C   LYS A  59       9.473  22.665  44.393  1.00 50.00           C  \n",
      "ATOM    178  N   TYR A  60       8.357  22.511  43.685  1.00 50.00           N  \n",
      "ATOM    179  CA  TYR A  60       7.074  23.069  44.107  1.00 50.00           C  \n",
      "ATOM    180  C   TYR A  60       6.395  22.224  45.185  1.00 50.00           C  \n",
      "ATOM    181  N   GLY A  61       5.356  22.784  45.794  1.00 50.00           N  \n",
      "ATOM    182  CA  GLY A  61       4.643  22.126  46.876  1.00 50.00           C  \n",
      "ATOM    183  C   GLY A  61       4.549  22.982  48.123  1.00 50.00           C  \n",
      "ATOM    184  N   LYS A  62       4.152  22.372  49.239  1.00 50.00           N  \n",
      "ATOM    185  CA  LYS A  62       3.933  23.106  50.488  1.00 50.00           C  \n",
      "ATOM    186  C   LYS A  62       4.239  22.310  51.750  1.00 50.00           C  \n",
      "ATOM    187  N   ILE A  63       4.591  23.040  52.809  1.00 50.00           N  \n",
      "ATOM    188  CA  ILE A  63       4.693  22.531  54.174  1.00 50.00           C  \n",
      "ATOM    189  C   ILE A  63       3.475  23.069  54.932  1.00 50.00           C  \n",
      "ATOM    190  N   ILE A  64       2.829  22.222  55.727  1.00 50.00           N  \n",
      "ATOM    191  CA  ILE A  64       1.634  22.601  56.480  1.00 50.00           C  \n",
      "ATOM    192  C   ILE A  64       1.854  22.242  57.933  1.00 50.00           C  \n",
      "ATOM    193  N   ILE A  65       1.874  23.239  58.804  1.00 50.00           N  \n",
      "ATOM    194  CA  ILE A  65       2.073  23.002  60.222  1.00 50.00           C  \n",
      "ATOM    195  C   ILE A  65       0.768  23.279  60.948  1.00 50.00           C  \n",
      "ATOM    196  N   ASN A  66       0.290  22.276  61.679  1.00 50.00           N  \n",
      "ATOM    197  CA  ASN A  66      -0.891  22.410  62.516  1.00 50.00           C  \n",
      "ATOM    198  C   ASN A  66      -0.453  22.986  63.852  1.00 50.00           C  \n",
      "ATOM    199  N   LEU A  67      -0.999  24.149  64.198  1.00 50.00           N  \n",
      "ATOM    200  CA  LEU A  67      -0.592  24.856  65.411  1.00 50.00           C  \n",
      "ATOM    201  C   LEU A  67      -1.758  24.951  66.382  1.00 50.00           C  \n",
      "ATOM    202  N   LYS A  68      -1.803  24.020  67.334  1.00 50.00           N  \n",
      "ATOM    203  CA  LYS A  68      -2.786  24.050  68.417  1.00 50.00           C  \n",
      "ATOM    204  C   LYS A  68      -4.230  23.887  67.897  1.00 50.00           C  \n",
      "ATOM    205  N   ASP A  69      -4.360  23.372  66.669  1.00 50.00           N  \n",
      "ATOM    206  CA  ASP A  69      -5.634  23.265  65.929  1.00 50.00           C  \n",
      "ATOM    207  C   ASP A  69      -6.324  24.609  65.583  1.00 50.00           C  \n",
      "ATOM    208  N   GLU A  70      -5.663  25.732  65.864  1.00 50.00           N  \n",
      "ATOM    209  CA  GLU A  70      -6.260  27.056  65.678  1.00 50.00           C  \n",
      "ATOM    210  C   GLU A  70      -5.705  27.817  64.472  1.00 50.00           C  \n",
      "ATOM    211  N   ASN A  71      -4.444  27.568  64.116  1.00 50.00           N  \n",
      "ATOM    212  CA  ASN A  71      -3.809  28.243  62.978  1.00 50.00           C  \n",
      "ATOM    213  C   ASN A  71      -2.991  27.288  62.116  1.00 50.00           C  \n",
      "ATOM    214  N   LYS A  72      -3.162  27.394  60.800  1.00 50.00           N  \n",
      "ATOM    215  CA  LYS A  72      -2.361  26.655  59.840  1.00 50.00           C  \n",
      "ATOM    216  C   LYS A  72      -1.250  27.571  59.379  1.00 50.00           C  \n",
      "ATOM    217  N   VAL A  73      -0.012  27.148  59.592  1.00 50.00           N  \n",
      "ATOM    218  CA  VAL A  73       1.134  27.861  59.060  1.00 50.00           C  \n",
      "ATOM    219  C   VAL A  73       1.540  27.179  57.763  1.00 50.00           C  \n",
      "ATOM    220  N   GLU A  74       1.322  27.867  56.648  1.00 50.00           N  \n",
      "ATOM    221  CA  GLU A  74       1.569  27.331  55.319  1.00 50.00           C  \n",
      "ATOM    222  C   GLU A  74       2.816  27.985  54.744  1.00 50.00           C  \n",
      "ATOM    223  N   ILE A  75       3.739  27.168  54.245  1.00 50.00           N  \n",
      "ATOM    224  CA  ILE A  75       4.974  27.643  53.638  1.00 50.00           C  \n",
      "ATOM    225  C   ILE A  75       5.075  27.087  52.232  1.00 50.00           C  \n",
      "ATOM    226  N   ASP A  76       5.435  27.939  51.278  1.00 50.00           N  \n",
      "ATOM    227  CA  ASP A  76       5.582  27.535  49.887  1.00 50.00           C  \n",
      "ATOM    228  C   ASP A  76       7.019  27.086  49.612  1.00 50.00           C  \n",
      "ATOM    229  N   LEU A  77       7.182  25.840  49.164  1.00 50.00           N  \n",
      "ATOM    230  CA  LEU A  77       8.497  25.282  48.815  1.00 50.00           C  \n",
      "ATOM    231  C   LEU A  77       9.102  25.895  47.545  1.00 50.00           C  \n",
      "ATOM    232  N   GLY A  78       8.257  26.493  46.707  1.00 50.00           N  \n",
      "ATOM    233  CA  GLY A  78       8.684  27.084  45.448  1.00 50.00           C  \n",
      "ATOM    234  C   GLY A  78       9.630  28.268  45.556  1.00 50.00           C  \n",
      "ATOM    235  N   ASP A  79       9.589  28.958  46.693  1.00 50.00           N  \n",
      "ATOM    236  CA  ASP A  79      10.472  30.092  46.956  1.00 50.00           C  \n",
      "ATOM    237  C   ASP A  79      11.106  29.975  48.338  1.00 50.00           C  \n",
      "ATOM    238  N   LYS A  80      12.167  30.750  48.566  1.00 50.00           N  \n",
      "ATOM    239  CA  LYS A  80      12.728  30.902  49.908  1.00 50.00           C  \n",
      "ATOM    240  C   LYS A  80      11.629  31.510  50.771  1.00 50.00           C  \n",
      "ATOM    241  N   LEU A  81      11.636  31.179  52.061  1.00 50.00           N  \n",
      "ATOM    242  CA  LEU A  81      10.659  31.704  53.009  1.00 50.00           C  \n",
      "ATOM    243  C   LEU A  81      10.728  33.227  53.030  1.00 50.00           C  \n",
      "ATOM    244  N   GLN A  82       9.570  33.873  53.104  1.00 50.00           N  \n",
      "ATOM    245  CA  GLN A  82       9.515  35.330  53.193  1.00 50.00           C  \n",
      "ATOM    246  C   GLN A  82      10.293  35.809  54.425  1.00 50.00           C  \n",
      "ATOM    247  N   PHE A  83      11.185  36.776  54.228  1.00 50.00           N  \n",
      "ATOM    248  CA  PHE A  83      12.059  37.285  55.284  1.00 50.00           C  \n",
      "ATOM    249  C   PHE A  83      11.295  37.788  56.510  1.00 50.00           C  \n",
      "ATOM    250  N   GLU A  84      10.113  38.360  56.292  1.00 50.00           N  \n",
      "ATOM    251  CA  GLU A  84       9.330  38.961  57.374  1.00 50.00           C  \n",
      "ATOM    252  C   GLU A  84       8.698  37.925  58.312  1.00 50.00           C  \n",
      "ATOM    253  N   ARG A  85       8.652  36.671  57.872  1.00 50.00           N  \n",
      "ATOM    254  CA  ARG A  85       8.095  35.575  58.668  1.00 50.00           C  \n",
      "ATOM    255  C   ARG A  85       9.135  34.888  59.555  1.00 50.00           C  \n",
      "ATOM    256  N   MET A  86      10.410  35.132  59.285  1.00 50.00           N  \n",
      "ATOM    257  CA  MET A  86      11.491  34.478  60.019  1.00 50.00           C  \n",
      "ATOM    258  C   MET A  86      11.537  34.850  61.498  1.00 50.00           C  \n",
      "ATOM    259  N   GLY A  87      11.017  36.025  61.838  1.00 50.00           N  \n",
      "ATOM    260  CA  GLY A  87      10.968  36.472  63.217  1.00 50.00           C  \n",
      "ATOM    261  C   GLY A  87       9.887  35.787  64.033  1.00 50.00           C  \n",
      "ATOM    262  N   ASP A  88       8.927  35.146  63.358  1.00 50.00           N  \n",
      "ATOM    263  CA  ASP A  88       7.840  34.406  64.012  1.00 50.00           C  \n",
      "ATOM    264  C   ASP A  88       8.372  33.328  64.939  1.00 50.00           C  \n",
      "ATOM    265  N   VAL A  89       7.615  33.049  65.989  1.00 50.00           N  \n",
      "ATOM    266  CA  VAL A  89       8.023  32.075  66.990  1.00 50.00           C  \n",
      "ATOM    267  C   VAL A  89       6.855  31.144  67.289  1.00 50.00           C  \n",
      "ATOM    268  N   LEU A  90       7.169  29.862  67.441  1.00 50.00           N  \n",
      "ATOM    269  CA  LEU A  90       6.172  28.825  67.663  1.00 50.00           C  \n",
      "ATOM    270  C   LEU A  90       6.445  28.123  68.993  1.00 50.00           C  \n",
      "ATOM    271  N   ASN A  91       5.377  27.728  69.680  1.00 50.00           N  \n",
      "ATOM    272  CA  ASN A  91       5.501  26.935  70.900  1.00 50.00           C  \n",
      "ATOM    273  C   ASN A  91       5.631  25.454  70.529  1.00 50.00           C  \n",
      "ATOM    274  N   SER A  92       6.795  24.871  70.788  1.00 50.00           N  \n",
      "ATOM    275  CA  SER A  92       7.123  23.525  70.314  1.00 50.00           C  \n",
      "ATOM    276  C   SER A  92       6.095  22.451  70.662  1.00 50.00           C  \n",
      "ATOM    277  N   LYS A  93       5.561  22.508  71.876  1.00 50.00           N  \n",
      "ATOM    278  CA  LYS A  93       4.570  21.536  72.326  1.00 50.00           C  \n",
      "ATOM    279  C   LYS A  93       3.201  21.718  71.648  1.00 50.00           C  \n",
      "ATOM    280  N   ASP A  94       2.923  22.913  71.128  1.00 50.00           N  \n",
      "ATOM    281  CA  ASP A  94       1.652  23.187  70.446  1.00 50.00           C  \n",
      "ATOM    282  C   ASP A  94       1.632  22.762  68.969  1.00 50.00           C  \n",
      "ATOM    283  N   ILE A  95       2.789  22.383  68.429  1.00 50.00           N  \n",
      "ATOM    284  CA  ILE A  95       2.870  21.882  67.064  1.00 50.00           C  \n",
      "ATOM    285  C   ILE A  95       2.364  20.439  67.062  1.00 50.00           C  \n",
      "ATOM    286  N   ARG A  96       1.193  20.247  66.467  1.00 50.00           N  \n",
      "ATOM    287  CA  ARG A  96       0.516  18.959  66.449  1.00 50.00           C  \n",
      "ATOM    288  C   ARG A  96       1.040  17.988  65.391  1.00 50.00           C  \n",
      "ATOM    289  N   GLY A  97       1.792  18.503  64.416  1.00 50.00           N  \n",
      "ATOM    290  CA  GLY A  97       2.232  17.724  63.262  1.00 50.00           C  \n",
      "ATOM    291  C   GLY A  97       2.554  18.585  62.052  1.00 50.00           C  \n",
      "ATOM    292  N   ILE A  98       3.234  18.007  61.060  1.00 50.00           N  \n",
      "ATOM    293  CA  ILE A  98       3.615  18.693  59.817  1.00 50.00           C  \n",
      "ATOM    294  C   ILE A  98       3.281  17.807  58.611  1.00 50.00           C  \n",
      "ATOM    295  N   SER A  99       2.643  18.356  57.575  1.00 50.00           N  \n",
      "ATOM    296  CA  SER A  99       2.351  17.598  56.358  1.00 50.00           C  \n",
      "ATOM    297  C   SER A  99       2.892  18.286  55.109  1.00 50.00           C  \n",
      "ATOM    298  N   VAL A 100       3.982  17.728  54.601  1.00 50.00           N  \n",
      "ATOM    299  CA  VAL A 100       4.601  18.204  53.376  1.00 50.00           C  \n",
      "ATOM    300  C   VAL A 100       4.010  17.490  52.171  1.00 50.00           C  \n",
      "ATOM    301  N   THR A 101       3.760  18.249  51.118  1.00 50.00           N  \n",
      "ATOM    302  CA  THR A 101       3.347  17.693  49.839  1.00 50.00           C  \n",
      "ATOM    303  C   THR A 101       4.191  18.336  48.750  1.00 50.00           C  \n",
      "ATOM    304  N   ILE A 102       5.075  17.546  48.154  1.00 50.00           N  \n",
      "ATOM    305  CA  ILE A 102       5.901  17.979  47.035  1.00 50.00           C  \n",
      "ATOM    306  C   ILE A 102       5.145  17.580  45.790  1.00 50.00           C  \n",
      "ATOM    307  N   ASN A 103       5.039  18.502  44.827  1.00 50.00           N  \n",
      "ATOM    308  CA  ASN A 103       4.241  18.315  43.607  1.00 50.00           C  \n",
      "ATOM    309  C   ASN A 103       5.122  18.248  42.353  1.00 50.00           C  \n",
      "ATOM    310  N   GLN A 104       5.904  17.168  42.248  1.00 50.00           N  \n",
      "ATOM    311  CA  GLN A 104       6.870  16.936  41.152  1.00 50.00           C  \n",
      "ATOM    312  C   GLN A 104       6.355  17.330  39.765  1.00 50.00           C  \n",
      "ATOM    313  N   ILE A 105       6.672  18.558  39.359  1.00 50.00           N  \n",
      "ATOM    314  CA  ILE A 105       6.251  19.120  38.073  1.00 50.00           C  \n",
      "ATOM    315  C   ILE A 105       7.339  20.084  37.570  1.00 50.00           C  \n",
      "\n"
     ]
    }
   ],
   "source": [
    "protein_row = processed_df.iloc[10]\n",
    "\n",
    "pdb_content = format_pdb_from_df(protein_row)\n",
    "print(pdb_content)\n",
    "\n",
    "# Optionally, write to a file\n",
    "with open(\"./data/sample_protein.pdb\", \"w\") as file:\n",
    "    file.write(pdb_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "413fba38-fba4-4306-b8f2-be50c64f0d56",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "de15c24b4ec94342936923a897e4cdab",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "from Bio.PDB import PDBParser\n",
    "\n",
    "# Load your PDB file\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure('Sample', './data/sample_protein.pdb')\n",
    "\n",
    "# Create NGLView widget for the structure\n",
    "view = nv.show_biopython(structure)\n",
    "view.clear_representations()\n",
    "\n",
    "# Add molecular graphics: e.g., cartoon for secondary structures and ball+stick for atoms\n",
    "view.add_representation('cartoon', selection='protein', color='blue')\n",
    "view.add_representation('ball+stick', selection='ligand')\n",
    "\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51384047-59e9-42f9-b088-0ae811a8e6b1",
   "metadata": {},
   "source": [
    "## add features for GNN\n",
    "### For nodes (amino acids) we will add some of their chemical properties)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5abdd08f-99f0-42f2-81dc-384e7eba1228",
   "metadata": {},
   "outputs": [],
   "source": [
    "def amino_acid_properties():\n",
    "    properties = {\n",
    "        'Node_Hydrophobic': \"AYCMVLWIF\",\n",
    "        'Node_Hydrophilic': \"RKNEPD\",\n",
    "        'Node_Polar': \"YWHKREQDNST\",\n",
    "        'Node_Small': \"VCAGDNSTP\",\n",
    "        # 'Node_Proline': \"P\",\n",
    "        'Node_Tiny': \"AGS\",\n",
    "        'Node_Aliphatic': \"ILV\",\n",
    "        'Node_Aromatic': \"FYWH\",\n",
    "        'Node_Positive': \"HKR\",\n",
    "        'Node_Negative': \"DE\",\n",
    "        'Node_Charged': \"HKRED\"\n",
    "    }\n",
    "    return properties\n",
    "\n",
    "def encode_properties(seq, properties):\n",
    "    # Initialize the dictionary for property encodings\n",
    "    property_encodings = {key: \"\" for key in properties}\n",
    "    \n",
    "    # Iterate through each amino acid in the sequence\n",
    "    for aa in seq:\n",
    "        for prop, aas in properties.items():\n",
    "            if aa in aas:\n",
    "                property_encodings[prop] += '1'\n",
    "            else:\n",
    "                property_encodings[prop] += '0'\n",
    "    \n",
    "    return property_encodings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4ac2d4f2-e1a4-4b7d-a856-c2d9207924eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Applying properties 1: 100%|█████████████| 50914/50914 [00:08<00:00, 5757.79it/s]\n",
      "Applying properties 2: 100%|█████████████| 50914/50914 [00:08<00:00, 5745.54it/s]\n",
      "Applying properties 3: 100%|█████████████| 50914/50914 [00:08<00:00, 5801.93it/s]\n",
      "Applying properties 4: 100%|█████████████| 50914/50914 [00:08<00:00, 5781.08it/s]\n",
      "Applying properties 5: 100%|█████████████| 50914/50914 [00:08<00:00, 5796.80it/s]\n",
      "Applying properties 6: 100%|█████████████| 50914/50914 [00:08<00:00, 5783.17it/s]\n",
      "Applying properties 7: 100%|█████████████| 50914/50914 [00:08<00:00, 5769.86it/s]\n",
      "Applying properties 8: 100%|█████████████| 50914/50914 [00:08<00:00, 5745.16it/s]\n",
      "Applying properties 9: 100%|█████████████| 50914/50914 [00:08<00:00, 5749.65it/s]\n",
      "Applying properties 10: 100%|████████████| 50914/50914 [00:08<00:00, 5750.77it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein_Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Node_Mask</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>...</th>\n",
       "      <th>Node_Hydrophobic</th>\n",
       "      <th>Node_Hydrophilic</th>\n",
       "      <th>Node_Polar</th>\n",
       "      <th>Node_Small</th>\n",
       "      <th>Node_Tiny</th>\n",
       "      <th>Node_Aliphatic</th>\n",
       "      <th>Node_Aromatic</th>\n",
       "      <th>Node_Positive</th>\n",
       "      <th>Node_Negative</th>\n",
       "      <th>Node_Charged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3LAF_1_A</td>\n",
       "      <td>DRWGSELESSHHHHHHGGRRSLHFVSEPSDAVTMRGGNVLLNCSAE...</td>\n",
       "      <td>0000000000000000001111111111111111111111111111...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.105...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.9999999999999999, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.037...</td>\n",
       "      <td>...</td>\n",
       "      <td>0010001000000000000001011000001101000011101010...</td>\n",
       "      <td>1100010100000000001100000011010000100100010001...</td>\n",
       "      <td>1110110111111111001110100110110010100100010101...</td>\n",
       "      <td>1001100011000000110010001101111110011110011110...</td>\n",
       "      <td>0001100011000000110010000100101000011000000110...</td>\n",
       "      <td>0000001000000000000001001000000100000011100000...</td>\n",
       "      <td>0010000000111111000000110000000000000000000000...</td>\n",
       "      <td>0100000000111111001100100000000000100000000000...</td>\n",
       "      <td>1000010100000000000000000010010000000000000001...</td>\n",
       "      <td>1100010100111111001100100010010000100000000001...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Protein_Name                                           Sequence  \\\n",
       "0     3LAF_1_A  DRWGSELESSHHHHHHGGRRSLHFVSEPSDAVTMRGGNVLLNCSAE...   \n",
       "\n",
       "                                           Node_Mask  \\\n",
       "0  0000000000000000001111111111111111111111111111...   \n",
       "\n",
       "                                                   x  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "\n",
       "                                                   y  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "\n",
       "                                                   z  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "\n",
       "                                                   A  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.105...   \n",
       "\n",
       "                                                   C  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "\n",
       "                                                   D  \\\n",
       "0  [0.9999999999999999, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "\n",
       "                                                   E  ...  \\\n",
       "0  [0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.037...  ...   \n",
       "\n",
       "                                    Node_Hydrophobic  \\\n",
       "0  0010001000000000000001011000001101000011101010...   \n",
       "\n",
       "                                    Node_Hydrophilic  \\\n",
       "0  1100010100000000001100000011010000100100010001...   \n",
       "\n",
       "                                          Node_Polar  \\\n",
       "0  1110110111111111001110100110110010100100010101...   \n",
       "\n",
       "                                          Node_Small  \\\n",
       "0  1001100011000000110010001101111110011110011110...   \n",
       "\n",
       "                                           Node_Tiny  \\\n",
       "0  0001100011000000110010000100101000011000000110...   \n",
       "\n",
       "                                      Node_Aliphatic  \\\n",
       "0  0000001000000000000001001000000100000011100000...   \n",
       "\n",
       "                                       Node_Aromatic  \\\n",
       "0  0010000000111111000000110000000000000000000000...   \n",
       "\n",
       "                                       Node_Positive  \\\n",
       "0  0100000000111111001100100000000000100000000000...   \n",
       "\n",
       "                                       Node_Negative  \\\n",
       "0  1000010100000000000000000010010000000000000001...   \n",
       "\n",
       "                                        Node_Charged  \n",
       "0  1100010100111111001100100010010000100000000001...  \n",
       "\n",
       "[1 rows x 37 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "df_features = processed_df#.copy()\n",
    "\n",
    "# Get amino acid properties\n",
    "props = amino_acid_properties()\n",
    "\n",
    "# Encode properties for each sequence\n",
    "for i, prop in enumerate(props):\n",
    "    tqdm.pandas(desc=f\"Applying properties {i+1}\")\n",
    "    df_features[prop] = df_features['Sequence'].progress_apply(lambda x: encode_properties(x, props)[prop])\n",
    "\n",
    "df_features.rename(columns={#\"Sequence\": \"Node_Sequence\",\n",
    "                            \"Mask\": \"Node_Mask\",\n",
    "                            \"ID\": \"Protein_Name\",}, inplace=True)\n",
    "df_features.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10a2c303-4f4f-46f9-b6d6-8a0e79543148",
   "metadata": {},
   "source": [
    "## Functions to compute global edge interactions (between all nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1ad62bfd-8af2-4092-a127-ea4cb30289d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from tqdm import tqdm\n",
    "\n",
    "def calculate_distances(coords):\n",
    "    \"\"\"Calculate the Euclidean distance matrix for a set of coordinates.\"\"\"\n",
    "    return squareform(pdist(coords, 'euclidean'))\n",
    "\n",
    "def calculate_centroid(coordinates):\n",
    "    \"\"\"Calculate the centroid from a list of 3D coordinates.\"\"\"\n",
    "    return np.mean(coordinates, axis=0)\n",
    "\n",
    "def calculate_residue_distance_matrix(centroid_list):\n",
    "    \"\"\"Calculate the Euclidean distance matrix for a list of centroids.\"\"\"\n",
    "    return squareform(pdist(centroid_list, 'euclidean'))\n",
    "\n",
    "def calculate_sequence_distance(seq_length):\n",
    "    \"\"\"Calculate the sequence distance matrix for a protein sequence.\"\"\"\n",
    "    indices = np.arange(seq_length)\n",
    "    seq_dist = np.abs(indices[:, None] - indices[None, :])\n",
    "    return seq_dist\n",
    "\n",
    "def calculate_contact_map(dist_matrix, threshold=16.0):\n",
    "    \"\"\"Calculate the contact map for a distance matrix given a threshold, ignoring distances of 0.\"\"\"\n",
    "    contact_map = (dist_matrix <= threshold).astype(int)\n",
    "    contact_map[dist_matrix == 0] = 0\n",
    "    return contact_map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4b833c1-071f-4425-8b43-7e1b0f8064f6",
   "metadata": {},
   "source": [
    "## Functions to compute local edge interactions (between neighboring nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "979ac273-5c15-4fc8-8a37-65b5674719f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def conservation_interaction_matrix(mask):\n",
    "    \"\"\"Generate a matrix where entries are 1 if both residues are conserved, 0 otherwise.\"\"\"\n",
    "    n = len(mask)\n",
    "    conservation_matrix = np.zeros((n, n), dtype=int)\n",
    "    for i in range(1, n):  # Start from 1 to be able to access i-1\n",
    "        if mask[i] == '1' and mask[i-1] == '1':\n",
    "            conservation_matrix[i, i-1] = 1  # Interaction with previous residue\n",
    "        if i < n - 1 and mask[i] == '1' and mask[i+1] == '1':\n",
    "            conservation_matrix[i, i+1] = 1  # Interaction with next residue\n",
    "    return conservation_matrix\n",
    "\n",
    "def calculate_charge_interactions(sequence):\n",
    "    \"\"\"Calculate electrostatic interaction matrix for a given amino acid sequence.\"\"\"\n",
    "    n = len(sequence)\n",
    "    charge_matrix = np.zeros((n, n), dtype=int)\n",
    "    charge_map = {'D': -1, 'E': -1, 'K': 1, 'R': 1, 'H': 1}  # Example simplified charges\n",
    "\n",
    "    for i in range(1, n):\n",
    "        if sequence[i] in charge_map and sequence[i-1] in charge_map:\n",
    "            if charge_map[sequence[i]] * charge_map[sequence[i-1]] == -1:\n",
    "                charge_matrix[i, i-1] = 1\n",
    "        if i < n - 1 and sequence[i] in charge_map and sequence[i+1] in charge_map:\n",
    "            if charge_map[sequence[i]] * charge_map[sequence[i+1]] == -1:\n",
    "                charge_matrix[i, i+1] = 1\n",
    "\n",
    "    return charge_matrix\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def assign_interaction_group(aa):\n",
    "    \"\"\"Assign interaction group to amino acids.\"\"\"\n",
    "    hydrophobic = {'L', 'I', 'M', 'P', 'V', 'A', 'G', 'F', 'W'}\n",
    "    hydrophilic = {'D', 'E', 'H', 'K', 'R', 'Q', 'N', 'C', 'T', 'S'}\n",
    "    if aa in hydrophobic: return -1  # Hydrophobic\n",
    "    elif aa in hydrophilic: return 1  # Hydrophilic\n",
    "    return 0  # Neither\n",
    "\n",
    "def calculate_water_interactions(sequence):\n",
    "    \"\"\"Calculate the interaction matrix for immediate neighbors in the sequence.\"\"\"\n",
    "    n = len(sequence)\n",
    "    interaction_scores = np.array([assign_interaction_group(aa) for aa in sequence])\n",
    "    \n",
    "    # Create an empty matrix with zeros\n",
    "    interaction_matrix = np.zeros((n, n), dtype=int)\n",
    "\n",
    "    # Only set values for immediate neighbors\n",
    "    for i in range(1, n):\n",
    "        if interaction_scores[i] == interaction_scores[i-1] and interaction_scores[i] != 0:\n",
    "            interaction_matrix[i, i-1] = interaction_scores[i]\n",
    "            interaction_matrix[i-1, i] = interaction_scores[i]  # Ensure symmetry\n",
    "\n",
    "    return interaction_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b8e1b36-21ed-46aa-be6a-1c7da15521e6",
   "metadata": {},
   "source": [
    "## apply all functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c53e08c3-56e1-43f6-9e9b-fa9d2ace383e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating x coords: 100%|██████████████| 50914/50914 [00:07<00:00, 6744.29it/s]\n",
      "Calculating y coords: 100%|██████████████| 50914/50914 [00:15<00:00, 3222.88it/s]\n",
      "Calculating z coords: 100%|██████████████| 50914/50914 [00:08<00:00, 6233.52it/s]\n",
      "Calculating zipped coords: 100%|████████| 50914/50914 [00:02<00:00, 22101.76it/s]\n"
     ]
    }
   ],
   "source": [
    "# rescale distances to Angstroms\n",
    "tqdm.pandas(desc=\"Calculating x coords\")\n",
    "df_features['x'] = df_features['x'].progress_apply(lambda x: [round(val / 100, 3) for val in x])\n",
    "tqdm.pandas(desc=\"Calculating y coords\")\n",
    "df_features['y'] = df_features['y'].progress_apply(lambda x: [round(val / 100, 3) for val in x])\n",
    "tqdm.pandas(desc=\"Calculating z coords\")\n",
    "df_features['z'] = df_features['z'].progress_apply(lambda x: [round(val / 100, 3) for val in x])\n",
    "tqdm.pandas(desc=\"Calculating zipped coords\")\n",
    "df_features['Atom_Coordinates'] = df_features.progress_apply(lambda row: list(zip(row['x'], row['y'], row['z'])), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "942a9a13-2fbe-49ce-a22c-1a92eac118d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating x coords: 100%|██████████████| 25299/25299 [00:03<00:00, 7201.35it/s]\n",
      "Calculating y coords: 100%|██████████████| 25299/25299 [00:03<00:00, 7205.87it/s]\n",
      "Calculating z coords: 100%|██████████████| 25299/25299 [00:03<00:00, 7173.25it/s]\n",
      "Calculating zipped coords: 100%|████████| 25299/25299 [00:00<00:00, 29881.12it/s]\n",
      "Calculating 3D atom distance matrices:  41%|▍| 10301/25299 [00:36<00:51, 289.01it"
     ]
    }
   ],
   "source": [
    "# Calculate atom distance matrix and add to dataframe\n",
    "tqdm.pandas(desc=\"Calculating 3D atom distance matrices\")\n",
    "df_features['Atom_3D_dists'] = df_features['Atom_Coordinates']\\\n",
    "                                 .progress_apply(lambda coords: calculate_distances(np.array(coords)))\n",
    "\n",
    "# Calculate centroid for each residue\n",
    "tqdm.pandas(desc=\"Calculating residue centroids\")\n",
    "df_features['Node_centroids'] = df_features['Atom_Coordinates']\\\n",
    "        .progress_apply(lambda coords: [calculate_centroid(coords[i:i+3]) for i in range(0, len(coords), 3)])\n",
    "    \n",
    "# Calculate distance matrix between centroids\n",
    "tqdm.pandas(desc=\"Calculating residue 3D distances\")\n",
    "df_features['Edge_3D_dists_global'] = df_features['Node_centroids']\\\n",
    "        .progress_apply(lambda centroids: calculate_residue_distance_matrix(np.array(centroids)))\n",
    "\n",
    "# Calculate sequence distance matrix and add to dataframe\n",
    "tqdm.pandas(desc=\"Calculating sequence distance matrices\")\n",
    "df_features['Edge_1D_dists_global'] = df_features['Sequence']\\\n",
    "                                    .progress_apply(lambda seq: calculate_sequence_distance(len(seq)))\n",
    "\n",
    "# Calculate which other atoms are within 16A, which is close enough to make a secondary structure\n",
    "tqdm.pandas(desc=\"Calculating atom contact map matrices\")\n",
    "df_features['Atom_proximity'] = df_features['Atom_3D_dists']\\\n",
    "                                  .progress_apply(lambda dist_matrix: calculate_contact_map(dist_matrix))\n",
    "\n",
    "# Calculate which other residues are within 16A, which is close enough to make a secondary structure\n",
    "tqdm.pandas(desc=\"Calculating residue contact map matrices\")\n",
    "df_features['Edge_proximity_global'] = df_features['Edge_3D_dists_global']\\\n",
    "        .progress_apply(lambda dist_matrix: calculate_contact_map(dist_matrix))\n",
    "\n",
    "# Calcluate which edges are highly conserved\n",
    "tqdm.pandas(desc=\"Calculating conservation matrices\")\n",
    "df_features['Edge_conserved_local'] = df_features['Node_Mask']\\\n",
    "                                .progress_apply(lambda mask: conservation_interaction_matrix(mask))\n",
    "\n",
    "# Calcluate charge interactions\n",
    "tqdm.pandas(desc=\"Calculating charge matrices\")\n",
    "df_features['Edge_charge_local'] = df_features['Sequence']\\\n",
    "                                .progress_apply(lambda seq: calculate_charge_interactions(seq))\n",
    "\n",
    "# Calcluate hydrophobicity interactions\n",
    "tqdm.pandas(desc=\"Calculating solubility matrices\")\n",
    "df_features['Edge_solubility_local'] = df_features['Sequence']\\\n",
    "                                .progress_apply(lambda seq: calculate_water_interactions(seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fd4e42a-7736-4f04-969c-e7821c6137ce",
   "metadata": {},
   "source": [
    "### export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "89a687ec-bacb-4c8e-8411-2c9df6da4612",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "# Assuming df is your DataFrame\n",
    "def protein_to_jsonl(row):\n",
    "    # Split atom coordinates into groups by atom type: N, CA, C\n",
    "    n_coords = row['Atom_Coordinates'][0::3]  # Assuming every third starting from 0 is N\n",
    "    ca_coords = row['Atom_Coordinates'][1::3]  # Assuming every third starting from 1 is CA\n",
    "    c_coords = row['Atom_Coordinates'][2::3]  # Assuming every third starting from 2 is C\n",
    "\n",
    "    # Create a dictionary to hold the protein data\n",
    "    protein_data = {\n",
    "        \"name\": row['Protein_Name'],\n",
    "        \"conserved\": row['Node_Mask'],\n",
    "        \"num_chains\": 1,  # Assuming all entries have one chain; adjust as necessary\n",
    "        \"seq\": row['Sequence'],\n",
    "        \"coords\": {\n",
    "            \"N\": n_coords,\n",
    "            \"CA\": ca_coords,\n",
    "            \"C\": c_coords\n",
    "        }\n",
    "    }\n",
    "    return json.dumps(protein_data)  # Convert dictionary to JSON string\n",
    "\n",
    "# Apply the function to each row and save the output as a JSONL file\n",
    "jsonl_data = df_features.apply(protein_to_jsonl, axis=1)\n",
    "\n",
    "# Write to a JSONL file\n",
    "with open('training_95_processed.jsonl', 'w') as outfile:\n",
    "    for entry in jsonl_data:\n",
    "        outfile.write(entry + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7574cabc-4b1f-4792-beca-4a3df1ad8a9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shuffle the DataFrame\n",
    "df_shuffled = df_features.sample(frac=1).reset_index(drop=True)\n",
    "\n",
    "# Calculate split indices\n",
    "train_end = int(len(df_shuffled) * 0.7)\n",
    "val_end = train_end + int(len(df_shuffled) * 0.15)\n",
    "\n",
    "# Split the DataFrame into train, validation, and test\n",
    "train_proteins = df_shuffled.iloc[:train_end]['Protein_Name'].tolist()\n",
    "val_proteins = df_shuffled.iloc[train_end:val_end]['Protein_Name'].tolist()\n",
    "test_proteins = df_shuffled.iloc[val_end:]['Protein_Name'].tolist()\n",
    "\n",
    "# Create dictionary to be written as JSON\n",
    "split_dict = {\n",
    "    'train': train_proteins,\n",
    "    'validation': val_proteins,\n",
    "    'test': test_proteins\n",
    "}\n",
    "\n",
    "# Write to JSON file\n",
    "with open('training_95_split.json', 'w') as json_file:\n",
    "    json.dump(split_dict, json_file, indent=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a41941e-3939-4bff-bb03-5e48725d1482",
   "metadata": {},
   "source": [
    "### convert strings to arrays and map amino acids to numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d9772a54-44bb-42cd-a38c-05620414bef8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_strings_to_char_arrays(df, columns):\n",
    "    aa_to_int = {'A': 1, 'C': 2, 'D': 3, 'E': 4, 'F': 5, 'G': 6, 'H': 7, 'I': 8, 'K': 9,\n",
    "                 'L': 10, 'M': 11, 'N': 12, 'P': 13, 'Q': 14, 'R': 15, 'S': 16, 'T': 17,\n",
    "                 'V': 18, 'W': 19, 'Y': 20}\n",
    "    # Apply conversion to the 'Sequence' column specifically\n",
    "    df['Sequence'] = df['Sequence'].apply(lambda seq: [aa_to_int[aa] for aa in seq])\n",
    "    \n",
    "    # Define a vectorized function to split strings into arrays of characters\n",
    "    vectorized_split = np.vectorize(list)\n",
    "    # Apply the vectorized function to the specified columns\n",
    "    for column in columns:\n",
    "        df[column] = df[column].apply(vectorized_split)\n",
    "    return df\n",
    "    \n",
    "# Specify the columns you want to convert\n",
    "columns_to_convert = ['Node_Mask', 'Node_Hydrophobic', 'Node_Hydrophilic',\n",
    "                      'Node_Polar', 'Node_Small', 'Node_Tiny', 'Node_Aliphatic',          \n",
    "                      'Node_Aromatic', 'Node_Positive', 'Node_Negative', 'Node_Charged']          \n",
    "\n",
    "# Convert the specified columns\n",
    "df_features = convert_strings_to_char_arrays(df_features, columns_to_convert)\n",
    "df_features.drop(columns=['x','y','z'],inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bafc3cbd-d076-48cc-be13-5b09fbd7830b",
   "metadata": {},
   "source": [
    "## Current features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "10b073c4-6b38-4733-808b-b6d165b9be0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Protein_Name                                                   90#2WXZ_2_C\n",
       "Sequence                 [3, 15, 18, 20, 8, 7, 13, 5, 7, 10, 10, 20, 20...\n",
       "Node_Mask                [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...\n",
       "A                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.012...\n",
       "C                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.002...\n",
       "D                        [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...\n",
       "E                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...\n",
       "F                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.016...\n",
       "G                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...\n",
       "H                        [0.0, 0.0, 0.0, 0.021224489795918365, 0.0, 1.0...\n",
       "I                        [0.0, 0.0, 0.012965050732807213, 0.0, 0.512043...\n",
       "K                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.011...\n",
       "L                        [0.0, 0.0, 0.02367531003382187, 0.0, 0.0, 0.0,...\n",
       "M                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...\n",
       "N                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.237...\n",
       "P                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, ...\n",
       "Q                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.012...\n",
       "R                        [0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...\n",
       "S                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.040...\n",
       "T                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.021...\n",
       "V                        [0.0, 0.0, 0.9633596392333709, 0.0, 0.48795669...\n",
       "W                        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...\n",
       "Y                        [0.0, 0.0, 0.0, 0.9787755102040815, 0.0, 0.0, ...\n",
       "Info                     [0.9999999988691709, 0.9999999988691709, 0.939...\n",
       "Node_Hydrophobic         [0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, ...\n",
       "Node_Hydrophilic         [1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, ...\n",
       "Node_Polar               [1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, ...\n",
       "Node_Small               [1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, ...\n",
       "Node_Tiny                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, ...\n",
       "Node_Aliphatic           [0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, ...\n",
       "Node_Aromatic            [0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, ...\n",
       "Node_Positive            [0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, ...\n",
       "Node_Negative            [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...\n",
       "Node_Charged             [1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, ...\n",
       "Atom_Coordinates         [(0.0, 0.0, 0.0), (0.0, 0.0, 0.0), (0.0, 0.0, ...\n",
       "Atom_3D_dists            [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...\n",
       "Node_centroids           [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, ...\n",
       "Edge_3D_dists_global     [[0.0, 0.0, 0.0, 0.0, 0.0, 50.39993083989964, ...\n",
       "Edge_1D_dists_global     [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13...\n",
       "Atom_proximity           [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...\n",
       "Edge_proximity_global    [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...\n",
       "Edge_conserved_local     [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...\n",
       "Edge_charge_local        [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...\n",
       "Edge_solubility_local    [[0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features.iloc[0].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "eab49b0b-b147-4014-b109-994338deeb1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Protein_Name', 'Sequence', 'Node_Mask', 'A', 'C', 'D', 'E', 'F', 'G',\n",
       "       'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y',\n",
       "       'Info', 'Node_Hydrophobic', 'Node_Hydrophilic', 'Node_Polar',\n",
       "       'Node_Small', 'Node_Tiny', 'Node_Aliphatic', 'Node_Aromatic',\n",
       "       'Node_Positive', 'Node_Negative', 'Node_Charged', 'Atom_Coordinates',\n",
       "       'Atom_3D_dists', 'Node_centroids', 'Edge_3D_dists_global',\n",
       "       'Edge_1D_dists_global', 'Atom_proximity', 'Edge_proximity_global',\n",
       "       'Edge_conserved_local', 'Edge_charge_local', 'Edge_solubility_local'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3780b7ba-0763-4e42-86c3-c6f4cae2e364",
   "metadata": {},
   "source": [
    "## model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d9b511d0-bd99-4084-aa20-dec5e4d195a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install torch torch-geometric ogb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d818654a-4c6a-42cc-aca2-0ee4d7f74361",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "from torch_geometric.data import Data, DataLoader\n",
    "from ogb.graphproppred import PygGraphPropPredDataset, Evaluator\n",
    "from torch_geometric.nn import GCNConv, global_mean_pool\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.preprocessing import StandardScaler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ecde2517-6563-479c-b18a-6f4ce4aee02f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df_features.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8997a2f6-d368-429a-a8ad-bd7f3b0fcce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def df_to_graph(row):\n",
    "    # node features (assuming they are already properly scaled and do not include the raw sequence anymore)\n",
    "    features = np.vstack([ row['Sequence'],\n",
    "        row['A'], row['C'], row['D'], row['E'], row['F'],\n",
    "        row['G'], row['H'], row['I'], row['K'], row['L'],\n",
    "        row['M'], row['N'], row['P'], row['Q'], row['R'],\n",
    "        row['S'], row['T'], row['V'], row['W'], row['Y'],\n",
    "        row['Node_Hydrophobic'], row['Node_Hydrophilic'],\n",
    "        row['Node_Polar'], row['Node_Small'], row['Node_Tiny'],\n",
    "        row['Node_Aliphatic'], row['Node_Aromatic'], row['Node_Positive'],\n",
    "        row['Node_Negative'], row['Node_Charged'], row['Info'],\n",
    "    ]).astype(float).T  # Ensure all data is float\n",
    "\n",
    "    # Combine sequence encoding with additional features\n",
    "    node_features = torch.tensor(features, dtype=torch.float)\n",
    "\n",
    "    # Edge indices for consecutive residues\n",
    "    edge_index = torch.tensor(np.vstack([np.arange(len(row['Sequence']) - 1), np.arange(1, len(row['Sequence']))]), dtype=torch.long)\n",
    "\n",
    "    # Edge attributes\n",
    "    edge_attributes = torch.tensor(np.vstack([\n",
    "        row['Edge_1D_dists_global'],\n",
    "        row['Edge_conserved_local'],\n",
    "        row['Edge_charge_local'],\n",
    "        row['Edge_solubility_local']\n",
    "    ]).T, dtype=torch.float)  # Transpose to align features with edges\n",
    "\n",
    "    # Target (residue centroids)\n",
    "    target = torch.tensor(row['Node_centroids'], dtype=torch.float)\n",
    "\n",
    "    return Data(x=node_features, edge_index=edge_index, edge_attr=edge_attr, y=target)\n",
    "\n",
    "# Apply this function to a DataFrame row by row\n",
    "# graphs = df.apply(df_to_graph, axis=1)  # This might still cause an error due to inconsistent feature lengths across rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "80a6f74c-35c3-4b4c-997d-aa58fc2bdb21",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'edge_attr' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[23], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Example of applying this function to your DataFrame\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m graphs \u001b[38;5;241m=\u001b[39m df\u001b[38;5;241m.\u001b[39mapply(df_to_graph, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n",
      "File \u001b[0;32m/opt/homebrew/anaconda3/lib/python3.11/site-packages/pandas/core/frame.py:10034\u001b[0m, in \u001b[0;36mDataFrame.apply\u001b[0;34m(self, func, axis, raw, result_type, args, by_row, **kwargs)\u001b[0m\n\u001b[1;32m  10022\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcore\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mapply\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m frame_apply\n\u001b[1;32m  10024\u001b[0m op \u001b[38;5;241m=\u001b[39m frame_apply(\n\u001b[1;32m  10025\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m  10026\u001b[0m     func\u001b[38;5;241m=\u001b[39mfunc,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m  10032\u001b[0m     kwargs\u001b[38;5;241m=\u001b[39mkwargs,\n\u001b[1;32m  10033\u001b[0m )\n\u001b[0;32m> 10034\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m op\u001b[38;5;241m.\u001b[39mapply()\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mapply\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m/opt/homebrew/anaconda3/lib/python3.11/site-packages/pandas/core/apply.py:837\u001b[0m, in \u001b[0;36mFrameApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    834\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw:\n\u001b[1;32m    835\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_raw()\n\u001b[0;32m--> 837\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_standard()\n",
      "File \u001b[0;32m/opt/homebrew/anaconda3/lib/python3.11/site-packages/pandas/core/apply.py:965\u001b[0m, in \u001b[0;36mFrameApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    964\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mapply_standard\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 965\u001b[0m     results, res_index \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_series_generator()\n\u001b[1;32m    967\u001b[0m     \u001b[38;5;66;03m# wrap results\u001b[39;00m\n\u001b[1;32m    968\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwrap_results(results, res_index)\n",
      "File \u001b[0;32m/opt/homebrew/anaconda3/lib/python3.11/site-packages/pandas/core/apply.py:981\u001b[0m, in \u001b[0;36mFrameApply.apply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    978\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m option_context(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmode.chained_assignment\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    979\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m i, v \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(series_gen):\n\u001b[1;32m    980\u001b[0m         \u001b[38;5;66;03m# ignore SettingWithCopy here in case the user mutates\u001b[39;00m\n\u001b[0;32m--> 981\u001b[0m         results[i] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfunc(v, \u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mkwargs)\n\u001b[1;32m    982\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(results[i], ABCSeries):\n\u001b[1;32m    983\u001b[0m             \u001b[38;5;66;03m# If we have a view on v, we need to make a copy because\u001b[39;00m\n\u001b[1;32m    984\u001b[0m             \u001b[38;5;66;03m#  series_generator will swap out the underlying data\u001b[39;00m\n\u001b[1;32m    985\u001b[0m             results[i] \u001b[38;5;241m=\u001b[39m results[i]\u001b[38;5;241m.\u001b[39mcopy(deep\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "Cell \u001b[0;32mIn[22], line 31\u001b[0m, in \u001b[0;36mdf_to_graph\u001b[0;34m(row)\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[38;5;66;03m# Target (residue centroids)\u001b[39;00m\n\u001b[1;32m     29\u001b[0m target \u001b[38;5;241m=\u001b[39m torch\u001b[38;5;241m.\u001b[39mtensor(row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNode_centroids\u001b[39m\u001b[38;5;124m'\u001b[39m], dtype\u001b[38;5;241m=\u001b[39mtorch\u001b[38;5;241m.\u001b[39mfloat)\n\u001b[0;32m---> 31\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m Data(x\u001b[38;5;241m=\u001b[39mnode_features, edge_index\u001b[38;5;241m=\u001b[39medge_index, edge_attr\u001b[38;5;241m=\u001b[39medge_attr, y\u001b[38;5;241m=\u001b[39mtarget)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'edge_attr' is not defined"
     ]
    }
   ],
   "source": [
    "# Example of applying this function to your DataFrame\n",
    "graphs = df.apply(df_to_graph, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a40176d6-ca42-47ee-95fb-fe8c929e29f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.iloc[20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24d75bd1-cda7-48cc-b7e6-26285d4f2cd1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0006588f-c339-4e9c-93d4-5406a5b55f87",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bb05472-4cd7-42ac-b5f1-1bcb88ac12fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f97db4ff-f756-4a19-9503-8b8d8f1a6487",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "04d7ef28-1b68-43b8-9e8e-2e993b2fd87a",
   "metadata": {},
   "source": [
    "### plotting functions from features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2785ed09-e0e9-439b-81d4-e0acdd3b3a52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy as np\n",
    "# import matplotlib.pyplot as plt\n",
    "# import seaborn as sns\n",
    "\n",
    "# def plot_distance_distribution_specific_atoms(dist_matrix, start_index=630, end_index=666):\n",
    "#     \"\"\"\n",
    "#     Plot the distribution of distances between a specific range of atoms in a protein.\n",
    "    \n",
    "#     Args:\n",
    "#     - dist_matrix (numpy.ndarray): The distance matrix of a protein.\n",
    "#     - start_index (int): Start index of the atom range.\n",
    "#     - end_index (int): End index of the atom range (inclusive).\n",
    "#     \"\"\"\n",
    "#     # Extract the specific submatrix for the desired range\n",
    "#     sub_matrix = dist_matrix[start_index:end_index+1, start_index:end_index+1]\n",
    "    \n",
    "#     # Flatten the upper triangle of the submatrix, ignoring diagonal (distance 0)\n",
    "#     distances = sub_matrix[np.triu_indices_from(sub_matrix, k=1)]\n",
    "    \n",
    "#     # Plot the distribution of distances\n",
    "#     plt.figure(figsize=(10, 6))\n",
    "#     sns.histplot(distances, bins=50, kde=True)\n",
    "#     plt.title(f'Distribution of Distances Between Atoms {start_index} to {end_index}')\n",
    "#     plt.xlabel('Distance (Å)')\n",
    "#     plt.ylabel('Frequency')\n",
    "#     plt.grid(True)\n",
    "#     plt.show()\n",
    "\n",
    "# # Example usage:\n",
    "# # Assuming 'dist_matrix' is already defined and contains the distances for all atoms\n",
    "# plot_distance_distribution_specific_atoms(dist_matrix)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a987e45-6969-485a-bb93-aa424a28e051",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy as np\n",
    "# import matplotlib.pyplot as plt\n",
    "# import seaborn as sns\n",
    "\n",
    "# def plot_distance_distribution_specific_atoms(dist_matrix, start_index=0, end_index=5000):\n",
    "#     \"\"\"\n",
    "#     Plot the distribution of distances between a specific range of atoms in a protein.\n",
    "    \n",
    "#     Args:\n",
    "#     - dist_matrix (numpy.ndarray): The distance matrix of a protein.\n",
    "#     - start_index (int): Start index of the atom range.\n",
    "#     - end_index (int): End index of the atom range (inclusive).\n",
    "#     \"\"\"\n",
    "#     # Extract the specific submatrix for the desired range\n",
    "#     sub_matrix = dist_matrix[start_index:end_index+1, start_index:end_index+1]\n",
    "    \n",
    "#     # Flatten the upper triangle of the submatrix, ignoring diagonal (distance 0)\n",
    "#     distances = sub_matrix[np.triu_indices_from(sub_matrix, k=1)]\n",
    "    \n",
    "#     # Plot the distribution of distances\n",
    "#     plt.figure(figsize=(4, 3))\n",
    "#     sns.histplot(distances, bins=2,)\n",
    "#     plt.title(f'Distribution of Distances Between Atoms {start_index} to {end_index}')\n",
    "#     plt.xlabel('Distance (Å)')\n",
    "#     plt.ylabel('Frequency')\n",
    "#     plt.grid(True)\n",
    "#     plt.show()\n",
    "\n",
    "# # Example usage:\n",
    "# # Assuming 'dist_matrix' is already defined and contains the distances for all atoms\n",
    "# plot_distance_distribution_specific_atoms(df_features['Edge_contact_map'].iloc[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "779ef7be-7a2b-48fa-bfe0-ebf39f3d417c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# contact matrix visualizer\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "def plot_section_of_array(array, start_row, start_col, max):\n",
    "    \"\"\"\n",
    "    Plot a 20x20 section of a numpy 2D array starting from specified row and column indices.\n",
    "    \n",
    "    Args:\n",
    "    - array (numpy.ndarray): The 2D array to visualize.\n",
    "    - start_row (int): The starting row index for the 20x20 section.\n",
    "    - start_col (int): The starting column index for the 20x20 section.\n",
    "    \"\"\"\n",
    "    if array.shape[0] < start_row + max or array.shape[1] < start_col + max:\n",
    "        raise ValueError(\"Array does not contain enough rows or columns for a 20x20 section from the starting index.\")\n",
    "    \n",
    "    # Extract the 20x20 section from the array\n",
    "    section = array[start_row:start_row + max, start_col:start_col + max]\n",
    "    \n",
    "    # Plot the section using imshow\n",
    "    plt.figure(figsize=(6, 6))\n",
    "    plt.imshow(section, cmap='viridis', interpolation='nearest')\n",
    "    plt.colorbar()\n",
    "    plt.title('20x20 Section of Array')\n",
    "    plt.show()\n",
    "\n",
    "# Example usage:\n",
    "# Assume 'large_array' is a numpy 2D array larger than 20x20\n",
    "mat = df_features['Edge_proximity_global'].iloc[0]\n",
    "plot_section_of_array(mat, 0, 0, len(mat))  # Adjust indices based on your array dimensions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9464cfc-ba34-44a0-bfc4-96417eed48f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "\n",
    "# def format_pdb_from_df1(row):\n",
    "#     pdb_str = \"\"\n",
    "#     atom_count = 1  # Starting index for ATOM records\n",
    "#     chain_id = 'A'  # Assuming a single chain for simplicity\n",
    "#     res_num = 1     # Residue number\n",
    "#     occupancy = 1.00\n",
    "#     t_factor = 50.00\n",
    "\n",
    "#     # Only keep the alpha carbon (CA)\n",
    "#     atom_types = ['CA']  # Only backbone alpha carbon atoms\n",
    "#     element_types = {'CA': 'C'}  # Element for alpha carbon\n",
    "    \n",
    "#     residue_mapping = {'A': 'ALA', 'R': 'ARG', 'N': 'ASN', 'D': 'ASP', 'C': 'CYS', \n",
    "#                        'E': 'GLU', 'Q': 'GLN', 'G': 'GLY', 'H': 'HIS', 'I': 'ILE', \n",
    "#                        'L': 'LEU', 'K': 'LYS', 'M': 'MET', 'F': 'PHE', 'P': 'PRO', \n",
    "#                        'S': 'SER', 'T': 'THR', 'W': 'TRP', 'Y': 'TYR', 'V': 'VAL'}\n",
    "\n",
    "#     for i in range(len(row['Sequence'])):\n",
    "#         residue = residue_mapping[row['Sequence'][i]]\n",
    "#         # We know the alpha carbon is the second atom in the list for each residue (1,4,7,...)\n",
    "#         idx = 3 * i + 1  # Update index to only point to alpha carbon\n",
    "#         x = row['x'][idx]/100\n",
    "#         y = row['y'][idx]/100\n",
    "#         z = row['z'][idx]/100\n",
    "#         element = element_types['CA']\n",
    "\n",
    "#         pdb_str += f\"ATOM  {atom_count:>5}  {atom_types[0]:<2}  {residue:>3} {chain_id}{res_num:>4}    {x:>8.3f}{y:>8.3f}{z:>8.3f}{occupancy:>6.2f}{t_factor:>6.2f}          {element:>2}  \\n\"\n",
    "#         atom_count += 1\n",
    "#         res_num += 1  # Increment residue number for each new amino acid\n",
    "\n",
    "#     return pdb_str\n",
    "\n",
    "# test = format_pdb_from_df1(protein_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "be9622b0-dbcb-4c24-9b12-f2dc0de1b4f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "def load_json_to_dict(filepath):\n",
    "    \"\"\"\n",
    "    Load JSON data from a file into a dictionary.\n",
    "\n",
    "    Args:\n",
    "    filepath (str): The path to the JSON file.\n",
    "\n",
    "    Returns:\n",
    "    dict: The dictionary containing the JSON data.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        with open(filepath, 'r') as file:\n",
    "            data_dict = json.load(file)\n",
    "        print(\"JSON data successfully loaded.\")\n",
    "        return data_dict\n",
    "    except FileNotFoundError:\n",
    "        print(\"The file was not found.\")\n",
    "        return None\n",
    "    except json.JSONDecodeError:\n",
    "        print(\"Error decoding JSON from the file.\")\n",
    "        return None\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {str(e)}\")\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "aaf89547-0cce-4195-8513-a0424b07a103",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JSON data successfully loaded.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys(['test', 'cath_nodes', 'train', 'validation'])"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example usage\n",
    "# Assuming the JSON structure is a list of objects\n",
    "json_filepath = '../neurips19-graph-protein-design/data/cath/chain_set_splits.json'\n",
    "sp = load_json_to_dict(json_filepath)\n",
    "sp.keys()\n",
    "# df_from_string = json_string_to_df(json_string)\n",
    "# print(\"DataFrame from JSON string:\")\n",
    "# print(df_from_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4572c8cb-11df-4fdf-883e-816672d864cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "def remove_key_from_jsonl(input_file_path, output_file_path, key_to_remove):\n",
    "    \"\"\"\n",
    "    Reads a JSONL file, removes a specified key from the 'coords' dictionary within each JSON object, and writes the result to a new JSONL file.\n",
    "\n",
    "    Args:\n",
    "    input_file_path (str): Path to the input JSONL file.\n",
    "    output_file_path (str): Path to the output JSONL file where the modified data will be saved.\n",
    "    key_to_remove (str): The key to remove from the 'coords' dictionary of each JSON object.\n",
    "    \"\"\"\n",
    "    with open(input_file_path, 'r') as input_file, open(output_file_path, 'w') as output_file:\n",
    "        for line in input_file:\n",
    "            data = json.loads(line)  # Convert JSON string to dictionary\n",
    "            if key_to_remove in data['coords']:\n",
    "                del data['coords'][key_to_remove]  # Remove the key from the 'coords' dictionary\n",
    "            output_file.write(json.dumps(data) + '\\n')  # Convert dictionary back to JSON string and write to output file\n",
    "\n",
    "\n",
    "# Usage example\n",
    "input_file_path = '../neurips19-graph-protein-design/data/cath/chain_set.jsonl'\n",
    "output_file_path = '../neurips19-graph-protein-design/data/cath/chain_set_updated.jsonl'\n",
    "remove_key_from_jsonl(input_file_path, output_file_path, 'O')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "31130379-76c0-41ce-98fe-fe31acfc1e2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "remove_set = {'2D6F_3_C', '2I7X_1_A', '4UDX_1_X', '3JBX_2_B', '4R4X_1_A', '3NG9_1_A', '3OHN_1_A', '5HCC_1_B', '1TMO_1_A', '1C96_1_A', '2QO3_1_A', '3GZS_1_A', '3EN9_1_A', '4OM9_1_A', '5C6G_1_A', '4DDU_1_A', '4KSI_1_A', '3FIM_1_B', '2D2O_1_A', '5DFZ_5_D', '4OUR_2_B', '3SAM_1_A', '4ZAJ_1_A', '3BGA_1_A', '3G25_1_A', '3WEO_1_A', '4RH3_1_A', '4G26_1_A', '3JZY_1_A', '4K7H_1_A', '2QYU_1_A', '4QNL_1_A', '4QBD_1_A', '4KF7_1_A', '5E5N_1_A', '5B2Q_1_A', '4CT0_1_A', '4GAA_1_A', '2QSF_1_A', '3U4J_1_A', '1KDG_1_A', '4HAW_3_C', '5DZT_1_A', '2AHV_1_A', '4CRM_1_P', '4K0J_1_A', '5AIJ_1_A', '2XFN_1_A', '3L0Q_1_A', '2XFR_1_A', '4BHW_1_A', '2VSE_1_A', '5IXQ_1_A', '4BOT_4_E', '4B2T_8_Z', '4TXI_1_A', '3UA4_1_A', '4U1C_1_C', '3UMV_1_A', '1KL7_1_A', '3IYL_1_A', '2C5G_1_A', '2GKS_1_A', '4WVM_1_A', '3QVP_1_A', '3WPU_1_A', '3FOT_1_A', '5GAS_8_N', '1EDQ_1_A', '4ZCF_2_C', '4OXI_1_A', '5FIR_1_A', '1ON9_1_A', '5EAD_1_A', '3AQP_1_A', '4DWS_4_D', '2QE7_1_A', '5ISU_1_A', '4UAQ_1_A', '3JB9_22_c', '4A82_1_A', '4C3X_1_A', '5A01_1_A', '3UOX_1_A', '3DJL_1_A', '1R6V_1_A', '5BN3_1_A', '4YBQ_1_A', '3J3R_2_A', '4V1T_1_A', '4QYR_1_A', '5EJB_1_B', '1W7P_3_D', '1TR2_1_A', '4ZIU_1_A', '4OK4_1_A', '5FJA_1_A', '3JCN_33_b', '1L9N_1_A', '1ZPU_1_A', '2WSD_1_A', '2DE0_1_X', '4H3S_1_A', '3PUK_1_A', '4EPL_1_A', '3AYJ_1_A', '4Y28_2_B', '5EPC_1_A', '5ELP_1_C', '1QVR_1_A', '1HN0_1_A', '4XMN_2_B', '4CMR_1_A', '3W0L_2_B', '5I6I_1_A', '3WXL_1_A', '2ZXR_1_A', '5C78_1_A', '3QGK_2_C', '4K17_1_A', '2VNU_2_D', '3J3I_1_A', '3FN9_1_A', '4BXZ_13_X', '5AM2_1_A', '2YEV_1_A', '4UPH_1_A', '3J26_1_A', '1GPE_1_A', '4B7O_1_A', '3OKY_2_B', '3BMX_1_A', '3JCR_7_B', '1ITZ_1_A', '2WQD_1_A', '2BVL_1_A', '4JDC_1_A', '4UFC_1_A', '2ZTG_1_A', '3IHG_1_A', '1YGY_1_A', '3IFQ_1_A', '2IUF_1_A', '5BV2_1_P', '4V5T_1_AA', '3CUZ_1_A', '3PUI_1_A', '4H5Y_1_A', '4ZW9_1_A', '1MTY_1_D', '5A1Y_2_C', '1Q57_1_A', '1Y4U_1_A', '4AT2_1_A', '2FOK_1_A', '4R0Z_1_A', '2E6K_1_A', '4TLM_2_B', '3J9V_1_b', '4J2D_1_A', '3AYX_1_A', '4UM2_1_A', '2A91_1_A', '3VIU_1_A', '1PIX_1_A', '3EDY_1_A', '3VA7_1_A', '3WE0_1_A', '5I6H_1_A', '2ZKM_1_X', '4UM8_2_B', '2MYS_1_A', '1H41_1_A', '2Z0F_1_A', '1X1N_1_A', '5DFM_1_A', '3B5X_1_A', '2YHE_1_A', '1TG7_1_A', '2VZS_1_A', '3PVM_2_B', '1Z1N_1_X', '4IV9_1_A', '4J2A_d4j2aa2', '4GUA_1_A', '3V2A_1_R', '3EPO_1_A', '3TCH_1_A', '1OHF_1_A', '2V9K_1_A', '1NTL_1_A', '1D0N_1_A', '3AK5_1_A', '3G4D_1_A', '4D04_1_A', '1IOK_1_A', '2WYO_1_A', '1YRZ_1_A', '2XZO_1_A', '4YDD_1_A', '5E9T_1_A', '2VXO_1_A', '4GZU_1_A', '1YQT_1_A', '4AM6_1_A', '4ZLF_1_A', '4NUF_1_A', '4AIE_1_A', '5A9Q_6_5', '4OO1_10_J', '2FGY_1_A', '2VEQ_1_A', '5B04_5_I', '4JE5_2_D', '4AD8_1_A', '4GY7_1_A', '2O1U_1_A', '1BF2_1_A', '4DXC_1_A', '4RF6_1_A', '4FYT_1_A', '3IVE_1_A', '1U6G_3_C', '3GAW_1_A', '1RH1_1_A', '4APN_1_A', '3IZO_2_F', '1R7A_1_A', '4FWG_1_A', '1D8Y_2_A', '2WVX_2_B', '3FHN_1_A', '5A9Q_4_3', '3X0V_1_A', '4DG6_1_A', '4MH1_1_A', '4FF3_1_A', '3NWA_1_B', '1JW1_1_A', '4IGL_2_B', '3LPP_1_A', '4QQW_1_A', '3BVU_1_A', '3BB0_1_A', '5A2R_1_A', '4WJ4_1_A', '4YPJ_1_A', '4R8A_1_A', '2XQ0_1_A', '4PQG_1_A', '1S58_1_A', '2DKH_1_A', '1W0P_1_A', '5D51_1_L', '1A4S_1_A', '3WOO_1_A', '1IV8_1_A', '4FUS_1_A', '3JAI_87_jj', '3D45_1_A', '3IZY_2_P', '3FHH_1_A', '4RJK_1_A', '4G9I_1_A', '2Z5U_1_A', '4OR0_1_A', '1UXV_1_A', '3MMP_2_G', '4FDD_1_A', '3WA2_1_X', '4FJQ_1_A', '5FL8_43_s', '3RCN_1_A', '4Z64_1_A', '2H1N_1_A', '5D0O_1_A', '2CXN_1_A', '3ILV_1_A', '3BQ6_1_A', '1YIQ_1_A', '3RA4_1_A', '1S0E_d1s0ea3', '2CKW_1_A', '2XQF_1_A', '4RGW_1_A', '4RPE_1_A', '4AP3_1_A', '4E57_1_A', '5HDT_1_A', '5DHK_1_A', '3CMX_3_A', '5I0N_1_A', '5CPS_1_A', '3EFM_1_A', '1V02_2_E', '4CJ0_1_A', '4QAW_1_A', '4QIW_2_B', '3ZUS_1_A', '1EA9_1_C', '1MW9_1_X', '3FY4_1_A', '3HX6_1_A', '4XHC_1_A', '1Q5C_1_A', '4AEF_1_A', '2RDY_1_A', '5CLW_1_A', '4RBN_1_A', '1E07_1_A', '4F4C_1_A', '4B2T_7_Q', '4I3G_1_A', '4WWX_1_B', '3GF3_1_A', '4WHJ_1_A', '4AUR_1_A', '2GP4_1_A', '4B9Q_1_A', '2NVO_1_A', '3VQT_1_A', '3U7Q_2_B', '2XE4_1_A', '1W07_1_A', '4HEA_13_L', '5FP2_1_A', '3OPB_1_A', '4BET_1_A', '2JGP_1_A', '4QFL_1_A', '5HB3_1_A', '4XWM_1_A', '1CHU_1_A', '3TTY_1_A', '2FHF_1_A', '4AUM_1_A', '3I4G_1_A', '5GAM_3_C', '4B2T_1_A', '4K6M_1_A', '3J9D_1_A', '3L4G_2_B', '3QYA_1_A', '2WTB_1_A', '4O5H_1_A', '2NVU_2_B', '2GJ4_1_A', '4PDX_d4pdxb1', '3ZXK_1_A', '1QI9_1_A', '1N62_2_B', '4PJ1_1_A', '3NAF_1_A', '3JSZ_1_A', '1EWQ_3_A', '3N2O_1_A', '1X9D_1_A', '1OAO_1_A', '4PF1_1_A', '2XWU_2_B', '1YQ2_1_A', '2WU2_1_A', '1DMS_1_A', '3ODU_1_A', '3HRZ_4_D', '2PFD_1_A', '4YD9_1_A', '4B3J_1_A', '3HWC_1_A', '2C42_1_A', '4BDV_1_A', '1S49_1_A', '5BP1_1_A', '4M5D_1_A', '2DU3_2_A', '2VQI_1_A', '5EY9_1_A', '5IRE_1_A', '1WDK_1_A', '1QO8_1_A', '5F1Q_1_A', '3J92_44_u', '2Q7Z_1_A', '3RDE_1_A', '1W7C_1_A', '4YE6_1_A', '1QH8_2_B', '2BX2_1_L', '4C00_1_A', '2OBE_1_A', '2EAB_1_A', '2WQ7_1_A', '3MT5_1_A', '4CWU_3_O', '1XDP_1_A', '3QA8_1_A', '1I7D_2_A', '3KIC_d3kict-', '4KRE_1_A', '5FGU_1_A', '4BQE_1_A', '5CXF_1_A', '1N7V_1_A', '2Z5L_1_A', '4FNM_1_A', '1G0D_1_A', '3C5E_1_A', '5IL7_1_A', '3N6R_1_A', '4B63_1_A', '4NBQ_1_A', '3UJZ_1_A', '4MC5_1_A', '4YFB_3_C', '4B7H_1_A', '3A1K_1_A', '3ZUK_1_A', '2FSH_1_A', '3MKQ_1_A', '3SDR_1_A', '3QDK_1_A', '2BPO_1_A', '1G8M_1_A', '3ITE_1_A', '2XN2_1_A', '3V83_1_A', '4RUL_1_A', '4RHB_1_A', '4EPA_1_A', '3H4Z_1_A', '3NSJ_1_A', '1RQB_1_A', '5FSG_1_A', '3OSR_1_A', '3BF0_1_A', '1SY7_1_A', '3UFB_1_A', '2WEU_1_A', '4MH8_1_A', '3LPF_1_A', '1FLG_1_A', '4QTO_1_A', '4LC9_1_A', '5AB0_1_A', '1FCB_1_A', '3MMP_1_A', '3IXW_1_A', '5DNC_1_A', '1ULV_1_A', '1RT8_1_A', '3KK7_1_A', '1M1Z_1_A', '4FL3_1_A', '4YE5_1_A', '1K4Y_1_A', '2CN3_1_A', '3P4P_1_A', '5AN8_1_A', '4RWF_1_A', '4OUA_2_B', '4KVL_1_A', '4QIW_1_A', '2W54_2_B', '5IKL_1_B', '5HAX_1_A', '4UHV_1_A', '3F6K_1_A', '1LLA_1_A', '5EAN_1_A', '5IJO_7_J', '5DFZ_2_C', '1K1X_1_A', '4QME_1_A', '2NLK_1_A', '1PGU_1_A', '2I0K_1_A', '4BY6_1_A'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "340786c9-1233-455a-87ac-cce27b4f3851",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated JSON file saved.\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "# Path to your JSON file\n",
    "json_file_path = '/Users/daniel/Dropbox/USF/generative/neurips19-graph-protein-design/data/cath/chain_set_splits.json'\n",
    "\n",
    "# Read the JSON data from the file\n",
    "with open(json_file_path, 'r') as file:\n",
    "    data = json.load(file)\n",
    "\n",
    "# Filter out the entries in the remove_set from each key\n",
    "for key in ['train', 'test', 'validation']:\n",
    "    data[key] = [item for item in data[key] if item not in remove_set]\n",
    "\n",
    "# Write the modified data back to the file\n",
    "with open(json_file_path, 'w') as file:\n",
    "    json.dump(data, file, indent=4)\n",
    "\n",
    "print(\"Updated JSON file saved.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b57a8fa-1977-4fea-96e2-1b99c176d21b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
